<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="David N. Nicholson" />
  <meta name="author" content="Daniel S. Himmelstein" />
  <meta name="author" content="Casey S. Greene" />
  <meta name="dcterms.date" content="2019-07-29" />
  <meta name="keywords" content="machine learning, weak supervision, natural language processing, heterogenous netowrks" />
  <title>Mining Heterogenous Relationships from Pubmed Abstracts Using Weak Supervision</title>
  <style>
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Mining Heterogenous Relationships from Pubmed Abstracts Using Weak Supervision</h1>
</header>
<p><small><em> This manuscript (<a href="https://greenelab.github.io/text_mined_hetnet_manuscript/v/77307b470aa2d92f7d31b4a5eebc6658154a5dc7/">permalink</a>) was automatically generated from <a href="https://github.com/greenelab/text_mined_hetnet_manuscript/tree/77307b470aa2d92f7d31b4a5eebc6658154a5dc7">greenelab/text_mined_hetnet_manuscript@77307b4</a> on July 29, 2019. </em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>David N. Nicholson</strong><br> <img src="images/orcid.svg" alt="ORCID icon" class="inline_icon" /> <a href="https://orcid.org/0000-0003-0002-5761">0000-0003-0002-5761</a> · <img src="images/github.svg" alt="GitHub icon" class="inline_icon" /> <a href="https://github.com/danich1">danich1</a><br> <small> Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania · Funded by GBMF4552 </small></p></li>
<li><p><strong>Daniel S. Himmelstein</strong><br> <img src="images/orcid.svg" alt="ORCID icon" class="inline_icon" /> <a href="https://orcid.org/0000-0002-3012-7446">0000-0002-3012-7446</a> · <img src="images/github.svg" alt="GitHub icon" class="inline_icon" /> <a href="https://github.com/dhimmel">dhimmel</a> · <img src="images/twitter.svg" alt="Twitter icon" class="inline_icon" /> <a href="https://twitter.com/dhimmel">dhimmel</a><br> <small> Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania · Funded by GBMF4552 </small></p></li>
<li><p><strong>Casey S. Greene</strong><br> <img src="images/orcid.svg" alt="ORCID icon" class="inline_icon" /> <a href="https://orcid.org/0000-0001-8713-9213">0000-0001-8713-9213</a> · <img src="images/github.svg" alt="GitHub icon" class="inline_icon" /> <a href="https://github.com/cgreene">cgreene</a> · <img src="images/twitter.svg" alt="Twitter icon" class="inline_icon" /> <a href="https://twitter.com/GreeneScientist">GreeneScientist</a><br> <small> Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania · Funded by GBMF4552 and R01 HG010067 </small></p></li>
</ul>
<h2 id="abstract" class="page_break_before">Abstract</h2>
<p>Knowledge bases support multiple research efforts including providing contextual information for biomedical entities, constructing networks, and supporting the interpretation of high-throughput analyses. Some knowledge bases are automatically constructed, but most are populated via some form of manual curation. Manual curation is time consuming and difficult to scale in the context of an increasing publication rate. A recently described “data programming” paradigm seeks to circumvent this arduous process by combining distant supervision with simple rules and heuristics written as labeling functions that can be automatically applied to inputs. Unfortunately writing useful label functions requires substantial error analysis and is a non trivial task: in early efforts to use data programming we found that producing each label function could take a few days. Producing a biomedical knowledge base with multiple node and edge types could take hundreds or thousands of label functions. In this paper we sought to evaluate the extent to which label functions could be re-used across edge types. We used a subset of Hetionet v1 that centered on disease, compound, and gene nodes to evaluate this approach. We compare a baseline distant supervision model with the same distant supervision resources added to edge-type-specific label functions, edge-type-mismatch label functions, and all label functions. We confirmed that adding additional edge-type-specific label functions improves performance. We also found that adding one or a few edge-type-mismatch label functions also nearly always improves performance. Adding a large number of edge-type-mismatch label functions produces more variable performance that depends on the edge type being predicted and the edge type that is the source of the label function. Lastly, we show that this approach, even on this subgraph of Hetionet, could certain novel edges to Hetionet v1 with high confidence. We expect that its use in practice would include additional filtering and scoring methods which would further enhance precision.</p>
<h2 id="introduction">Introduction</h2>
<p>Knowledge bases are important resources that hold complex structured and unstructed information. These resources have been used in important tasks such as network analysis for drug repurposing discovery <span class="citation" data-cites="u8pIAt5j bPvC638e O21tn8vf">[<a href="#ref-u8pIAt5j" role="doc-biblioref">1</a>,<a href="#ref-bPvC638e" role="doc-biblioref">2</a>,<a href="#ref-O21tn8vf" role="doc-biblioref">3</a>]</span> or as a source of training labels for text mining systems <span class="citation" data-cites="EHeTvZht CVHSURuI HS4ARwmZ">[<a href="#ref-EHeTvZht" role="doc-biblioref">4</a>,<a href="#ref-CVHSURuI" role="doc-biblioref">5</a>,<a href="#ref-HS4ARwmZ" role="doc-biblioref">6</a>]</span>. Populating knowledge bases often requires highly-trained scientists to read biomedical literature and summarize the results <span class="citation" data-cites="N1Ai0gaI">[<a href="#ref-N1Ai0gaI" role="doc-biblioref">7</a>]</span>. This manual curation process requires a significant amount of effort and time: in 2007 researchers estimated that filling in the missing annotations at that point would require approximately 8.4 years <span class="citation" data-cites="UdzvLgBM">[<a href="#ref-UdzvLgBM" role="doc-biblioref">8</a>]</span>). The rate of publications has continued to increase exponentially <span class="citation" data-cites="1DBISRlwN">[<a href="#ref-1DBISRlwN" role="doc-biblioref">9</a>]</span>. This has been recognized as a considerable challenge and leads to gaps in knowledge bases <span class="citation" data-cites="UdzvLgBM">[<a href="#ref-UdzvLgBM" role="doc-biblioref">8</a>]</span>.<br />
Relationship extraction has been studied as a solution towards handling this problem <span class="citation" data-cites="N1Ai0gaI">[<a href="#ref-N1Ai0gaI" role="doc-biblioref">7</a>]</span>. This process consists of creating a machine learning system to automatically scan and extract relationships from textual sources. Machine learning methods often leverage a large corpus of well-labeled training data, which still requires manual curation. Distant supervision is one technique to sidestep the requirement of well-annotated sentences: with distant supervision one makes the assumption that that all sentences containing an entity pair found in a selected database provide evidence for a relationship <span class="citation" data-cites="EHeTvZht">[<a href="#ref-EHeTvZht" role="doc-biblioref">4</a>]</span>. Distant supervision provides many labeled examples; however it is accompanied by a decrease in the quality of the labels.<br />
Ratner et al. <span class="citation" data-cites="5Il3kN32">[<a href="#ref-5Il3kN32" role="doc-biblioref">10</a>]</span> recently introduced “data programming” as a solution. Data programming combines distant supervision with the automated labeling of text using hand-written label functions. The distant supervision sources and label functions are integrated using a noise aware generative model, which is used to produce training labels. Combining distant supervision with label functions can dramatically reduce the time required to acquire sufficient training data. However, constructing a knowledge base of heterogeneous relationships through this framework still requires tens of hand-written label functions for each relationship type. Writing useful label functions requires significant error analysis, which can be a time-consuming process.</p>
<p>In this paper, we aim to address the question: to what extent can label functions be re-used across different relationship types? We hypothesized that sentences describing one relationship type may share information in the form of keywords or sentence structure with sentences that indicate other relationship types. We designed a series of experiments to determine the extent to which label function re-use enhanced performance over distant supervision alone. We examine relationships that indicate similar types of physical interactions (i.e., gene-binds-gene and compound-binds-gene) as well as different types (i.e., disease-associates-gene and compound-treats-disease). The re-use of label functions could dramatically reduce the number required to generate and update a heterogeneous knowledge graph.</p>
<h2 id="recent-work">Recent Work</h2>
<p>Talk about what has been done in the field in regards to text mining and knowledge base integration</p>
<style> 
span.gene_color { color:#02b3e4 } 
span.disease_color { color:#875442 }
</style>
<h2 id="materials-and-methods">Materials and Methods</h2>
<h3 id="hetionet">Hetionet</h3>
<p>Hetionet <span class="citation" data-cites="O21tn8vf">[<a href="#ref-O21tn8vf" role="doc-biblioref">3</a>]</span> is a large heterogenous network that contains pharmacological and biological information. This network depicts information in the form of nodes and edges of different types: nodes that represent biological and pharmacological entities and edges which represent relationships between entities. Hetionet v1.0 contains 47,031 nodes with 11 different data types and 2,250,197 edges that represent 24 different relationship types (Figure <a href="#fig:hetionet">1</a>). Edges in Hetionet were obtained from open databases, such as the GWAS Catalog <span class="citation" data-cites="16cIDAXhG">[<a href="#ref-16cIDAXhG" role="doc-biblioref">11</a>]</span> and DrugBank <span class="citation" data-cites="1FI8iuYiQ">[<a href="#ref-1FI8iuYiQ" role="doc-biblioref">12</a>]</span>. For this project, we analyzed performance over a subset of the Hetionet relationship types: disease associates with a gene (DaG), compound binds to a gene (CbG), gene interacts with gene (GiG) and compound treating a disease (CtD).</p>
<figure>
<img src="images/figures/hetionet/metagraph_highlighted_edges.png" alt="Figure 1: A metagraph (schema) of Hetionet where pharmacological, biological and disease entities are represented as nodes and the relationships between them are represented as edges. This project only focuses on the information shown in bold; however, we can extend this work to incorporate the faded out information as well." id="fig:hetionet" /><figcaption><span>Figure 1:</span> A metagraph (schema) of Hetionet where pharmacological, biological and disease entities are represented as nodes and the relationships between them are represented as edges. This project only focuses on the information shown in bold; however, we can extend this work to incorporate the faded out information as well.</figcaption>
</figure>
<h3 id="dataset">Dataset</h3>
<p>We used PubTator <span class="citation" data-cites="13vw5RIy4">[<a href="#ref-13vw5RIy4" role="doc-biblioref">13</a>]</span> as input to our analysis. PubTator provides MEDLINE abstracts that have been annotated with well-established entity recognition tools including DNorm <span class="citation" data-cites="vtuZ3Wx7">[<a href="#ref-vtuZ3Wx7" role="doc-biblioref">14</a>]</span> for disease mentions, GeneTUKit <span class="citation" data-cites="4S2HMNpa">[<a href="#ref-4S2HMNpa" role="doc-biblioref">15</a>]</span> for gene mentions, Gnorm <span class="citation" data-cites="1AkC7QdyP">[<a href="#ref-1AkC7QdyP" role="doc-biblioref">16</a>]</span> for gene normalizations and a dictionary based look system for compound mentions <span class="citation" data-cites="r501gnuM">[<a href="#ref-r501gnuM" role="doc-biblioref">17</a>]</span>. We downloaded PubTator on June 30, 2017, at which point it contained 10,775,748 abstracts. Then we filtered out mention tags that were not contained in hetionet. We used the Stanford CoreNLP parser <span class="citation" data-cites="RQkLuc5t">[<a href="#ref-RQkLuc5t" role="doc-biblioref">18</a>]</span> to tag parts of speech and generate dependency trees. We extracted sentences with two or more mentions, termed candidate sentences. Each candidates sentence was stratified by co-mention pair to produce a training set, tuning set and a testing set (shown in Table <a href="#tbl:candidate-sentences">1</a>). Each unique co-mention pair is sorted into four categories: (1) in hetionet and has sentences, (2) in hetionet and doesn’t have sentences, (3) not in hetionet and does have sentences and (4) not in hetionet and doesn’t have sentences. Within these four categories each pair receives their own individual partition rank (continuous number between 0 and 1). Any rank lower than 0.7 is sorted into training set, while any rank greater than 0.7 and lower than 0.9 is assigned to tuning set. The rest of the pairs with a rank greater than or equal to 0.9 is assigned to the test set. Sentences that contain more than one co-mention pair are treated as multiple individual candidates. We hand labeled five hundred to a thousand candidate sentences of each relationship to obtain to obtain a ground truth set (Table <a href="#tbl:candidate-sentences">1</a>, <a href="https://github.com/greenelab/text_mined_hetnet_manuscript/tree/master/supplementary_materials/annotated_sentences">dataset</a>).</p>
<a name="tbl:candidate-sentences"></a>
<table>
<caption><span>Table 1:</span> Statistics of Candidate Sentences. We sorted each candidate sentence into a training, tuning and testing set. Numbers in parentheses show the number of positives and negatives that resulted from the hand-labeling process. </caption>
<thead>
<tr class="header">
<th style="text-align: left;">Relationship</th>
<th style="text-align: center;">Train</th>
<th style="text-align: center;">Tune</th>
<th style="text-align: center;">Test</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Disease Associates Gene</td>
<td style="text-align: center;">2.35 M</td>
<td style="text-align: center;">31K (397+, 603-)</td>
<td style="text-align: center;">313K (351+, 649-)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Compound Binds Gene</td>
<td style="text-align: center;">1.7M</td>
<td style="text-align: center;">468K (37+, 463-)</td>
<td style="text-align: center;">227k (31+, 469-)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Compound Treats Disease</td>
<td style="text-align: center;">1.013M</td>
<td style="text-align: center;">96K (96+, 404-)</td>
<td style="text-align: center;">32K (112+, 388-)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Gene Interacts Gene</td>
<td style="text-align: center;">12.6M</td>
<td style="text-align: center;">1.056M (60+, 440-)</td>
<td style="text-align: center;">257K (76+, 424-)</td>
</tr>
</tbody>
</table>
<h3 id="label-functions-for-annotating-sentences">Label Functions for Annotating Sentences</h3>
<p>A common challenge in natural language processing is having too few ground truth annotations, even when textual data are abundant. Data programming circumvents this issue by quickly annotating large datasets by using multiple noisy signals emitted by label functions <span class="citation" data-cites="5Il3kN32">[<a href="#ref-5Il3kN32" role="doc-biblioref">10</a>]</span>. Label functions are simple pythonic functions that emit: a positive label (1), a negative label (-1) or abstain from emitting a label (0). We combine these functions using a generative model to output a single annotation, which is a consensus probability score bounded between 0 (low chance of mentioning a relationship) and 1 (high chance of mentioning a relationship). We used these annotations to train a discriminator model that makes the final classification step. Our label functions fall into three categories: databases, text patterns and domain heuristics. We provide examples for the categories, described below, using the following candidate sentence: “<span class="gene_color">PTK6</span> may be a novel therapeutic target for <span class="disease_color">pancreatic cancer</span>.”</p>
<p><strong>Databases</strong>: These label functions incorporate existing databases to generate a signal, as seen in distant supervision <span class="citation" data-cites="EHeTvZht">[<a href="#ref-EHeTvZht" role="doc-biblioref">4</a>]</span>. These functions detect if a candidate sentence’s co-mention pair is present in a given database. If the pair is present, emit a positive label and abstain otherwise. If the pair isn’t present in any existing database, then a separate label function will emit a negative label. We use a separate label function to prevent the label imbalance problem. This problem occurs when candidates, that scarcely appear in databases, are drowned out by negative labels. The multitude of negative labels increases the likelihood of misclassification when training the generative model.</p>
<p><span class="math display">\[ \Lambda_{DB}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 1 &amp; (\color{#875442}{D}, \color{#02b3e4}{G}) \in DB \\
0 &amp; otherwise \\
\end{cases} \]</span></p>
<p><span class="math display">\[ \Lambda_{\neg DB}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 -1 &amp; (\color{#875442}{D}, \color{#02b3e4}{G}) \notin DB \\
0 &amp; otherwise \\
\end{cases} \]</span></p>
<p><strong>Text Patterns</strong>: These label functions are designed to use keywords and sentence context to generate a signal. For example, a label function could focus on the number of words between two mentions or focus on the grammatical structure of a sentence. These functions emit a positive or negative label depending on the situation. In general, those focused on keywords emit positives and those focused on negation emit negatives.</p>
<p><span class="math display">\[ \Lambda_{TP}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 1 &amp; &quot;target&quot; \&gt; \in Candidate \&gt; Sentence \\
 0 &amp; otherwise \\
\end{cases} \]</span></p>
<p><span class="math display">\[ \Lambda_{TP}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 -1 &amp;   &quot;VB&quot; \&gt; \notin pos\_tags(Candidate \&gt; Sentence) \\
 0 &amp; otherwise \\
\end{cases} \]</span></p>
<p><strong>Domain Heuristics</strong>: These label functions use the other experiment results to generate a signal. For this category, we used dependency path cluster themes generated by Percha et al <span class="citation" data-cites="CSiMoOrI">[<a href="#ref-CSiMoOrI" role="doc-biblioref">19</a>]</span>. If a candidate sentence’s dependency path belongs to a previously generated cluster, then the label function will emit a positive label and abstain otherwise.</p>
<p><span class="math display">\[
\Lambda_{DH}(\color{#875442}{D}, \color{#02b3e4}{G}) = \begin{cases}
    1 &amp; Candidate \&gt; Sentence \in Cluster \&gt; Theme\\
    0 &amp; otherwise \\
    \end{cases}
\]</span></p>
<p>Roughly half of our label functions are based on text patterns, while the others are distributed across the databases and domain heuristics (Table <a href="#tbl:label-functions">2</a>).</p>
<a name="tbl:label-functions"></a>
<table>
<caption><span>Table 2:</span> The distribution of each label function per relationship. </caption>
<thead>
<tr class="header">
<th>Relationship</th>
<th style="text-align: center;">Databases (DB)</th>
<th style="text-align: center;">Text Patterns (TP)</th>
<th style="text-align: center;">Domain Heuristics (DH)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Disease associates Gene (DaG)</td>
<td style="text-align: center;">7</td>
<td style="text-align: center;">20</td>
<td style="text-align: center;">10</td>
</tr>
<tr class="even">
<td>Compound treats Disease (CtD)</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">15</td>
<td style="text-align: center;">7</td>
</tr>
<tr class="odd">
<td>Compound binds Gene (CbG)</td>
<td style="text-align: center;">9</td>
<td style="text-align: center;">13</td>
<td style="text-align: center;">7</td>
</tr>
<tr class="even">
<td>Gene interacts Gene (GiG)</td>
<td style="text-align: center;">9</td>
<td style="text-align: center;">20</td>
<td style="text-align: center;">8</td>
</tr>
</tbody>
</table>
<h3 id="training-models">Training Models</h3>
<h4 id="generative-model">Generative Model</h4>
<p>The generative model is a core part of this automatic annotation framework. It integrates multiple signals emitted by label functions and assigns a training class to each candidate sentence. This model assigns training classes by estimating the joint probability distribution of the latent true class (<span class="math inline">\(Y\)</span>) and label function signals (<span class="math inline">\(\Lambda\)</span>), <span class="math inline">\(P(\Lambda, Y)\)</span>. Assuming each label function is conditionally independent, the joint distribution is defined as follows:</p>
<p><span class="math display">\[
P(\Lambda, Y) = \frac{\exp(\sum_{i=1}^{m} \theta^{T}F_{i}(\Lambda, y))}
{\sum_{\Lambda&#39;}\sum_{y&#39;} \exp(\sum_{i=1}^{m} \theta^{T}F_{i}(\Lambda&#39;, y&#39;))}
\]</span></p>
<p>where <span class="math inline">\(m\)</span> is the number of candidate sentences, <span class="math inline">\(F\)</span> is the vector of summary statistics and <span class="math inline">\(\theta\)</span> is a vector of weights for each summary statistic. The summary statistics used by the generative model are as follows:</p>
<p><span class="math display">\[F^{Lab}_{i,j}(\Lambda, Y) = \unicode{x1D7D9}\{\Lambda_{i,j} \neq 0\}\]</span> <span class="math display">\[F^{Acc}_{i,j}(\Lambda, Y) = \unicode{x1D7D9}\{\Lambda_{i,j} = y_{i,j}\}\]</span></p>
<p><em>Lab</em> is the label function’s propensity (the frequency of a label function emitting a signal). <em>Acc</em> is the individual label function’s accuracy given the training class. This model optimizes the weights (<span class="math inline">\(\theta\)</span>) by minimizing the negative log likelihood:</p>
<p><span class="math display">\[\hat{\theta} = argmin_{\theta} -\sum_{\Lambda} log \sum_{Y}P(\Lambda, Y)\]</span></p>
<p>In the framework we used predictions from the generative model, <span class="math inline">\(\hat{Y} = P(Y \mid \Lambda)\)</span>, as training classes for our dataset <span class="citation" data-cites="9Jo1af7Z vzoBuh4l">[<a href="#ref-9Jo1af7Z" role="doc-biblioref">20</a>,<a href="#ref-vzoBuh4l" role="doc-biblioref">21</a>]</span>.</p>
<h4 id="word-embeddings">Word Embeddings</h4>
<p>Word embeddings are representations that map individual words to real valued vectors of user-specified dimensions. These embeddings have been shown to capture the semantic and syntactic information between words <span class="citation" data-cites="u5iJzbp9">[<a href="#ref-u5iJzbp9" role="doc-biblioref">22</a>]</span>. Using all candidate sentences for each individual relationship pair, we trained facebook’s fastText <span class="citation" data-cites="qUpCDz2v">[<a href="#ref-qUpCDz2v" role="doc-biblioref">23</a>]</span> to generate word embeddings. The fastText model uses a skipgram model <span class="citation" data-cites="1GhHIDxuW">[<a href="#ref-1GhHIDxuW" role="doc-biblioref">24</a>]</span> that aims to predict the context given a candidate word and pairs the model with a novel scoring function that treats each word as a bag of character n-grams. We trained this model for 20 epochs using a window size of 2 and generated 300-dimensional word embeddings. We use the optimized word embeddings to train a discriminative model.</p>
<h4 id="discriminative-model">Discriminative Model</h4>
<p>The discriminative model is a neural network, which we train to predict labels from the generative model. The expectation is that the discriminative model can learn more complete features of the text than the label functions that are used in the generative model. We used a convolutional neural network with multiple filters as our discriminative model. This network uses multiple filters with fixed widths of 300 dimensions and a fixed height of 7 (Figure <a href="#fig:convolutional_network">2</a>), because this height provided the best performance in terms of relationship classification <span class="citation" data-cites="fs8rAHoJ">[<a href="#ref-fs8rAHoJ" role="doc-biblioref">25</a>]</span>. We trained this model for 20 epochs using the adam optimizer <span class="citation" data-cites="c6d3lKFX">[<a href="#ref-c6d3lKFX" role="doc-biblioref">26</a>]</span> with a learning rate of 0.001. This optimizer used pytorch’s default parameter settings. We added a L2 penalty on the network weights to prevent overfitting. Lastly, we added a dropout layer (p=0.25) between the fully connected layer and the softmax layer.</p>
<figure>
<img src="images/figures/convolutional_neural_network/convolutional_neural_nework.png" alt="Figure 2: The architecture of the discriminative model is a convolutional neural network. We perform a convolution step using multiple filters. These filters generate a feature map that is sent into a maximum pooling layer. This layer extracts the largest feature in each of these maps. The extracted features are concatenated into a singular vector that is passed into a fully connected network. The fully connected network has 300 neurons for the first layer, 100 neurons for the second layer and 50 neurons for the last layer. From the fully connected network the last step is to generate predictions using the softmax layer." id="fig:convolutional_network" /><figcaption><span>Figure 2:</span> The architecture of the discriminative model is a convolutional neural network. We perform a convolution step using multiple filters. These filters generate a feature map that is sent into a maximum pooling layer. This layer extracts the largest feature in each of these maps. The extracted features are concatenated into a singular vector that is passed into a fully connected network. The fully connected network has 300 neurons for the first layer, 100 neurons for the second layer and 50 neurons for the last layer. From the fully connected network the last step is to generate predictions using the softmax layer.</figcaption>
</figure>
<h4 id="calibration-of-the-discriminative-model">Calibration of the Discriminative Model</h4>
<p>Often many tasks require a machine learning model to output reliable probability predictions. A model is well calibrated if the probabilities emitted from the model match the observed probabilities: a well-calibrated model that assigns a class label with 80% probability should have that class appear 80% of the time. Deep neural network models can often be poorly calibrated <span class="citation" data-cites="QJ6hYH8N rLVjMJ5l">[<a href="#ref-QJ6hYH8N" role="doc-biblioref">27</a>,<a href="#ref-rLVjMJ5l" role="doc-biblioref">28</a>]</span>. These models are usually over-confident in their predictions. As a result, we calibrated our convolutional neural network using temperature scaling. Temperature scaling uses a parameter T to scale each value of the logit vector (z) before being passed into the softmax (SM) function.</p>
<p><span class="math display">\[\sigma_{SM}(\frac{z_{i}}{T}) = \frac{\exp(\frac{z_{i}}{T})}{\sum_{i}\exp(\frac{z_{i}}{T})}\]</span></p>
<p>We found the optimal T by minimizing the negative log likelihood (NLL) of a held out validation set. The benefit of using this method is the model becomes more reliable and the accuracy of the model doesn’t change <span class="citation" data-cites="QJ6hYH8N">[<a href="#ref-QJ6hYH8N" role="doc-biblioref">27</a>]</span>.</p>
<h3 id="experimental-design">Experimental Design</h3>
<p>Being able to re-use label functions across edge types would substantially reduce the number of label functions required to extract multiple relationship types from biomedical literature. We first established a baseline by training a generative model using only distant supervision label functions designed for the target edge type. As an example, for the gene-interacts-gene edge type we used label functions that returned a <code>1</code> if the pair of genes were included in the Human Interaction database <span class="citation" data-cites="LCyCrr7W">[<a href="#ref-LCyCrr7W" role="doc-biblioref">29</a>]</span>, the iRefIndex database <span class="citation" data-cites="gtV3bOpd">[<a href="#ref-gtV3bOpd" role="doc-biblioref">30</a>]</span> or in the Incomplete Interactome database <span class="citation" data-cites="2jkcXYxN">[<a href="#ref-2jkcXYxN" role="doc-biblioref">31</a>]</span>. Then we compared models that also included text and domain-heuristic label functions. Using a sampling with replacement approach, we sampled these text and domain-heuristic label functions separately within edge types, across edge types, and from a pool of all label functions. We compared within-edge-type performance to across-edge-type and all-edge-type performance. For each edge type we sampled a fixed number of label functions consisting of five evenly-spaced numbers between one and the total number of possible label functions. We repeated this sampling process 50 times for each point. We evaluated both generative and discriminative models at each point, and we report performance of each in terms of the area under the receiver operating characteristic curve (AUROC) and the area under the precision-recall curve (AUPR).</p>
<h4 id="adding-random-noise-to-generative-model">Adding Random Noise to Generative Model</h4>
<p>We discovered in the course of this work that adding a single label function from a mismatched type would often improve the performance of the generative model (see Results). We designed an experiment to test whether adding a noisy label function also increased performance. This label function emitted a positive or negative label at varying frequencies, which were evenly spaced from zero to one. Zero is the same as distant supervision alone. We trained the generative model with these label functions added and report results in terms of AUROC and AUPR.</p>
<h2 id="results">Results</h2>
<h3 id="generative-model-using-randomly-sampled-label-functions">Generative Model Using Randomly Sampled Label Functions</h3>
<figure>
<img src="https://raw.githubusercontent.com/greenelab/snorkeling/master/figures/label_sampling_experiment/transfer_test_set_auroc.png" alt="Figure 3: Grid of Area Under the Receiver Operating Curve (AUROC) scores for each generative model trained on randomly sampled label functions. The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function is sampled from. For example, the top-left most square depicts the generative model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. The square towards the right depicts the generative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship. This pattern continues filling out the rest of the grid. The last most column consists of pooling every relationship specific label function and proceeding as above." id="fig:gen_model_auroc" /><figcaption><span>Figure 3:</span> Grid of Area Under the Receiver Operating Curve (AUROC) scores for each generative model trained on randomly sampled label functions. The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function is sampled from. For example, the top-left most square depicts the generative model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. The square towards the right depicts the generative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship. This pattern continues filling out the rest of the grid. The last most column consists of pooling every relationship specific label function and proceeding as above.</figcaption>
</figure>
<p>We added randomly sampled label functions to a baseline for each edge type to evaluate the feasibility of label function re-use. Our baseline model consisted of a generative model trained with only the edge type’s distant supervision label functions. We report the results in the form of area under the precision recall curve (AUPR) (Figure <a href="#fig:gen_model_auprc">4</a>) and area under the receiver operating curve (AUROC) (Figure <a href="#fig:gen_model_auroc">3</a>).<br />
The on-diagonal plots of figure <a href="#fig:gen_model_auprc">4</a>) and figure <a href="#fig:gen_model_auprc">4</a> show performance when edge-specific label functions are added on top of edge-specific baselines. The general trend is performance increases in this setting. The Compound-treats-Disease (CtD) edge type is a quintessential example of this trend. The baseline model starts off with an AUROC score of 52% and an AUPRC of 28%, which increase to 76% and 49% respectively as more CtD label functions are included. Disease-associates-Gene (DaG) edges have a similar trend: performance starting off with a AUROC of 56% and AUPRC of 41%, which increase to 62% and 45% respectively. Both the Compound-binds-Gene (CbG) and Gene-interacts-Gene (GiG) edges have an increasing trend but plateau after a few label functions are added.</p>
<p>The off-diagonals in figure <a href="#fig:gen_model_auprc">4</a>) and figure <a href="#fig:gen_model_auprc">4</a> show how performance varies when label functions from one edge type are added to a different edge type’s baseline. In certain cases (apparent for DaG), performance increases regardless of the edge type used for label functions. In other cases (apparent with CtD), one label function appears to improve performance; however, adding more label functions does not improve performance (AUROC) or decreases it (AUPRC). In certain cases, the source of the label functions appear to be important: for CbG edges performance decreases when using label functions from the DaG and CtD categories.</p>
<p>Our initial hypothesis was based on the idea that certain edge types capture similar physical relationships and that these cases would be particularly amenable for label function transfer. For example, Compound-binds-Gene (CbG) and Gene-interacts-Gene (GiG) both describe physical interactions. We observed that performance increased as assessed by both AUPRC and AUPRC when using label functions from the GiG edge type to predict CbG edges. A similar trend was observed when predicting the GiG edge; however, the performance differences were small for this edge type making the importance difficult to assess.<br />
The last column shows performance when sampling from all label functions. Performance increased (AUROC and AUPRC) for both DaG and CtD, when sampling from the full pool of label functions. CbG and GiG also had increased performance when one random label function was sampled, but performance decreased drastically as more label functions were added. It is possible that a small number of irrelevant label functions are able to overwhelm the distant supervision label functions in these cases.</p>
<figure>
<img src="https://raw.githubusercontent.com/greenelab/snorkeling/master/figures/label_sampling_experiment/transfer_test_set_auprc.png" alt="Figure 4: Grid of Area Under the Precision Recall Curve (AUPRC) scores for each generative model trained on randomly sampled label functions. The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function is sampled from. For example, the top-left most square depicts the generative model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. The square towards the right depicts the generative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship. This pattern continues filling out the rest of the grid. The last most column consists of pooling every relationship specific label function and proceeding as above." id="fig:gen_model_auprc" /><figcaption><span>Figure 4:</span> Grid of Area Under the Precision Recall Curve (AUPRC) scores for each generative model trained on randomly sampled label functions. The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function is sampled from. For example, the top-left most square depicts the generative model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. The square towards the right depicts the generative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship. This pattern continues filling out the rest of the grid. The last most column consists of pooling every relationship specific label function and proceeding as above.</figcaption>
</figure>
<h3 id="random-label-function-gen-model-analysis">Random Label Function Gen Model Analysis</h3>
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/f8962788e462b783be05a6dec5eec7fe0f0259e7/figures/gen_model_error_analysis/transfer_test_set_auroc.png" alt="Figure 5: A grid of area under the receiver operating curve (AUROC) for each edge type. Each plot consists of adding a single label function on top of the baseline model. This label function emits a positive (top row) or negative (bottom row) label at specified frequencies, and performance at zero is equivalent to not having a randomly emitting label function. The error bars represent 95% confidence intervals for AUROC (y-axis) at each emission frequency." id="fig:random_label_function_auroc" /><figcaption><span>Figure 5:</span> A grid of area under the receiver operating curve (AUROC) for each edge type. Each plot consists of adding a single label function on top of the baseline model. This label function emits a positive (top row) or negative (bottom row) label at specified frequencies, and performance at zero is equivalent to not having a randomly emitting label function. The error bars represent 95% confidence intervals for AUROC (y-axis) at each emission frequency.</figcaption>
</figure>
<p>We observed that including one label function of a mismatched type to distant supervision often improved performance, so we evaluated the effects of adding a random label function in the same setting. We found that adding random noise did not usually improve performance (Figures <a href="#fig:random_label_function_auprc">6</a> and <a href="#fig:random_label_function_auprc">6</a>). For the CbG edge type we did observe slightly increased performance via AUPR (Figure <a href="#fig:random_label_function_auprc">6</a>). However, in general the performance changes were smaller than those observed with mismatched label types.</p>
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/f8962788e462b783be05a6dec5eec7fe0f0259e7/figures/gen_model_error_analysis/transfer_test_set_auprc.png" alt="Figure 6: A grid of area under the precision recall curve (AUPR) for each edge type. Each plot consists of adding a single label function on top of the baseline model. This label function emits a positive (top row) or negative (bottom row) label at specified frequencies. The error bars represent 95% confidence intervals for AUPR (y-axis) at emission frequency." id="fig:random_label_function_auprc" /><figcaption><span>Figure 6:</span> A grid of area under the precision recall curve (AUPR) for each edge type. Each plot consists of adding a single label function on top of the baseline model. This label function emits a positive (top row) or negative (bottom row) label at specified frequencies. The error bars represent 95% confidence intervals for AUPR (y-axis) at emission frequency.</figcaption>
</figure>
<h3 id="discriminative-model-performance">Discriminative Model Performance</h3>
<p>In this framework we used a generative model trained over label functions to produce probabilistic training labels for each sentence. Then we train a discriminative model, which has full access to a representation of the text of the sentence, to predict the generated labels. The discriminative model is a convolutional neural network trained over word embeddings. We report the results of the discriminative model using AUPR (Figure <a href="#fig:discriminative_model_auprc">8</a>) and AUROC (Figure <a href="#fig:discriminative_model_auroc">7</a>).<br />
We found that the discriminative model under-performed the generative model in most cases. Only for the CtD edge does the discriminative model appear to provide performance above the generative model and that increased performance is only with modest numbers of label functions. With the full set of label functions, the performance of both remains similar. The trend observed in the generative model that one or a few mismatched label functions (off-diagonal) improves performance is retained despite the limited performance of the discriminative model.</p>
<figure>
<img src="https://raw.githubusercontent.com/greenelab/snorkeling/master/figures/label_sampling_experiment/disc_performance_test_set_auroc.png" alt="Figure 7: Grid of Area Under the Receiver Operating Curve (AUROC) scores for each discriminative model trained using generated labels from the generative models. The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function was sampled from. For example, the top-left most square depicts the discriminator model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. The error bars over the points represents the standard deviation between sampled runs. The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship. This pattern continues filling out the rest of the grid. The last most column consists of pooling every relationship specific label function and proceeding as above." id="fig:discriminative_model_auroc" /><figcaption><span>Figure 7:</span> Grid of Area Under the Receiver Operating Curve (AUROC) scores for each discriminative model trained using generated labels from the generative models. The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function was sampled from. For example, the top-left most square depicts the discriminator model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. The error bars over the points represents the standard deviation between sampled runs. The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship. This pattern continues filling out the rest of the grid. The last most column consists of pooling every relationship specific label function and proceeding as above.</figcaption>
</figure>
<figure>
<img src="https://raw.githubusercontent.com/greenelab/snorkeling/master/figures/label_sampling_experiment/disc_performance_test_set_auprc.png" alt="Figure 8: Grid of Area Under the Receiver Operating Curve (AUROC) scores for each discriminative model trained using generated labels from the generative models. The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function was sampled from. For example, the top-left most square depicts the discriminator model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. The error bars over the points represents the standard deviation between sampled runs. The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship. This pattern continues filling out the rest of the grid. The last most column consists of pooling every relationship specific label function and proceeding as above." id="fig:discriminative_model_auprc" /><figcaption><span>Figure 8:</span> Grid of Area Under the Receiver Operating Curve (AUROC) scores for each discriminative model trained using generated labels from the generative models. The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function was sampled from. For example, the top-left most square depicts the discriminator model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. The error bars over the points represents the standard deviation between sampled runs. The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship. This pattern continues filling out the rest of the grid. The last most column consists of pooling every relationship specific label function and proceeding as above.</figcaption>
</figure>
<h3 id="discriminative-model-calibration">Discriminative Model Calibration</h3>
<p>Even deep learning models with high precision and recall can be poorly calibrated, and the overconfidence of these models has been noted <span class="citation" data-cites="QJ6hYH8N rLVjMJ5l">[<a href="#ref-QJ6hYH8N" role="doc-biblioref">27</a>,<a href="#ref-rLVjMJ5l" role="doc-biblioref">28</a>]</span>. We attempted to calibrate the best performing discriminative model so that we could directly use the emitted probabilities. We examined the calibration of our existing model (Figure <a href="#fig:discriminative_model_calibration">9</a>, blue line). We found that the DaG and CtG edge types were, though not perfectly calibrated, were somewhat aligned with the ideal calibration lines. The CbG and GiG edges were poorly calibrated and increasing model certainty did not always lead to an increase in precision. Applying the calibration algorithm (orange line) did not appear to bring predictions in line with the ideal calibration line, but did capture some of the uncertainty in the GiG edge type. For this reason we use the measured precision instead of the predicted probabilities when determining how many edges could be added to existing knowledge bases with specified levels of confidence.</p>
<figure>
<img src="https://raw.githubusercontent.com/greenelab/snorkeling/master/figures/model_calibration_experiment/model_calibration.png" alt="Figure 9: Calibration plots for the discriminative model. A perfectly calibrated model would follow the dashed diagonal line. The blue line represents the predictions before calibration and the orange line shows predictions after calibration." id="fig:discriminative_model_calibration" /><figcaption><span>Figure 9:</span> Calibration plots for the discriminative model. A perfectly calibrated model would follow the dashed diagonal line. The blue line represents the predictions before calibration and the orange line shows predictions after calibration.</figcaption>
</figure>
<h2 id="reconstructing-hetionet">Reconstructing Hetionet</h2>
<p>place figure of number of new edges that can be added to hetionet as well as edges we can reconstruct using this method</p>
<h1 id="discussion">Discussion</h1>
<p>Here mention why performnace increases in the beginning for the generative model then decreases</p>
<p>Discuss discriminator model performance given generative model</p>
<p>Mention Take home messages</p>
<ol type="1">
<li>have a centralized set of negative label functions and focus more on contstructing positive label functions</li>
</ol>
<h1 id="conclusion-and-future-direction">Conclusion and Future Direction</h1>
<p>Recap the original problem - takes a long time to create useful label function</p>
<p>Proposed solution - reuse label functions</p>
<p>Mention incorporating more relationships Mention creating a centralized multitask text extractor using this method.</p>
<h2 id="references" class="page_break_before">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" role="doc-bibliography">
<div id="ref-u8pIAt5j">
<p>1. <strong>Graph Theory Enables Drug Repurposing – How a Mathematical Model Can Drive the Discovery of Hidden Mechanisms of Action</strong><br />
Ruggero Gramatica, T. Di Matteo, Stefano Giorgetti, Massimo Barbiani, Dorian Bevec, Tomaso Aste<br />
<em>PLoS ONE</em> (2014-01-09) <a href="https://doi.org/gf45zp">https://doi.org/gf45zp</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0084912">10.1371/journal.pone.0084912</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24416311">24416311</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3886994">PMC3886994</a></p>
</div>
<div id="ref-bPvC638e">
<p>2. <strong>Drug repurposing through joint learning on knowledge graphs and literature</strong><br />
Mona Alshahrani, Robert Hoehndorf<br />
<em>Cold Spring Harbor Laboratory</em> (2018-08-06) <a href="https://doi.org/gf45zk">https://doi.org/gf45zk</a><br />
DOI: <a href="https://doi.org/10.1101/385617">10.1101/385617</a></p>
</div>
<div id="ref-O21tn8vf">
<p>3. <strong>Systematic integration of biomedical knowledge prioritizes drugs for repurposing</strong><br />
Daniel Scott Himmelstein, Antoine Lizee, Christine Hessler, Leo Brueggeman, Sabrina L Chen, Dexter Hadley, Ari Green, Pouya Khankhanian, Sergio E Baranzini<br />
<em>eLife</em> (2017-09-22) <a href="https://doi.org/cdfk">https://doi.org/cdfk</a><br />
DOI: <a href="https://doi.org/10.7554/elife.26726">10.7554/elife.26726</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28936969">28936969</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5640425">PMC5640425</a></p>
</div>
<div id="ref-EHeTvZht">
<p>4. <strong>Distant supervision for relation extraction without labeled data</strong><br />
Mike Mintz, Steven Bills, Rion Snow, Dan Jurafsky<br />
<em>Proceedings of the Joint Conference of the 47th Annual Meeting of the ACL and the 4th International Joint Conference on Natural Language Processing of the AFNLP: Volume 2 - ACL-IJCNLP ’09</em> (2009) <a href="https://doi.org/fg9q43">https://doi.org/fg9q43</a><br />
DOI: <a href="https://doi.org/10.3115/1690219.1690287">10.3115/1690219.1690287</a></p>
</div>
<div id="ref-CVHSURuI">
<p>5. <strong>CoCoScore: Context-aware co-occurrence scoring for text mining applications using distant supervision</strong><br />
Alexander Junge, Lars Juhl Jensen<br />
<em>Cold Spring Harbor Laboratory</em> (2018-10-16) <a href="https://doi.org/gf45zm">https://doi.org/gf45zm</a><br />
DOI: <a href="https://doi.org/10.1101/444398">10.1101/444398</a></p>
</div>
<div id="ref-HS4ARwmZ">
<p>6. <strong>Knowledge-guided convolutional networks for chemical-disease relation extraction</strong><br />
Huiwei Zhou, Chengkun Lang, Zhuang Liu, Shixian Ning, Yingyu Lin, Lei Du<br />
<em>BMC Bioinformatics</em> (2019-05-21) <a href="https://doi.org/gf45zn">https://doi.org/gf45zn</a><br />
DOI: <a href="https://doi.org/10.1186/s12859-019-2873-7">10.1186/s12859-019-2873-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31113357">31113357</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6528333">PMC6528333</a></p>
</div>
<div id="ref-N1Ai0gaI">
<p>7. <strong>Facts from text: can text mining help to scale-up high-quality manual curation of gene products with ontologies?</strong><br />
R. Winnenburg, T. Wachter, C. Plake, A. Doms, M. Schroeder<br />
<em>Briefings in Bioinformatics</em> (2008-07-11) <a href="https://doi.org/bfsnwg">https://doi.org/bfsnwg</a><br />
DOI: <a href="https://doi.org/10.1093/bib/bbn043">10.1093/bib/bbn043</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19060303">19060303</a></p>
</div>
<div id="ref-UdzvLgBM">
<p>8. <strong>Manual curation is not sufficient for annotation of genomic databases</strong><br />
William A. Baumgartner Jr, K. Bretonnel Cohen, Lynne M. Fox, George Acquaah-Mensah, Lawrence Hunter<br />
<em>Bioinformatics</em> (2007-07-01) <a href="https://doi.org/dtck86">https://doi.org/dtck86</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btm229">10.1093/bioinformatics/btm229</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17646325">17646325</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2516305">PMC2516305</a></p>
</div>
<div id="ref-1DBISRlwN">
<p>9. <strong>Growth rates of modern science: A bibliometric analysis based on the number of publications and cited references</strong><br />
Lutz Bornmann, Rüdiger Mutz<br />
<em>Journal of the Association for Information Science and Technology</em> (2015-04-29) <a href="https://doi.org/gfj5zc">https://doi.org/gfj5zc</a><br />
DOI: <a href="https://doi.org/10.1002/asi.23329">10.1002/asi.23329</a></p>
</div>
<div id="ref-5Il3kN32">
<p>10. <strong>Data Programming: Creating Large Training Sets, Quickly</strong><br />
Alexander Ratner, Christopher De Sa, Sen Wu, Daniel Selsam, Christopher Ré<br />
<em>arXiv</em> (2016-05-25) <a href="https://arxiv.org/abs/1605.07723v3">https://arxiv.org/abs/1605.07723v3</a></p>
</div>
<div id="ref-16cIDAXhG">
<p>11. <strong>The new NHGRI-EBI Catalog of published genome-wide association studies (GWAS Catalog)</strong><br />
Jacqueline MacArthur, Emily Bowler, Maria Cerezo, Laurent Gil, Peggy Hall, Emma Hastings, Heather Junkins, Aoife McMahon, Annalisa Milano, Joannella Morales, … Helen Parkinson<br />
<em>Nucleic Acids Research</em> (2016-11-29) <a href="https://doi.org/f9v7cp">https://doi.org/f9v7cp</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkw1133">10.1093/nar/gkw1133</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27899670">27899670</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5210590">PMC5210590</a></p>
</div>
<div id="ref-1FI8iuYiQ">
<p>12. <strong>DrugBank 5.0: a major update to the DrugBank database for 2018</strong><br />
David S Wishart, Yannick D Feunang, An C Guo, Elvis J Lo, Ana Marcu, Jason R Grant, Tanvir Sajed, Daniel Johnson, Carin Li, Zinat Sayeeda, … Michael Wilson<br />
<em>Nucleic Acids Research</em> (2017-11-08) <a href="https://doi.org/gcwtzk">https://doi.org/gcwtzk</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkx1037">10.1093/nar/gkx1037</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29126136">29126136</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5753335">PMC5753335</a></p>
</div>
<div id="ref-13vw5RIy4">
<p>13. <strong>PubTator: a web-based text mining tool for assisting biocuration</strong><br />
Chih-Hsuan Wei, Hung-Yu Kao, Zhiyong Lu<br />
<em>Nucleic Acids Research</em> (2013-05-22) <a href="https://doi.org/f475th">https://doi.org/f475th</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkt441">10.1093/nar/gkt441</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23703206">23703206</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3692066">PMC3692066</a></p>
</div>
<div id="ref-vtuZ3Wx7">
<p>14. <strong>DNorm: disease name normalization with pairwise learning to rank</strong><br />
R. Leaman, R. Islamaj Dogan, Z. Lu<br />
<em>Bioinformatics</em> (2013-08-21) <a href="https://doi.org/f5gj9n">https://doi.org/f5gj9n</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btt474">10.1093/bioinformatics/btt474</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23969135">23969135</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3810844">PMC3810844</a></p>
</div>
<div id="ref-4S2HMNpa">
<p>15. <strong>GeneTUKit: a software for document-level gene normalization</strong><br />
M. Huang, J. Liu, X. Zhu<br />
<em>Bioinformatics</em> (2011-02-08) <a href="https://doi.org/dng2cb">https://doi.org/dng2cb</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btr042">10.1093/bioinformatics/btr042</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21303863">21303863</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3065680">PMC3065680</a></p>
</div>
<div id="ref-1AkC7QdyP">
<p>16. <strong>Cross-species gene normalization by species inference</strong><br />
Chih-Hsuan Wei, Hung-Yu Kao<br />
<em>BMC Bioinformatics</em> (2011-10-03) <a href="https://doi.org/dnmvds">https://doi.org/dnmvds</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-12-s8-s5">10.1186/1471-2105-12-s8-s5</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22151999">22151999</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3269940">PMC3269940</a></p>
</div>
<div id="ref-r501gnuM">
<p>17. <strong>Collaborative biocuration–text-mining development task for document prioritization for curation</strong><br />
T. C. Wiegers, A. P. Davis, C. J. Mattingly<br />
<em>Database</em> (2012-11-22) <a href="https://doi.org/gbb3zw">https://doi.org/gbb3zw</a><br />
DOI: <a href="https://doi.org/10.1093/database/bas037">10.1093/database/bas037</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23180769">23180769</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3504477">PMC3504477</a></p>
</div>
<div id="ref-RQkLuc5t">
<p>18. <strong>The Stanford CoreNLP Natural Language Processing Toolkit</strong><br />
Christopher Manning, Mihai Surdeanu, John Bauer, Jenny Finkel, Steven Bethard, David McClosky<br />
<em>Proceedings of 52nd Annual Meeting of the Association for Computational Linguistics: System Demonstrations</em> (2014) <a href="https://doi.org/gf3xhp">https://doi.org/gf3xhp</a><br />
DOI: <a href="https://doi.org/10.3115/v1/p14-5010">10.3115/v1/p14-5010</a></p>
</div>
<div id="ref-CSiMoOrI">
<p>19. <strong>A global network of biomedical relationships derived from text</strong><br />
Bethany Percha, Russ B Altman<br />
<em>Bioinformatics</em> (2018-02-27) <a href="https://doi.org/gc3ndk">https://doi.org/gc3ndk</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/bty114">10.1093/bioinformatics/bty114</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29490008">29490008</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6061699">PMC6061699</a></p>
</div>
<div id="ref-9Jo1af7Z">
<p>20. <strong>Snorkel MeTaL</strong><br />
Alex Ratner, Braden Hancock, Jared Dunnmon, Roger Goldman, Christopher Ré<br />
<em>Proceedings of the Second Workshop on Data Management for End-To-End Machine Learning - DEEM’18</em> (2018) <a href="https://doi.org/gf3xk7">https://doi.org/gf3xk7</a><br />
DOI: <a href="https://doi.org/10.1145/3209889.3209898">10.1145/3209889.3209898</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30931438">30931438</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6436830">PMC6436830</a></p>
</div>
<div id="ref-vzoBuh4l">
<p>21. <strong>Snorkel</strong><br />
Alexander Ratner, Stephen H. Bach, Henry Ehrenberg, Jason Fries, Sen Wu, Christopher Ré<br />
<em>Proceedings of the VLDB Endowment</em> (2017-11-01) <a href="https://doi.org/ch44">https://doi.org/ch44</a><br />
DOI: <a href="https://doi.org/10.14778/3157794.3157797">10.14778/3157794.3157797</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29770249">29770249</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5951191">PMC5951191</a></p>
</div>
<div id="ref-u5iJzbp9">
<p>22. <strong>Distributed Representations of Words and Phrases and their Compositionality</strong><br />
Tomas Mikolov, Ilya Sutskever, Kai Chen, Greg Corrado, Jeffrey Dean<br />
<em>arXiv</em> (2013-10-16) <a href="https://arxiv.org/abs/1310.4546v1">https://arxiv.org/abs/1310.4546v1</a></p>
</div>
<div id="ref-qUpCDz2v">
<p>23. <strong>Enriching Word Vectors with Subword Information</strong><br />
Piotr Bojanowski, Edouard Grave, Armand Joulin, Tomas Mikolov<br />
<em>arXiv</em> (2016-07-15) <a href="https://arxiv.org/abs/1607.04606v2">https://arxiv.org/abs/1607.04606v2</a></p>
</div>
<div id="ref-1GhHIDxuW">
<p>24. <strong>Efficient Estimation of Word Representations in Vector Space</strong><br />
Tomas Mikolov, Kai Chen, Greg Corrado, Jeffrey Dean<br />
<em>arXiv</em> (2013-01-16) <a href="https://arxiv.org/abs/1301.3781v3">https://arxiv.org/abs/1301.3781v3</a></p>
</div>
<div id="ref-fs8rAHoJ">
<p>25. <strong>A Sensitivity Analysis of (and Practitioners’ Guide to) Convolutional Neural Networks for Sentence Classification</strong><br />
Ye Zhang, Byron Wallace<br />
<em>arXiv</em> (2015-10-13) <a href="https://arxiv.org/abs/1510.03820v4">https://arxiv.org/abs/1510.03820v4</a></p>
</div>
<div id="ref-c6d3lKFX">
<p>26. <strong>Adam: A Method for Stochastic Optimization</strong><br />
Diederik P. Kingma, Jimmy Ba<br />
<em>arXiv</em> (2014-12-22) <a href="https://arxiv.org/abs/1412.6980v9">https://arxiv.org/abs/1412.6980v9</a></p>
</div>
<div id="ref-QJ6hYH8N">
<p>27. <strong>On Calibration of Modern Neural Networks</strong><br />
Chuan Guo, Geoff Pleiss, Yu Sun, Kilian Q. Weinberger<br />
<em>arXiv</em> (2017-06-14) <a href="https://arxiv.org/abs/1706.04599v2">https://arxiv.org/abs/1706.04599v2</a></p>
</div>
<div id="ref-rLVjMJ5l">
<p>28. <strong>Accurate Uncertainties for Deep Learning Using Calibrated Regression</strong><br />
Volodymyr Kuleshov, Nathan Fenner, Stefano Ermon<br />
<em>arXiv</em> (2018-07-01) <a href="https://arxiv.org/abs/1807.00263v1">https://arxiv.org/abs/1807.00263v1</a></p>
</div>
<div id="ref-LCyCrr7W">
<p>29. <strong>A Proteome-Scale Map of the Human Interactome Network</strong><br />
Thomas Rolland, Murat Taşan, Benoit Charloteaux, Samuel J. Pevzner, Quan Zhong, Nidhi Sahni, Song Yi, Irma Lemmens, Celia Fontanillo, Roberto Mosca, … Marc Vidal<br />
<em>Cell</em> (2014-11) <a href="https://doi.org/f3mn6x">https://doi.org/f3mn6x</a><br />
DOI: <a href="https://doi.org/10.1016/j.cell.2014.10.050">10.1016/j.cell.2014.10.050</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25416956">25416956</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4266588">PMC4266588</a></p>
</div>
<div id="ref-gtV3bOpd">
<p>30. <strong>iRefIndex: A consolidated protein interaction database with provenance</strong><br />
Sabry Razick, George Magklaras, Ian M Donaldson<br />
<em>BMC Bioinformatics</em> (2008) <a href="https://doi.org/b99bjj">https://doi.org/b99bjj</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-9-405">10.1186/1471-2105-9-405</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/18823568">18823568</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2573892">PMC2573892</a></p>
</div>
<div id="ref-2jkcXYxN">
<p>31. <strong>Uncovering disease-disease relationships through the incomplete interactome</strong><br />
J. Menche, A. Sharma, M. Kitsak, S. D. Ghiassian, M. Vidal, J. Loscalzo, A.-L. Barabasi<br />
<em>Science</em> (2015-02-19) <a href="https://doi.org/f3mn6z">https://doi.org/f3mn6z</a><br />
DOI: <a href="https://doi.org/10.1126/science.1257601">10.1126/science.1257601</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25700523">25700523</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4435741">PMC4435741</a></p>
</div>
</div>
<style> span.gene_color { color:#02b3e4 } span.disease_color { color:#875442 } span.compound_color { color:#e91e63 } </style>
<h2 id="supplemental-figures" class="page_break_before">Supplemental Figures</h2>
<h3 id="confidence-scores-for-each-edge-type">Confidence Scores for Each Edge Type</h3>
<table>
<caption>Contains the top ten Disease-associates-Gene confidence scores before and after model calbration. Disease mentions are highlighted in <span class="disease_color">brown</span> and Gene mentions are highlighted in <span class="gene_color">blue</span>.{#tbl:dg_top_ten_table}</caption>
<colgroup>
<col style="width: 7%" />
<col style="width: 3%" />
<col style="width: 80%" />
<col style="width: 4%" />
<col style="width: 4%" />
</colgroup>
<thead>
<tr class="header">
<th>Disease Name</th>
<th>Gene Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibraiton</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adrenal gland cancer</td>
<td>TP53</td>
<td>the mechanisms of adrenal tumorigenesis remain poorly established ; the r337h germline mutation in the <span class="gene_color">p53</span> gene has previously been associated with <span class="disease_color">acts</span> in brazilian children .</td>
<td>1.0</td>
<td>0.882</td>
</tr>
<tr class="even">
<td>breast cancer</td>
<td>ERBB2</td>
<td>in <span class="disease_color">breast cancer</span> , overexpression of <span class="gene_color">her2</span> is associated with an aggressive tumor phenotype and poor prognosis .</td>
<td>1.0</td>
<td>0.845</td>
</tr>
<tr class="odd">
<td>lung cancer</td>
<td>TP53</td>
<td>however , both adenine ( a ) and guanine ( g ) mutations are found in the <span class="gene_color">p53</span> gene in cr exposure-related <span class="disease_color">lung cancer</span> .</td>
<td>1.0</td>
<td>0.83</td>
</tr>
<tr class="even">
<td>malignant glioma</td>
<td>BAX</td>
<td>these data suggest that the combination of tra-8 treatment with specific overexpression of <span class="gene_color">bax</span> using advegfbax may be an effective approach for the treatment of human <span class="disease_color">malignant gliomas</span> .</td>
<td>0.999</td>
<td>0.827</td>
</tr>
<tr class="odd">
<td>polycystic ovary syndrome</td>
<td>SERPINE1</td>
<td>4 g allele in <span class="gene_color">pai-1</span> gene was more frequent in <span class="disease_color">pcos</span> and the 4g/4 g genotype was associated with increased pai-1 levels .</td>
<td>0.999</td>
<td>0.814</td>
</tr>
<tr class="even">
<td>systemic lupus erythematosus</td>
<td>PRL</td>
<td>results : <span class="disease_color">sle</span> patients showed a significantly higher serum level of <span class="gene_color">prl</span> than healthy subjects , which was especially obvious in the active stage of the disease ( p = 0.000 .</td>
<td>0.999</td>
<td>0.813</td>
</tr>
<tr class="odd">
<td>hematologic cancer</td>
<td>TNF</td>
<td>the mean <span class="gene_color">tnf-alpha</span> plasma concentration in the patients with <span class="disease_color">cll</span> was significantly higher than in the healthy control population ( 16.4 versus 8.7 pg/ml ; p &lt; .0001 ) .</td>
<td>0.999</td>
<td>0.81</td>
</tr>
<tr class="even">
<td>lung cancer</td>
<td>MUC16</td>
<td>the mean concentration of <span class="gene_color">ca 125</span> was higher in patients with <span class="disease_color">lung cancer</span> ( 37 + / - 81 u/ml ) than in those with nonmalignant disease ( 4.2 + / - 5.7 u/ml ) ( p less than 0.01 ) .</td>
<td>0.999</td>
<td>0.806</td>
</tr>
<tr class="odd">
<td>prostate cancer</td>
<td>AR</td>
<td>the <span class="gene_color">androgen receptor</span> was expressed in all primary and metastatic <span class="disease_color">prostate cancer</span> tissues and no mutations were identified .</td>
<td>0.999</td>
<td>0.801</td>
</tr>
<tr class="even">
<td>breast cancer</td>
<td>ERBB2</td>
<td>the results of multiple linear regression analysis , with her2 as the dependent variable , showed that family history of <span class="disease_color">breast cancer</span> was significantly associated with elevated <span class="gene_color">her2</span> levels in the tumors ( p = 0.0038 ) , after controlling for the effects of age , tumor estrogen receptor , and dna index .</td>
<td>0.999</td>
<td>0.8</td>
</tr>
</tbody>
</table>
<a name="tbl:dg_bottom_ten_table"></a>
<table>
<caption><span>Table 3:</span> Contains the bottom ten Disease-associates-Gene confidence scores before and after model calbration. Disease mentions are highlighted in <span class="disease_color">brown</span> and Gene mentions are highlighted in <span class="gene_color">blue</span>. </caption>
<colgroup>
<col style="width: 4%" />
<col style="width: 3%" />
<col style="width: 83%" />
<col style="width: 4%" />
<col style="width: 4%" />
</colgroup>
<thead>
<tr class="header">
<th>Disease Name</th>
<th>Gene Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibraiton</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>breast cancer</td>
<td>NAT2</td>
<td>[ the relationship between passive smoking , <span class="disease_color">breast cancer</span> risk and <span class="gene_color">n-acetyltransferase 2</span> ( nat2 ) ] .</td>
<td>0.012</td>
<td>0.287</td>
</tr>
<tr class="even">
<td>schizophrenia</td>
<td>EP300</td>
<td>ventricle size and <span class="gene_color">p300</span> in <span class="disease_color">schizophrenia</span> .</td>
<td>0.012</td>
<td>0.286</td>
</tr>
<tr class="odd">
<td>hematologic cancer</td>
<td>CD33</td>
<td>in the 2 ( nd ) study of <span class="gene_color">cd33</span> + <span class="disease_color">sr-aml</span> 2 doses of go ( 4.5 - 9 mg/m ( 2 ) ) were administered &gt; = 60d post reduced intensity conditioning ( ric ) allosct ( 8 wks apart ) .</td>
<td>0.01</td>
<td>0.281</td>
</tr>
<tr class="even">
<td>Crohn’s disease</td>
<td>PTPN2</td>
<td>in this sample , we were able to confirm an association between <span class="disease_color">cd</span> and <span class="gene_color">ptpn2</span> ( genotypic p = 0.019 and allelic p = 0.011 ) , and phenotypic analysis showed an association of this snp with late age at first diagnosis , inflammatory and penetrating cd behaviour , requirement of bowel resection and being a smoker at diagnosis .</td>
<td>0.008</td>
<td>0.268</td>
</tr>
<tr class="odd">
<td>breast cancer</td>
<td>ERBB2</td>
<td><span class="disease_color">long-term efficacy and safety of adjuvant trastuzumab for <span class="gene_color">her2-positive</span> early [breast cancer ]</span> .</td>
<td>0.007</td>
<td>0.262</td>
</tr>
<tr class="even">
<td>hematologic cancer</td>
<td>CD40LG</td>
<td>we examined the direct effect of lenalidomide on <span class="disease_color">cll-cell proliferation</span> induced by <span class="gene_color">cd154-expressing</span> accessory cells in media containing interleukin-4 and -10 .</td>
<td>0.006</td>
<td>0.259</td>
</tr>
<tr class="odd">
<td>hematologic cancer</td>
<td>MLANA</td>
<td>methods : the sln sections ( n = 214 ) were assessed by qrt assay for 4 established messenger rna biomarkers : <span class="gene_color">mart-1</span> , mage-a3 , <span class="disease_color">galnac-t</span> , and pax3 .</td>
<td>0.005</td>
<td>0.252</td>
</tr>
<tr class="even">
<td>breast cancer</td>
<td>ERBB2</td>
<td>the keywords erbb2 or her2 or erbb-2 or <span class="gene_color">her-2</span> and <span class="disease_color">breast cancer</span> and ( country ) were used to search pubmed , international and local conference abstracts and local-language journals from the year 2000 onwards .</td>
<td>0.003</td>
<td>0.225</td>
</tr>
<tr class="odd">
<td>hepatitis B</td>
<td>PKD2</td>
<td>conversely , a significant enhancement of activation was observed for afb1 in cases of mild cah and especially for <span class="gene_color">trp-p-2</span> in <span class="disease_color">hepatitis b</span> virus carriers , irrespective of their histologic diagnosis .</td>
<td>0.002</td>
<td>0.217</td>
</tr>
<tr class="even">
<td>hematologic cancer</td>
<td>C7</td>
<td>serum antibody responses to four haemophilus influenzae type b capsular polysaccharide-protein conjugate vaccines ( prp-d , hboc , <span class="gene_color">c7p</span> , and <span class="disease_color">prp-t )</span> were studied and compared in 175 infants , 85 adults and 140 2-year-old children .</td>
<td>0.002</td>
<td>0.208</td>
</tr>
</tbody>
</table>
<table>
<caption>Contains the top ten Compound-treats-Disease confidence scores after model calbration. Disease mentions are highlighted in <span class="disease_color">brown</span> and Compound mentions are highlighted in <span class="compound_color">red</span>.{#tbl:cd_top_ten_table}</caption>
<colgroup>
<col style="width: 7%" />
<col style="width: 12%" />
<col style="width: 65%" />
<col style="width: 7%" />
<col style="width: 7%" />
</colgroup>
<thead>
<tr class="header">
<th>Compound Name</th>
<th>Disease Name</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Methylprednisolone</td>
<td>asthma</td>
<td>use of tao without <span class="compound_color">methylprednisolone</span> in the treatment of severe <span class="disease_color">asthma</span> .</td>
<td>1.0</td>
<td>0.895</td>
</tr>
<tr class="even">
<td>Methyldopa</td>
<td>hypertension</td>
<td>atenolol and <span class="compound_color">methyldopa</span> in the treatment of <span class="disease_color">hypertension</span> .</td>
<td>1.0</td>
<td>0.888</td>
</tr>
<tr class="odd">
<td>Prednisone</td>
<td>asthma</td>
<td><span class="compound_color">prednisone</span> and beclomethasone for treatment of <span class="disease_color">asthma</span> .</td>
<td>1.0</td>
<td>0.885</td>
</tr>
<tr class="even">
<td>Prazosin</td>
<td>hypertension</td>
<td>experience with <span class="compound_color">prazosin</span> in the treatment of <span class="disease_color">hypertension</span> .</td>
<td>1.0</td>
<td>0.883</td>
</tr>
<tr class="odd">
<td>Prazosin</td>
<td>hypertension</td>
<td><span class="compound_color">prazosin</span> in the treatment of <span class="disease_color">hypertension</span> .</td>
<td>1.0</td>
<td>0.878</td>
</tr>
<tr class="even">
<td>Prazosin</td>
<td>hypertension</td>
<td><span class="disease_color"><span class="compound_color">prazosin</span> in the treatment of [hypertension ]</span> .</td>
<td>1.0</td>
<td>0.878</td>
</tr>
<tr class="odd">
<td>Methyldopa</td>
<td>hypertension</td>
<td>oxprenolol plus cyclopenthiazide-kcl versus <span class="compound_color">methyldopa</span> in the treatment of <span class="disease_color">hypertension</span> .</td>
<td>1.0</td>
<td>0.877</td>
</tr>
<tr class="even">
<td>Prednisolone</td>
<td>lymphatic system cancer</td>
<td>peptichemio : a new oncolytic drug in combination with vincristine and <span class="compound_color">prednisolone</span> in the treatment of <span class="disease_color">non-hodgkin lymphomas</span> .</td>
<td>1.0</td>
<td>0.871</td>
</tr>
<tr class="odd">
<td>Methyldopa</td>
<td>hypertension</td>
<td>methyldopate , the ethyl ester hydrochloride salt of <span class="compound_color">alpha-methyldopa</span> ( alpha-md ) , is used extensively in the treatment of severe <span class="disease_color">hypertension</span> .</td>
<td>1.0</td>
<td>0.851</td>
</tr>
<tr class="even">
<td>Haloperidol</td>
<td>Gilles de la Tourette syndrome</td>
<td>a comparison of pimozide and <span class="compound_color">haloperidol</span> in the treatment of gilles de la <span class="disease_color">tourette ’s syndrome</span> .</td>
<td>1.0</td>
<td>0.839</td>
</tr>
</tbody>
</table>
<table>
<caption>Contains the bottom ten Compound-treats-Disease confidence scores before and after model calbration. Disease mentions are highlighted in <span class="disease_color">brown</span> and Compound mentions are highlighted in <span class="compound_color">red</span>.{#tbl:cd_bottom_ten_table}</caption>
<colgroup>
<col style="width: 5%" />
<col style="width: 4%" />
<col style="width: 83%" />
<col style="width: 3%" />
<col style="width: 3%" />
</colgroup>
<thead>
<tr class="header">
<th>Compound Name</th>
<th>Disease Name</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Dexamethasone</td>
<td>hypertension</td>
<td><span class="compound_color">dexamethasone</span> and <span class="disease_color">hypertension</span> in preterm infants .</td>
<td>0.011</td>
<td>0.34</td>
</tr>
<tr class="even">
<td>Reserpine</td>
<td>hypertension</td>
<td><span class="compound_color">reserpine</span> in <span class="disease_color">hypertension</span> : present status .</td>
<td>0.01</td>
<td>0.336</td>
</tr>
<tr class="odd">
<td>Creatine</td>
<td>coronary artery disease</td>
<td>scintiphotographic findings were compared with the size of <span class="disease_color">myocardial infarcts</span> calculated from measurements of the activity of mb isoenzymes of <span class="compound_color">creatine</span> kinase ( ck-mb ) in serum and in the myocardium at autopsy , as described by sobel ’s method .</td>
<td>0.009</td>
<td>0.334</td>
</tr>
<tr class="even">
<td>Hydrocortisone</td>
<td>brain cancer</td>
<td>to explore the effects of repeated episodes of hypercortisolemia on <span class="disease_color">hypothalamic-pituitary-adrenal axis</span> regulation , we studied plasma acth and <span class="compound_color">cortisol</span> ( cort ) responses to 100 micrograms human crh ( hcrh ) in 10 dexamethasone ( 1.5 mg ) - pretreated elderly endurance athletes who had abstained from physical activity for at least 48 h before testing and 13 sedentary age-matched controls .</td>
<td>0.009</td>
<td>0.333</td>
</tr>
<tr class="odd">
<td>Hydrocortisone</td>
<td>brain cancer</td>
<td>basal activity of the <span class="disease_color">hypothalamic-pituitary-adrenal axis</span> was estimated by determinations of 24-h urinary free cortisol-excretion , evening basal plasma total and free <span class="compound_color">cortisol</span> concentrations , and the cortisol binding globulin-binding capacity .</td>
<td>0.008</td>
<td>0.328</td>
</tr>
<tr class="even">
<td>Creatine</td>
<td>coronary artery disease</td>
<td>during successful and uncomplicated angioplasty ( ptca ) , we studied the effect of a short lasting <span class="disease_color">myocardial ischemia</span> on plasma creatine kinase , creatine kinase mb-activity , and <span class="compound_color">creatine</span> kinase mm-isoforms ( mm1 , mm2 , mm3 ) in 23 patients .</td>
<td>0.006</td>
<td>0.318</td>
</tr>
<tr class="odd">
<td>Benzylpenicillin</td>
<td>epilepsy syndrome</td>
<td>it was shown in experiments on cats under nembutal anesthesia that a lesion of the medial forebrain bundle ( mfb ) and partly of the preoptic region at the side of local penicillin application on the cerebral cortex ( g. suprasylvius medius ) results in depression of the <span class="disease_color">epileptiform activity</span> in the <span class="compound_color">penicillin-induced</span> focus , as well as in the secondary `` mirror ’’ focus , which appeared in the symmetrical cortex area of the other hemisphere .</td>
<td>0.005</td>
<td>0.315</td>
</tr>
<tr class="even">
<td>Indomethacin</td>
<td>hypertension</td>
<td>effects of <span class="compound_color">indomethacin</span> in rabbit <span class="disease_color">renovascular hypertension</span> .</td>
<td>0.004</td>
<td>0.308</td>
</tr>
<tr class="odd">
<td>Cyclic Adenosine Monophosphate</td>
<td>ovarian cancer</td>
<td>the hormonal regulation of steroidogenesis and <span class="compound_color">adenosine 3 ’ :5 ’ - cyclic monophosphate</span> in <span class="disease_color">embryonic-chick ovary</span> .</td>
<td>0.002</td>
<td>0.292</td>
</tr>
<tr class="even">
<td>Dobutamine</td>
<td>coronary artery disease</td>
<td>two-dimensional echocardiography can detect regional wall motion abnormalities resulting from <span class="disease_color">myocardial ischemia</span> produced by <span class="compound_color">dobutamine</span> infusion .</td>
<td>0.002</td>
<td>0.287</td>
</tr>
</tbody>
</table>
<table>
<caption>Contains the top ten Compound-treats-Disease confidence scores before and after model calbration. Gene mentions are highlighted in <span class="gene_color">blue</span> and Compound mentions are highlighted in <span class="compound_color">red</span>.{#tbl:cg_top_ten_table}</caption>
<colgroup>
<col style="width: 3%" />
<col style="width: 2%" />
<col style="width: 85%" />
<col style="width: 4%" />
<col style="width: 4%" />
</colgroup>
<thead>
<tr class="header">
<th>Compound Name</th>
<th>Gene Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Hydrocortisone</td>
<td>SHBG</td>
<td>serum concentrations of testicular and adrenal androgens and androgen precursors , <span class="compound_color">cortisol</span> , unconjugated ( e1 ) and total estrone ( te1 ; greater than or equal to 85 % e1 sulfate ) , pituitary hormones , sex hormone binding globulin ( <span class="gene_color">shbg )</span> and albumin were measured in 14 male patients with non-diabetic end stage renal disease and in 28 age-matched healthy controls .</td>
<td>0.997</td>
<td>0.745</td>
</tr>
<tr class="even">
<td>Minoxidil</td>
<td>EGFR</td>
<td>direct measurement of the ability of minoxidil to compete for binding to the egf receptor indicated that <span class="compound_color">minoxidil</span> probably does not bind to the <span class="gene_color">egf receptor</span> .</td>
<td>0.99</td>
<td>0.706</td>
</tr>
<tr class="odd">
<td>Hydrocortisone</td>
<td>SHBG</td>
<td>gonadotropin , testosterone , sex hormone binding globulin ( <span class="gene_color">shbg )</span> , dehydroepiandrosterone sulphate , androstenedione , estradiol , prolactin , <span class="compound_color">cortisol</span> , thyrotropin , and free thyroxine levels were determined .</td>
<td>0.988</td>
<td>0.7</td>
</tr>
<tr class="even">
<td>Cholecalciferol</td>
<td>DBP</td>
<td><span class="compound_color">cholecalciferol</span> ( vitamin d3 ) and its 25-hydroxy metabolite are transported in plasma bound to a specific protein , the binding protein for cholecalciferol and its metabolites ( <span class="gene_color">dbp )</span> .</td>
<td>0.983</td>
<td>0.685</td>
</tr>
<tr class="odd">
<td>Indomethacin</td>
<td>AGT</td>
<td><span class="compound_color">indomethacin</span> , a potent inhibitor of prostaglandin synthesis , is known to increase the maternal blood pressure response to <span class="gene_color">angiotensin ii</span> infusion .</td>
<td>0.982</td>
<td>0.68</td>
</tr>
<tr class="even">
<td>Tretinoin</td>
<td>RXRA</td>
<td>the vitamin a derivative <span class="compound_color">retinoic acid</span> exerts its effects on transcription through two distinct classes of nuclear receptors , the retinoic acid receptor ( rar ) and the <span class="gene_color">retinoid x receptor</span> ( rxr ) .</td>
<td>0.975</td>
<td>0.668</td>
</tr>
<tr class="odd">
<td>Dopamine</td>
<td>NTS</td>
<td><span class="gene_color">neurotensin</span> binding was not modified by the addition of <span class="compound_color">dopamine</span> .</td>
<td>0.97</td>
<td>0.659</td>
</tr>
<tr class="even">
<td>D-Tyrosine</td>
<td>PLCG1</td>
<td>epidermal growth factor ( egf ) or platelet-derived growth factor binding to their receptor on fibroblasts induces tyrosine phosphorylation of plc gamma 1 and stable association of <span class="gene_color">plc gamma 1</span> with the receptor protein <span class="compound_color">tyrosine</span> kinase .</td>
<td>0.969</td>
<td>0.659</td>
</tr>
<tr class="odd">
<td>D-Tyrosine</td>
<td>PLCG1</td>
<td><span class="compound_color">tyrosine</span> phosphorylation of plc-ii was stimulated by low physiological concentrations of egf ( 1 nm ) , was quantitative , and was already maximal after a 30 sec incubation with 50 nm egf at 37 degrees c. interestingly , antibodies specific for plc-ii were able to coimmunoprecipitate the egf receptor and antibodies against egf receptor also coimmunoprecipitated <span class="gene_color">plc-ii</span> .</td>
<td>0.964</td>
<td>0.651</td>
</tr>
<tr class="even">
<td>Ketamine</td>
<td>C5</td>
<td>additionally , reduction of glycine binding by the c-5 antagonists was reversed by both nmda receptor agonists and c-7 competitive <span class="compound_color">nmda</span> antagonists , providing evidence that the site of action of these <span class="gene_color">c-5</span> antagonists is the nmda recognition site , resulting in indirect modulation of the glycine site .</td>
<td>0.957</td>
<td>0.643</td>
</tr>
</tbody>
</table>
<table>
<caption>Contains the bottom ten Compound-binds-Gene confidence scores before and after model calbration. Gene mentions are highlighted in <span class="gene_color">blue</span> and Compound mentions are highlighted in <span class="compound_color">red</span>.{#tbl:cg_bottom_ten_table}</caption>
<colgroup>
<col style="width: 3%" />
<col style="width: 2%" />
<col style="width: 84%" />
<col style="width: 4%" />
<col style="width: 4%" />
</colgroup>
<thead>
<tr class="header">
<th>Compound Name</th>
<th>Gene Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Iron</td>
<td>NDUFB3</td>
<td>since gastric acid plays an important role in the absorption process of <span class="compound_color">iron</span> and vitamin b12 , we determined levels of iron , ferritin , vitamin <span class="gene_color">b12</span> , and folic acid in 75 serum samples obtained during continuous omeprazole therapy ( 6-48 months after start of therapy ) from 34 patients with peptic diseases ( primarily reflux esophagitis ) .</td>
<td>0.006</td>
<td>0.276</td>
</tr>
<tr class="even">
<td>D-Tyrosine</td>
<td>PLAU</td>
<td>either the 55 kda u-pa form and the lower mw form ( 33 kda ) derived from the 55 kda <span class="gene_color">u-pa</span> are tyr-phosphorylated also the u-pa secreted in the culture media of human fibrosarcoma cells ( ht-1080 ) is phosphorylated in <span class="compound_color">tyrosine</span> as well as u-pa present in tissue extracts of tumors induced in nude mice by ht-1080 cells .</td>
<td>0.006</td>
<td>0.276</td>
</tr>
<tr class="odd">
<td>D-Leucine</td>
<td>POMC</td>
<td>cross-reactivities of <span class="compound_color">leucine-enkephalin</span> and <span class="gene_color">beta-endorphin</span> with the eia were less than 0.1 % , while that with gly-gly-phe-met and oxidized gly-gly-phe-met were 2.5 % and 10.2 % , respectively .</td>
<td>0.006</td>
<td>0.273</td>
</tr>
<tr class="even">
<td>Eprazinone</td>
<td>GAST</td>
<td>in patients with renal failure there exists the inhibition of the gastrin acid secretion which is the cause of the weakening of the mechanism of the feedback connection between <span class="compound_color">hcl</span> and <span class="gene_color">gastrin</span> , while because of a permanent stimulation of g-cells , the hyperplasia of these cells develops , as well as the increased secretory activity , and hypergastrinemia .</td>
<td>0.005</td>
<td>0.271</td>
</tr>
<tr class="odd">
<td>Hydrocortisone</td>
<td>GH1</td>
<td>luteinizing hormone responses to luteinizing hormone releasing hormone , and <span class="gene_color">growth hormone</span> and <span class="compound_color">cortisol</span> responses to insulin induced hypoglycaemia in functional secondary amenorrhoea .</td>
<td>0.005</td>
<td>0.271</td>
</tr>
<tr class="even">
<td>Hydrocortisone</td>
<td>GH1</td>
<td>group iv patients had normal basal levels of lh and normal lh , <span class="gene_color">gh</span> and <span class="compound_color">cortisol</span> responses .</td>
<td>0.005</td>
<td>0.269</td>
</tr>
<tr class="odd">
<td>Bupivacaine</td>
<td>AVP</td>
<td>plasma renin activity and <span class="gene_color">vasopressin</span> concentration , arterial pressure , and serum osmolality were measured in 17 patients before and after random epidural injection of either 6.7 ml of 0.75 % <span class="compound_color">bupivacaine</span> ( n = 7 ) or the same volume of saline ( n = 10 ) .</td>
<td>0.004</td>
<td>0.26</td>
</tr>
<tr class="even">
<td>Epinephrine</td>
<td>INS</td>
<td>thermogenic effect of thyroid hormones : interactions with <span class="compound_color">epinephrine</span> and <span class="gene_color">insulin</span> .</td>
<td>0.004</td>
<td>0.259</td>
</tr>
<tr class="odd">
<td>Hydrocortisone</td>
<td>GH1</td>
<td><span class="compound_color">cortisol</span> and <span class="gene_color">growth hormone</span> ( gh ) secretion ( spontaneous variations at night and the release induced by insulin hypoglycaemia ) were investigated in 69 children and adolescents .</td>
<td>0.002</td>
<td>0.241</td>
</tr>
<tr class="even">
<td>Estriol</td>
<td>LGALS1</td>
<td>[ diagnostic value of serial determination of <span class="compound_color">estriol</span> and <span class="gene_color">hpl</span> in plasma and of total estrogens in 24-h-urine compared to single values for diagnosis of fetal danger ] .</td>
<td>0.0</td>
<td>0.181</td>
</tr>
</tbody>
</table>
<table>
<caption>Contains the top ten Gene-interacts-Gene confidence scores before and after model calbration. Both gene mentions highlighted in <span class="gene_color">blue</span>.{#tbl:gg_top_ten_table}</caption>
<colgroup>
<col style="width: 4%" />
<col style="width: 4%" />
<col style="width: 79%" />
<col style="width: 6%" />
<col style="width: 5%" />
</colgroup>
<thead>
<tr class="header">
<th>Gene1 Symbol</th>
<th>Gene2 Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>INS</td>
<td>HSPA4</td>
<td>conclusions : intact insulin only weakly interacts with the <span class="gene_color">hsp70</span> chaperone dnak whereas monomeric proinsulin and peptides from 3 distinct <span class="gene_color">proinsulin</span> regions show substantial chaperone binding .</td>
<td>0.834</td>
<td>0.574</td>
</tr>
<tr class="even">
<td>NMT1</td>
<td>S100B</td>
<td>values for k ( cat ) indicated that , once gag or <span class="gene_color">nef</span> binds to the enzyme , myristoylation by <span class="gene_color">nmt1</span> and nmt2 proceeds at comparable rates .</td>
<td>0.826</td>
<td>0.571</td>
</tr>
<tr class="odd">
<td>VEGFA</td>
<td>HIF1A</td>
<td>mechanistically , we demonstrated that resveratrol inhibited <span class="gene_color">hif-1alpha</span> and <span class="gene_color">vegf</span> expression through multiple mechanisms .</td>
<td>0.82</td>
<td>0.569</td>
</tr>
<tr class="even">
<td>ITGAV</td>
<td>PECAM1</td>
<td>antigens expressed on emp and ec were assayed flow cytometrically and included constitutive markers ( <span class="gene_color">cd31</span> , <span class="gene_color">cd51/61</span> , cd105 ) , inducible markers ( cd54 , cd62e and cd106 ) , and annexin v binding .</td>
<td>0.81</td>
<td>0.566</td>
</tr>
<tr class="odd">
<td>F10</td>
<td>PF4</td>
<td>these compounds inhibit both <span class="gene_color">factor xa</span> and thrombin , in the presence of antithrombin , while they are devoid of undesirable non-specific interactions , particularly with <span class="gene_color">platelet factor 4</span> ( pf4 ) .</td>
<td>0.766</td>
<td>0.554</td>
</tr>
<tr class="even">
<td>NFKB2</td>
<td>RELB</td>
<td>the results indicate that dystrophic muscle is characterized by increases in the whole cell expression of ikappab-alpha , p65 , p50 , <span class="gene_color">relb</span> , <span class="gene_color">p100</span> , p52 , ikk , and traf-3 .</td>
<td>0.76</td>
<td>0.553</td>
</tr>
<tr class="odd">
<td>SSSCA1</td>
<td>CDKN1B</td>
<td>conclusion : hl-60 / ht cells have lower <span class="gene_color">p27 (</span> <span class="gene_color">kip1 )</span> expression compared with hl-60 cells .</td>
<td>0.757</td>
<td>0.552</td>
</tr>
<tr class="even">
<td>PTH2R</td>
<td>PTH2</td>
<td>thus , the juxtamembrane receptor domain specifies the signaling and binding selectivity of <span class="gene_color">tip39</span> for the <span class="gene_color">pth2 receptor</span> over the pth1 receptor .</td>
<td>0.749</td>
<td>0.55</td>
</tr>
<tr class="odd">
<td>MMP9</td>
<td>MMP2</td>
<td>all these factors markedly influenced the secretion and/or activation of <span class="gene_color">mmp-2</span> and <span class="gene_color">mmp-9</span> .</td>
<td>0.738</td>
<td>0.547</td>
</tr>
<tr class="even">
<td>CCND1</td>
<td>ABL1</td>
<td>synergy with <span class="gene_color">v-abl</span> depended on a motif in <span class="gene_color">cyclin d1</span> that mediates its binding to the retinoblastoma protein , suggesting that abl oncogenes in part mediate their mitogenic effects via a retinoblastoma protein-dependent pathway .</td>
<td>0.736</td>
<td>0.547</td>
</tr>
</tbody>
</table>
<table>
<caption>Contains the bottom ten Gene-interacts-Gene confidence scores before and after model calbration. Both gene mentions highlighted in <span class="gene_color">blue</span>.{#tbl:gg_bottom_ten_table}</caption>
<colgroup>
<col style="width: 4%" />
<col style="width: 4%" />
<col style="width: 78%" />
<col style="width: 6%" />
<col style="width: 5%" />
</colgroup>
<thead>
<tr class="header">
<th>Gene1 Symbol</th>
<th>Gene2 Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>IFNG</td>
<td>IL6</td>
<td>in the control group , the positive rate for il-4 , <span class="gene_color">il-6</span> , il-10 were 0/10 , 2/10 and 1/10 , respectively , and those for il-2 and <span class="gene_color">ifn-gamma</span> were both 1/10 .</td>
<td>0.012</td>
<td>0.306</td>
</tr>
<tr class="even">
<td>ACHE</td>
<td>BCHE</td>
<td>anticholinesterase activity was determined against acetylcholinesterase ( <span class="gene_color">ache )</span> and butyrylcholinesterase ( <span class="gene_color">bche )</span> , the enzymes vital for alzheimer ’s disease , at 50 , 100 and 200 g ml ( -1 ) .</td>
<td>0.011</td>
<td>0.306</td>
</tr>
<tr class="odd">
<td>CCL2</td>
<td>AGT</td>
<td>we found no significant increase in <span class="gene_color">mcp-1</span> concentrations by <span class="gene_color">ang ii</span> alone ; but it enhanced the tnf-alpha-induced mcp-1 mrna expression in a dose-dependent manner .</td>
<td>0.011</td>
<td>0.306</td>
</tr>
<tr class="even">
<td>CXCL8</td>
<td>IL1B</td>
<td>furthermore , somatostatin completely abrogated the increased secretion of <span class="gene_color">il-8</span> and <span class="gene_color">il-1beta</span> after invasion by salmonella .</td>
<td>0.011</td>
<td>0.303</td>
</tr>
<tr class="odd">
<td>SULT1A2</td>
<td>SULT1A3</td>
<td>to date , the laboratory has cloned seven unique human sulfotransferases ; five aryl sulfotransferases ( hast1 , hast2 , <span class="gene_color">hast3</span> , <span class="gene_color">hast4</span> and hast4v ) , an estrogen sulfotransferase and a dehydroepiandrosterone sulfotransferase .</td>
<td>0.009</td>
<td>0.295</td>
</tr>
<tr class="even">
<td>IFNG</td>
<td>IL10</td>
<td>results : we found weak mrna expression of interleukin-4 ( il-4 ) and il-5 , and strong expression of il-6 , <span class="gene_color">il-10</span> and <span class="gene_color">ifn-gamma</span> before therapy .</td>
<td>0.008</td>
<td>0.292</td>
</tr>
<tr class="odd">
<td>IL2</td>
<td>IFNG</td>
<td>prostaglandin e2 at priming of naive cd4 + t cells inhibits acquisition of ability to produce <span class="gene_color">ifn-gamma</span> and <span class="gene_color">il-2</span> , but not il-4 and il-5 .</td>
<td>0.007</td>
<td>0.289</td>
</tr>
<tr class="even">
<td>IL2</td>
<td>IFNG</td>
<td>the detailed distribution of lymphokine-producing cells showed that <span class="gene_color">il-2</span> and <span class="gene_color">ifn-gamma-producing</span> cells were located mainly in the follicular areas .</td>
<td>0.007</td>
<td>0.287</td>
</tr>
<tr class="odd">
<td>IL2</td>
<td>IFNG</td>
<td>pbl of ms patients produced more pro-inflammatory cytokines , <span class="gene_color">il-2</span> , <span class="gene_color">ifn-gamma</span> and tnf/lymphotoxin , and less anti-inflammatory cytokine , tgf-beta , during wk 2 to 4 in culture than pbl of normal controls .</td>
<td>0.006</td>
<td>0.283</td>
</tr>
<tr class="even">
<td>NFKB1</td>
<td>TNF</td>
<td><span class="gene_color">nf-kappab-dependent</span> reporter gene transcription activated by <span class="gene_color">tnf</span> was also suppressed by calagualine .</td>
<td>0.005</td>
<td>0.276</td>
</tr>
</tbody>
</table>
<!-- default theme -->

<style>
    /* import google fonts */
    @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
    @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

    /* -------------------------------------------------- */
    /* global */
    /* -------------------------------------------------- */

    /* all elements */
    * {
        /* force sans-serif font unless specified otherwise */
        font-family: "Open Sans", "Helvetica", sans-serif;

        /* prevent text inflation on some mobile browsers */
        -webkit-text-size-adjust: none !important;
        -moz-text-size-adjust: none !important;
        -o-text-size-adjust: none !important;
        text-size-adjust: none !important;
    }

    @media only screen {
        /* "page" element */
        body {
            position: relative;
            box-sizing: border-box;
            font-size: 12pt;
            line-height: 1.5;
            max-width: 8.5in;
            margin: 20px auto;
            padding: 40px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* "page" element */
        body {
            padding: 20px;
            margin: 0;
            border-radius: 0;
            border: none;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
            background: none;
        }
    }

    /* -------------------------------------------------- */
    /* headings */
    /* -------------------------------------------------- */

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
        margin: 20px 0;
        padding: 0;
        font-weight: bold;
    }

    /* biggest heading */
    h1 {
        margin: 40px 0;
        text-align: center;
    }

    /* second biggest heading */
    h2 {
        margin-top: 30px;
        padding-bottom: 5px;
        border-bottom: solid 1px #bdbdbd;
    }

    /* -------------------------------------------------- */
    /* manuscript header */
    /* -------------------------------------------------- */

    /* manuscript title */
    header > h1 {
        margin: 0;
    }

    /* manuscript title caption text (ie "automatically generated on") */
    header + p {
        text-align: center;
        margin-top: 10px;
    }

    /* -------------------------------------------------- */
    /* text elements */
    /* -------------------------------------------------- */

    /* links */
    a {
        color: #2196f3;
    }

    /* normal links (not empty, not button link, not syntax highlighting link) */
    a:not(:empty):not(.button):not(.sourceLine) {
        padding-left: 1px;
        padding-right: 1px;
    }

    /* superscripts and subscripts */
    sub,
    sup {
        /* prevent from affecting line height */
        line-height: 0;
    }

    /* unordered and ordered lists*/
    ul,
    ol {
        padding-left: 20px;
    }

    /* class for styling text semibold */
    .semibold {
        font-weight: 600;
    }

    /* class for styling elements horizontally left aligned */
    .left {
        display: block;
        text-align: left;
        margin-left: auto;
        margin-right: 0;
        justify-content: left;
    }

    /* class for styling elements horizontally centered */
    .center {
        display: block;
        text-align: center;
        margin-left: auto;
        margin-right: auto;
        justify-content: center;
    }

    /* class for styling elements horizontally right aligned */
    .right {
        display: block;
        text-align: right;
        margin-left: 0;
        margin-right: auto;
        justify-content: right;
    }

    /* -------------------------------------------------- */
    /* section elements */
    /* -------------------------------------------------- */

    /* horizontal divider line */
    hr {
        border: none;
        height: 1px;
        background: #bdbdbd;
    }

    /* paragraphs, horizontal dividers, figures, tables, code */
    p,
    hr,
    figure,
    table,
    pre {
        /* treat all as "paragraphs", with consistent vertical margins */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* figures */
    /* -------------------------------------------------- */

    /* figure */
    figure {
        max-width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure caption */
    figcaption {
        padding: 0;
        padding-top: 10px;
    }

    /* figure image element */
    figure img {
        max-width: 100%;
        display: block;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure auto-number */
    img + figcaption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* tables */
    /* -------------------------------------------------- */

    /* table */
    table {
        border-collapse: collapse;
        border-spacing: 0;
        width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* table cells */
    th,
    td {
        border: solid 1px #bdbdbd;
        padding: 10px;
        /* squash table if too wide for page by forcing line breaks */
        word-break: break-all;
        word-break: break-word;
    }

    /* header row and even rows */
    th,
    tr:nth-child(2n) {
        background-color: #fafafa;
    }

    /* odd rows */
    tr:nth-child(2n + 1) {
        background-color: #ffffff;
    }

    /* table caption */
    caption {
        text-align: left;
        padding: 0;
        padding-bottom: 10px;
    }

    /* table auto-number */
    table > caption > span:first-of-type,
    div.table_wrapper > table > caption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* code */
    /* -------------------------------------------------- */

    /* multi-line code block */
    pre {
        padding: 10px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
        break-inside: avoid;
        text-align: left;
    }

    /* inline code, ie code within normal text */
    :not(pre) > code {
        padding: 0 4px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
    }

    /* code text */
    /* apply all children, to reach syntax highlighting sub-elements */
    code,
    code * {
        /* force monospace font */
        font-family: "Source Code Pro", "Courier New", monospace;
    }

    /* -------------------------------------------------- */
    /* quotes */
    /* -------------------------------------------------- */

    /* quoted text */
    blockquote {
        margin: 0;
        padding: 0;
        border-left: 4px solid #bdbdbd;
        padding-left: 16px;
        break-inside: avoid;
    }

    /* -------------------------------------------------- */
    /* banners */
    /* -------------------------------------------------- */

    /* info banners */
    .banner {
        box-sizing: border-box;
        display: block;
        position: relative;
        width: 100%;
        margin-top: 20px;
        margin-bottom: 20px;
        padding: 20px;
        text-align: center;
    }

    /* paragraph in banner */
    .banner > p {
        margin: 0;
    }

    /* -------------------------------------------------- */
    /* highlight colors */
    /* -------------------------------------------------- */

    .white {
        background: #ffffff;
    }
    .lightgrey {
        background: #eeeeee;
    }
    .grey {
        background: #757575;
    }
    .darkgrey {
        background: #424242;
    }
    .black {
        background: #000000;
    }
    .lightred {
        background: #ffcdd2;
    }
    .lightyellow {
        background: #ffecb3;
    }
    .lightgreen {
        background: #dcedc8;
    }
    .lightblue {
        background: #e3f2fd;
    }
    .lightpurple {
        background: #f3e5f5;
    }
    .red {
        background: #f44336;
    }
    .orange {
        background: #ff9800;
    }
    .yellow {
        background: #ffeb3b;
    }
    .green {
        background: #4caf50;
    }
    .blue {
        background: #2196f3;
    }
    .purple {
        background: #9c27b0;
    }
    .white,
    .lightgrey,
    .lightred,
    .lightyellow,
    .lightgreen,
    .lightblue,
    .lightpurple,
    .orange,
    .yellow,
    .white a,
    .lightgrey a,
    .lightred a,
    .lightyellow a,
    .lightgreen a,
    .lightblue a,
    .lightpurple a,
    .orange a,
    .yellow a {
        color: #000000;
    }
    .grey,
    .darkgrey,
    .black,
    .red,
    .green,
    .blue,
    .purple,
    .grey a,
    .darkgrey a,
    .black a,
    .red a,
    .green a,
    .blue a,
    .purple a {
        color: #ffffff;
    }

    /* -------------------------------------------------- */
    /* buttons */
    /* -------------------------------------------------- */

    /* class for styling links like buttons */
    .button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        margin: 5px;
        padding: 10px 20px;
        font-size: 0.75em;
        font-weight: 600;
        text-transform: uppercase;
        text-decoration: none;
        letter-spacing: 1px;
        background: none;
        color: #2196f3;
        border: solid 1px #bdbdbd;
        border-radius: 5px;
    }

    /* buttons when hovered */
    .button:hover:not([disabled]),
    .icon_button:hover:not([disabled]) {
        cursor: pointer;
        background: #f5f5f5;
    }

    /* buttons when disabled */
    .button[disabled],
    .icon_button[disabled] {
        opacity: 0.35;
        pointer-events: none;
    }

    /* class for styling buttons containg only single icon */
    .icon_button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        text-decoration: none;
        margin: 0;
        padding: 0;
        background: none;
        border-radius: 5px;
        border: none;
        width: 20px;
        height: 20px;
        min-width: 20px;
        min-height: 20px;
    }

    /* icon button inner svg image */
    .icon_button > svg {
        height: 16px;
    }

    /* -------------------------------------------------- */
    /* icons */
    /* -------------------------------------------------- */

    /* class for styling icons inline with text */
    .inline_icon {
        height: 1em;
        position: relative;
        top: 0.125em;
    }

    /* -------------------------------------------------- */
    /* print control */
    /* -------------------------------------------------- */

    @media print {
        @page {
            /* suggested printing margin */
            margin: 0.5in;
        }

        /* document and "page" elements */
        html, body {
            margin: 0;
            padding: 0;
            width: 100%;
            height: 100%;
        }

        /* "page" element */
        body {
            font-size: 11pt !important;
            line-height: 1.35;
        }

        /* all headings */
        h1,
        h2,
        h3,
        h4,
        h5,
        h6 {
            margin: 15px 0;
        }

        /* heading 1 */
        h1 {
            font-size: 1.75em;
        }

        /* heading 2 */
        h2 {
            font-size: 1.25em;
            margin-top: 0;
        }

        /* heading 3 */
        h3 {
            font-size: 1.10em;
        }

        /* figures and tables */
        figure, table {
            font-size: 0.85em;
        }

        /* table cells */
        th,
        td {
            padding: 5px;
        }

        /* shrink font awesome icons */
        i.fas,
        i.fab,
        i.far,
        i.fal {
            transform: scale(0.85);
        }

        /* decrease banner margins */
        .banner {
            margin-top: 15px;
            margin-bottom: 15px;
            padding: 15px;
        }

        /* class for centering an element vertically on its own page */
        .page_center {
            margin: auto;
            width: 100%;
            height: 100%;
            display: flex;
            align-items: center;
            vertical-align: middle;
            break-before: page;
            break-after: page;
        }

        /* always insert a page break before the element */
        .page_break_before {
            break-before: page;
        }

        /* always insert a page break after the element */
        .page_break_after {
            break-after: page;
        }

        /* avoid page break before the element */
        .page_break_before_avoid {
            break-before: avoid;
        }

        /* avoid page break after the element */
        .page_break_after_avoid {
            break-after: avoid;
        }

        /* avoid page break inside the element */
        .page_break_inside_avoid {
            break-inside: avoid;
        }
    }

    /* -------------------------------------------------- */
    /* override pandoc css quirks */
    /* -------------------------------------------------- */

    .sourceCode {
        /* prevent unsightly overflow in wide code blocks */
        overflow: auto !important;
    }

    div.sourceCode {
        /* prevent background fill on top-most code block  container */
        background: none !important;
    }

    .sourceCode * {
        /* force consistent line spacing */
        line-height: 1.5 !important;
    }

    div.sourceCode {
        /* style code block margins same as <pre> element */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* mathjax */
    /* -------------------------------------------------- */

    /* mathjax containers */
    .math.display > span:not(.MathJax_Preview) {
        /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
        display: flex !important;
        overflow-x: auto !important;
        overflow-y: hidden !important;
        justify-content: center;
        align-items: center;
        margin: 0 !important;
    }

    /* right click menu */
    .MathJax_Menu {
        border-radius: 5px !important;
        border: solid 1px #bdbdbd !important;
        box-shadow: none !important;
    }

    /* equation auto-number */
    span[id^="eq:"] > span.math.display + span {
        font-weight: 600;
    }

    /* equation */
    span[id^="eq:"] > span.math.display > span {
        /* nudge to make room for equation auto-number */
        margin-right: 40px !important;
    }

    /* -------------------------------------------------- */
    /* table scroll plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* table wrapper */
        .table_wrapper {
            /* show scrollbar on tables if necessary to prevent overflow */
            overflow: auto;
            width: 100%;
            margin: 20px 0;
        }

        /* table within table wrapper */
        .table_wrapper table,
        .table_wrapper table * {
            /* don't break table words */
            word-break: normal !important;
        }

        .table_wrapper > table {
            /* move margins from table to table_wrapper to allow margin collapsing */
            margin: 0;
        }
    }

    /* -------------------------------------------------- */
    /* anchors plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anchor button */
        .anchor {
            opacity: 0;
            margin-left: 5px;
        }

        /* anchor buttons within <h2>'s */
        h2 .anchor {
            margin-left: 10px;
        }

        /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
        *:hover > .anchor,
        .anchor:hover,
        .anchor:focus {
            opacity: 1;
        }

        /* anchor button when hovered */
        .anchor:hover {
            cursor: pointer;
        }
    }

    /* always show anchor button on devices with no mouse/hover ability */
    @media (hover: none) {
        .anchor {
            opacity: 1;
        }
    }

    /* always hide anchor button on print */
    @media only print {
        .anchor {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* accordion plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* accordion arrow button */
        .accordion_arrow {
            margin-right: 10px;
        }

        /* arrow icon when <h2> data-collapsed attribute true */
        h2[data-collapsed="true"] > .accordion_arrow > svg {
            transform: rotate(-90deg);
        }

        /* all elements (except <h2>'s) when data-collapsed attribute true */
        *:not(h2)[data-collapsed="true"] {
            display: none;
        }

        /* accordion arrow button when hovered and <h2>'s when hovered */
        .accordion_arrow:hover,
        h2[data-collapsed="true"]:hover,
        h2[data-collapsed="false"]:hover {
            cursor: pointer;
        }
    }

    /* always hide accordion arrow button on print */
    @media only print {
        .accordion_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* tooltips plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* tooltip container */
        #tooltip {
            position: absolute;
            width: 50%;
            min-width: 240px;
            max-width: 75%;
            z-index: 1;
        }

        /* tooltip content */
        #tooltip_content {
            margin-bottom: 5px;
            padding: 20px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
            word-break: break-word;
        }

        /* tooltip copy of paragraphs and figures */
        #tooltip_content > p,
        #tooltip_content > figure {
            margin: 0;
            max-height: 320px;
            overflow-y: auto;
        }

        /* tooltip copy of <img> */
        #tooltip_content > figure > img {
            max-height: 260px;
        }

        /* navigation bar */
        #tooltip_nav_bar {
            margin-top: 10px;
            text-align: center;
        }

        /* navigation bar previous/next buton */
        #tooltip_nav_bar > .icon_button {
            position: relative;
            top: 3px;
        }

        /* navigation bar previous button */
        #tooltip_nav_bar > .icon_button:first-of-type {
            margin-right: 5px;
        }

        /* navigation bar next button */
        #tooltip_nav_bar > .icon_button:last-of-type {
            margin-left: 5px;
        }
    }

    /* always hide tooltip on print */
    @media only print {
        #tooltip {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* jump to first plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* jump button */
        .jump_arrow {
            position: relative;
            top: 0.125em;
            margin-right: 5px;
        }
    }

    /* always hide jump button on print */
    @media only print {
        .jump_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* link highlight plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anything with data-highlighted attribute true */
        [data-highlighted="true"] {
            background: #ffeb3b;
        }

        /* anything with data-selected attribute true */
        [data-selected="true"] {
            background: #ff8a65 !important;
        }

        /* animation definition for glow */
        @keyframes highlight_glow {
            0% {
                background: none;
            }
            10% {
                background: #bbdefb;
            }
            100% {
                background: none;
            }
        }

        /* anything with data-glow attribute true */
        [data-glow="true"] {
            animation: highlight_glow 2s;
        }
    }

    /* -------------------------------------------------- */
    /* table of contents plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* toc panel */
        #toc_panel {
            box-sizing: border-box;
            position: fixed;
            top: 0;
            left: 0;
            background: #ffffff;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            z-index: 2;
        }

        /* toc panel when closed */
        #toc_panel[data-open="false"] {
            min-width: 60px;
            width: 60px;
            height: 60px;
            border-right: solid 1px #bdbdbd;
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc panel when open */
        #toc_panel[data-open="true"] {
            min-width: 260px;
            max-width: 480px;
            /* keep panel edge consistent distance away from "page" edge */
            width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
            bottom: 0;
            border-right: solid 1px #bdbdbd;
        }

        /* toc panel header */
        #toc_header {
            box-sizing: border-box;
            display: flex;
            flex-direction: row;
            align-items: center;
            height: 60px;
            margin: 0;
            padding: 20px;
        }

        /* toc panel header when hovered */
        #toc_header:hover {
            cursor: pointer;
        }

        /* toc panel header when panel open */
        #toc_panel[data-open="true"] > #toc_header {
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc open/close header button */
        #toc_button {
            margin-right: 20px;
        }

        /* hide toc list and header text when closed */
        #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
        #toc_panel[data-open="false"] > #toc_list {
            display: none;
        }

        /* toc list of entries */
        #toc_list {
            box-sizing: border-box;
            width: 100%;
            padding: 20px;
            position: absolute;
            top: calc(60px + 1px);
            bottom: 0;
            overflow: auto;
        }

        /* toc entry, link to section in document */
        .toc_link {
            display: block;
            padding: 5px;
            position: relative;
            font-weight: 600;
            text-decoration: none;
        }

        /* toc entry when hovered or when "viewed" */
        .toc_link:hover,
        .toc_link[data-viewing="true"] {
            background: #f5f5f5;
        }

        /* toc entry, level 1 indentation */
        .toc_link[data-level="1"] {
            margin-left: 0;
        }

        /* toc entry, level 2 indentation */
        .toc_link[data-level="2"] {
            margin-left: 20px;
        }

        /* toc entry, level 3 indentation */
        .toc_link[data-level="3"] {
            margin-left: 40px;
        }

        /* toc entry, level 4 indentation */
        .toc_link[data-level="4"] {
            margin-left: 60px;
        }

        /* toc entry bullets */
        #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
            position: absolute;
            left: -15px;
            top: -1px;
            font-size: 1.5em;
        }

        /* toc entry, level 2 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
            content: "\2022";
        }

        /* toc entry, level 3 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
            content: "\25AB";
        }

        /* toc entry, level 4 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
            content: "-";
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* push <body> ("page") element down to make room for toc icon */
        .toc_body_nudge {
            padding-top: 60px;
        }

        /* toc icon when panel closed and not hovered */
        #toc_panel[data-open="false"]:not(:hover) {
            background: rgba(255, 255, 255, 0.75);
        }
    }

    /* always hide toc panel on print */
    @media only print {
        #toc_panel {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* lightbox plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* regular <img> in document when hovered */
        .lightbox_document_img:hover {
            cursor: pointer;
        }

        .body_no_scroll {
            overflow: hidden !important;
        }

        /* screen overlay */
        #lightbox_overlay {
            display: flex;
            flex-direction: column;
            position: fixed;
            left: 0;
            top: 0;
            right: 0;
            bottom: 0;
            background: rgba(0, 0, 0, 0.75);
            z-index: 3;
        }

        /* middle area containing lightbox image */
        #lightbox_image_container {
            flex-grow: 1;
            display: flex;
            justify-content: center;
            align-items: center;
            overflow: hidden;
            position: relative;
            padding: 20px;
        }

        /* bottom area containing caption */
        #lightbox_bottom_container {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 100px;
            min-height: 100px;
            max-height: 100px;
            background: rgba(0, 0, 0, 0.5);
        }

        /* image number info text box */
        #lightbox_number_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            left: 2px;
            top: 0;
            z-index: 4;
        }

        /* zoom info text box */
        #lightbox_zoom_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            right: 2px;
            top: 0;
            z-index: 4;
        }

        /* copy of image caption */
        #lightbox_caption {
            box-sizing: border-box;
            display: inline-block;
            width: 100%;
            max-height: 100%;
            padding: 10px 0;
            text-align: center;
            overflow-y: auto;
            color: #ffffff;
        }

        /* navigation previous/next button */
        .lightbox_button {
            width: 100px;
            height: 100%;
            min-width: 100px;
            min-height: 100%;
            color: #ffffff;
        }

        /* navigation previous/next button when hovered */
        .lightbox_button:hover {
            background: none !important;
        }

        /* navigation button icon */
        .lightbox_button > svg {
            height: 25px;
        }

        /* figure auto-number */
        #lightbox_caption > span:first-of-type {
            font-weight: bold;
            margin-right: 5px;
        }

        /* lightbox image when hovered */
        #lightbox_img:hover {
            cursor: grab;
        }

        /* lightbox image when grabbed */
        #lightbox_img:active {
            cursor: grabbing;
        }
    }

    /* when on screen < 480px wide */
    @media only screen and (max-width: 480px) {
        /* make navigation buttons skinnier on small screens to make more room for caption text */
        .lightbox_button {
            width: 50px;
            min-width: 50px;
        }
    }

    /* always hide lightbox on print */
    @media only print {
        #lightbox_overlay {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* hypothesis (annotations) plugin */
    /* -------------------------------------------------- */

    /* side panel */
    .annotator-frame {
        width: 280px !important;
        z-index: 0 !important;
    }

    /* match highlight color to rest of theme */
    .annotator-highlights-always-on .annotator-hl {
        background-color: #ffeb3b !important;
    }

    /* match focused color to rest of theme */
    .annotator-hl.annotator-hl-focused {
        background-color: #ff8a65 !important;
    }

    /* match bucket bar color to rest of theme */
    .annotator-bucket-bar {
        background: #f5f5f5 !important;
    }

    /* always hide toolbar and tooltip on print */
    @media only print {
        .annotator-frame {
            display: none !important;
        }

        hypothesis-adder {
            display: none !important;
        }
    }
</style>
<!-- table scroll plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows tables that are too wide to fit within
        // the page to have a scrollbar instead of being squashed to fit.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tableScroll';

        // default plugin options
        const options = {
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // wrap each table in a container
            const tables = document.querySelectorAll('table');
            for (const table of tables)
                wrapElement(table).classList.add('table_wrapper');
            // table_wrapper CSS class in theme file provides scroll
            // functionality
        }

        // wrap element in div and return div
        function wrapElement(element) {
            const parent = element.parentNode;
            const wrapper = document.createElement('div');
            parent.replaceChild(wrapper, element);
            wrapper.appendChild(element);
            return wrapper;
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- anchors plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds an anchor next to each of a certain type
        // of element that provides a human-readable url to that specific
        // item/position in the document (eg "manuscript.html#abstract"). It
        // also makes it such that scrolling out of view of a target removes
        // its identifier from the url.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'anchors';

        // default plugin options
        const options = {
            // which types of elements to add anchors next to, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3, figure, table',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // add anchor to each element of specified types
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements)
                addAnchor(element);

            // attach scroll listener to window
            window.addEventListener('scroll', onScroll);
        }

        // when window is scrolled
        function onScroll() {
            // if url has hash and user has scrolled out of view of hash
            // target, remove hash from url
            const target = getHashTarget();
            if (target) {
                if (
                    target.getBoundingClientRect().top > window.innerHeight ||
                    target.getBoundingClientRect().bottom < 0
                )
                    history.pushState(null, null, ' ');
            }
        }

        // add anchor to element
        function addAnchor(element) {
            let withId; // element with unique id
            let addTo; // element to add anchor button to

            // if figure or table, modify withId and addTo to get expected
            // elements
            if (element.tagName.toLowerCase() === 'figure') {
                withId = element.querySelector('img');
                addTo = element.querySelector('figcaption');
            } else if (element.tagName.toLowerCase() === 'table') {
                withId =
                    element.previousElementSibling ||
                    element.parentNode.previousElementSibling;
                addTo = element.querySelector('caption');
            }

            withId = withId || element;
            addTo = addTo || element;
            const id = withId.id || withId.name || null;

            // do not add anchor if element doesn't have assigned id.
            // id is generated by pandoc and is assumed to be unique and
            // human-readable
            if (!id)
                return;

            // create anchor button
            const anchor = document.createElement('a');
            anchor.innerHTML = document.querySelector('.icon_link').innerHTML;
            anchor.title = 'Link to this part of the document';
            anchor.classList.add('icon_button', 'anchor');
            anchor.dataset.ignore = 'true';
            anchor.href = '#' + id;
            addTo.appendChild(anchor);
        }

        // get element that is target of link or url hash
        function getHashTarget() {
            const hash = window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector(
                '[id="' + id + '"], [name="' + id + '"]'
            );
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (hash.indexOf('#fig:') === 0)
                target = target.parentNode;
            if (hash.indexOf('#tbl:') === 0)
                target = target.nextElementSibling;

            return target;
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- link icon -->

<template class="icon_link">
    <!-- modified from: https://fontawesome.com/icons/link -->
    <svg width="16" height="16" viewBox="0 0 512 512">
        <path
            fill="currentColor"
            d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
        ></path>
    </svg>
</template>
<!-- accordion plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows sections of content under <h2> headings
        // to be collapsible.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'accordion';

        // default plugin options
        const options = {
            // whether to always start expanded ('false'), always start
            // collapsed ('true'), or start collapsed when screen small ('auto')
            startCollapsed: 'auto',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <h2> heading
            const headings = document.querySelectorAll('h2');
            for (const heading of headings) {
                addArrow(heading);

                // start expanded/collapsed based on option
                if (
                    options.startCollapsed === 'true' ||
                    (options.startCollapsed === 'auto' && isSmallScreen())
                )
                    collapseHeading(heading);
                else
                    expandHeading(heading);
            }

            // attach hash change listener to window
            window.addEventListener('hashchange', onHashChange);
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                goToElement(target);
        }

        // add arrow to heading
        function addArrow(heading) {
            // add arrow button
            const arrow = document.createElement('button');
            arrow.innerHTML = document.querySelector(
                '.icon_angle_down'
            ).innerHTML;
            arrow.classList.add('icon_button', 'accordion_arrow');
            heading.insertBefore(arrow, heading.firstChild);

            // attach click listener to heading and button
            heading.addEventListener('click', onHeadingClick);
            arrow.addEventListener('click', onArrowClick);
        }

        // determine if on mobile-like device with small screen
        function isSmallScreen() {
            return Math.min(window.innerWidth, window.innerHeight) < 480;
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get element that is target of hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector(
                '[id="' + id + '"], [name="' + id + '"]'
            );
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (hash.indexOf('#fig:') === 0)
                target = target.parentNode;
            if (hash.indexOf('#tbl:') === 0)
                target = target.nextElementSibling;

            return target;
        }

        // when <h2> heading is clicked
        function onHeadingClick(event) {
            // only collapse if <h2> itself is target of click (eg, user did
            // not click on anchor within <h2>)
            if (event.target === this)
                toggleCollapse(this);
        }

        // when arrow button is clicked
        function onArrowClick() {
            toggleCollapse(this.parentNode);
        }

        // collapse section if expanded, expand if collapsed
        function toggleCollapse(heading) {
            if (heading.dataset.collapsed === 'false')
                collapseHeading(heading);
            else
                expandHeading(heading);
        }

        // elements to exclude from collapse, such as table of contents panel,
        // hypothesis panel, etc
        const exclude = '#toc_panel, div.annotator-frame, #lightbox_overlay';

        // collapse section
        function collapseHeading(heading) {
            heading.setAttribute('data-collapsed', 'true');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'true');
        }

        // expand section
        function expandHeading(heading) {
            heading.setAttribute('data-collapsed', 'false');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'false');
        }

        // get list of elements between this <h2> and next <h2> or <h1>
        // ("children" of the <h2> section)
        function getChildren(heading) {
            return nextUntil(heading, 'h2, h1', exclude);
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get list of elements after a start element up to element matching
        // query
        function nextUntil(element, query, exclude) {
            const elements = [];
            while (element = element.nextElementSibling, element) {
                if (element.matches(query))
                    break;
                if (!element.matches(exclude))
                    elements.push(element);
            }
            return elements;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
    <!-- modified from: https://fontawesome.com/icons/angle-down -->
    <svg width="16" height="16" viewBox="0 0 448 512">
        <path
            fill="currentColor"
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        ></path>
    </svg>
</template>
<!-- tooltips plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when the user hovers or
        // focuses a link to a citation or figure, a tooltip appears with a
        // preview of the reference content, along with arrows to navigate
        // between instances of the same reference in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tooltips';

        // default plugin options
        const options = {
            // whether user must click off to close tooltip instead of just
            // un-hovering
            clickClose: 'false',
            // delay (in ms) between opening and closing tooltip
            delay: '100',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach hover and focus listeners to link
                link.addEventListener('mouseover', onLinkHover);
                link.addEventListener('mouseleave', onLinkUnhover);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('touchend', onLinkTouch);
            }

            // attach mouse, key, and resize listeners to window
            window.addEventListener('mousedown', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('keyup', onKeyUp);
            window.addEventListener('resize', onResize);
        }

        // when link is hovered
        function onLinkHover() {
            // function to open tooltip
            const delayOpenTooltip = function() {
                openTooltip(this);
            }.bind(this);

            // run open function after delay
            this.openTooltipTimer = window.setTimeout(
                delayOpenTooltip,
                options.delay
            );
        }

        // when mouse leaves link
        function onLinkUnhover() {
            // cancel opening tooltip
            window.clearTimeout(this.openTooltipTimer);

            // don't close on unhover if option specifies
            if (options.clickClose === 'true')
                return;

            // function to close tooltip
            const delayCloseTooltip = function() {
                // if tooltip open and if mouse isn't over tooltip, close
                const tooltip = document.getElementById('tooltip');
                if (tooltip && !tooltip.matches(':hover'))
                    closeTooltip();
            };

            // run close function after delay
            this.closeTooltipTimer = window.setTimeout(
                delayCloseTooltip,
                options.delay
            );
        }

        // when link is focused (tabbed to)
        function onLinkFocus(event) {
            openTooltip(this);
        }

        // when link is touched on touch screen
        function onLinkTouch(event) {
            // attempt to force hover state on first tap always, and trigger
            // regular link click (and navigation) on second tap
            if (event.target === document.activeElement)
                event.target.click();
            else {
                document.activeElement.blur();
                event.target.focus();
            }
            if (event.cancelable)
                event.preventDefault();
            event.stopPropagation();
            return false;
        }

        // when mouse is clicked anywhere in window
        function onClick(event) {
            closeTooltip();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'tooltip_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'tooltip_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeTooltip();
                    break;
            }
        }

        // when window is resized or zoomed
        function onResize() {
            closeTooltip();
        }

        // get all links of types we wish to handle
        function getLinks() {
            const queries = [];
            // exclude buttons, anchor links, toc links, etc
            const exclude =
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            queries.push('a[href^="#ref-"]' + exclude); // citation links
            queries.push('a[href^="#fig:"]' + exclude); // figure links
            const query = queries.join(', ');
            return document.querySelectorAll(query);
        }

        // get links with same target, get index of link in set, get total
        // same links
        function getSameLinks(link) {
            const sameLinks = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    sameLinks.push(otherLink);
            }

            return {
                elements: sameLinks,
                index: sameLinks.indexOf(link),
                total: sameLinks.length
            };
        }

        // open tooltip
        function openTooltip(link) {
            // delete tooltip if it exists, start fresh
            closeTooltip();

            // make tooltip element
            const tooltip = makeTooltip(link);

            // if source couldn't be found and tooltip not made, exit
            if (!tooltip)
                return;

            // make navbar elements
            const navBar = makeNavBar(link);
            if (navBar)
                tooltip.firstElementChild.appendChild(navBar);

            // attach tooltip to page
            document.body.appendChild(tooltip);

            // position tooltip
            const position = function() {
                positionTooltip(link);
            };
            position();

            // if tooltip contains images, position again after they've loaded
            const imgs = tooltip.querySelectorAll('img');
            for (const img of imgs)
                img.addEventListener('load', position);
        }

        // close (delete) tooltip
        function closeTooltip() {
            const tooltip = document.getElementById('tooltip');
            if (tooltip)
                tooltip.remove();
        }

        // make tooltip
        function makeTooltip(link) {
            // get target element that link points to
            const source = getSource(link);

            // if source can't be found, exit
            if (!source)
                return;

            // create new tooltip
            const tooltip = document.createElement('div');
            tooltip.id = 'tooltip';
            const tooltipContent = document.createElement('div');
            tooltipContent.id = 'tooltip_content';
            tooltip.appendChild(tooltipContent);

            // make copy of source node and put in tooltip
            const sourceCopy = makeCopy(source);
            tooltipContent.appendChild(sourceCopy);

            // attach mouse event listeners
            tooltip.addEventListener('click', onTooltipClick);
            tooltip.addEventListener('mousedown', onTooltipClick);
            tooltip.addEventListener('touchstart', onTooltipClick);
            tooltip.addEventListener('mouseleave', onTooltipUnhover);

            // (for interaction with lightbox plugin)
            // transfer click on tooltip copied img to original img
            const sourceImg = source.querySelector('img');
            const sourceCopyImg = sourceCopy.querySelector('img');
            if (sourceImg && sourceCopyImg) {
                const clickImg = function() {
                    sourceImg.click();
                    closeTooltip();
                };
                sourceCopyImg.addEventListener('click', clickImg);
            }

            return tooltip;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // when tooltip is clicked
        function onTooltipClick(event) {
            // when user clicks on tooltip, stop click from transferring
            // outside of tooltip (eg, click off to close tooltip, or eg click
            // off to unhighlight same refs)
            event.stopPropagation();
        }

        // when tooltip is unhovered
        function onTooltipUnhover(event) {
            if (options.clickClose === 'true')
                return;

            // make sure new mouse/touch/focus no longer over tooltip or any
            // element within it
            const tooltip = document.getElementById('tooltip');
            if (!tooltip)
                return;
            if (this.contains(event.relatedTarget))
                return;

            closeTooltip();
        }

        // make nav bar to go betwen prev/next instances of same reference
        function makeNavBar(link) {
            // find other links to the same source
            const sameLinks = getSameLinks(link);

            // don't show nav bar when singular reference
            if (sameLinks.total <= 1)
                return;

            // find prev/next links with same target
            const prevLink = getPrevLink(link, sameLinks);
            const nextLink = getNextLink(link, sameLinks);

            // create nav bar
            const navBar = document.createElement('div');
            navBar.id = 'tooltip_nav_bar';
            const text = sameLinks.index + 1 + ' of ' + sameLinks.total;

            // create nav bar prev/next buttons
            const prevButton = document.createElement('button');
            const nextButton = document.createElement('button');
            prevButton.id = 'tooltip_prev_button';
            nextButton.id = 'tooltip_next_button';
            prevButton.title =
                'Jump to the previous occurence of this item in the document [←]';
            nextButton.title =
                'Jump to the next occurence of this item in the document [→]';
            prevButton.classList.add('icon_button');
            nextButton.classList.add('icon_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;
            navBar.appendChild(prevButton);
            navBar.appendChild(document.createTextNode(text));
            navBar.appendChild(nextButton);

            // attach click listeners to buttons
            prevButton.addEventListener('click', function() {
                onPrevNextClick(link, prevLink);
            });
            nextButton.addEventListener('click', function() {
                onPrevNextClick(link, nextLink);
            });

            return navBar;
        }

        // get previous link with same target
        function getPrevLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if < 1
            let index;
            if (sameLinks.index - 1 >= 0)
                index = sameLinks.index - 1;
            else
                index = sameLinks.total - 1;
            return sameLinks.elements[index];
        }

        // get next link with same target
        function getNextLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if > total
            let index;
            if (sameLinks.index + 1 <= sameLinks.total - 1)
                index = sameLinks.index + 1;
            else
                index = 0;
            return sameLinks.elements[index];
        }

        // get element that is target of link or url hash
        function getSource(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector(
                '[id="' + id + '"], [name="' + id + '"]'
            );
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (hash.indexOf('#ref-') === 0)
                target = target.querySelector('p');
            else if (hash.indexOf('#fig:') === 0)
                target = target.parentNode;
            else if (hash.indexOf('#tbl:') === 0)
                return;

            return target;
        }

        // when prev/next arrow button is clicked
        function onPrevNextClick(link, prevNextLink) {
            if (link && prevNextLink)
                goToElement(prevNextLink, window.innerHeight * 0.5);
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // determine position to place tooltip based on link position in
        // viewport and tooltip size
        function positionTooltip(link, left, top) {
            const tooltipElement = document.getElementById('tooltip');
            if (!tooltipElement)
                return;

            // get convenient vars for position/dimensions of
            // link/tooltip/page/view
            link = getRectInPage(link);
            const tooltip = getRectInPage(tooltipElement);
            const view = getRectInPage();

            // horizontal positioning
            if (left)
                // use explicit value
                left = left;
            else if (link.left + tooltip.width < view.right)
                // fit tooltip to right of link
                left = link.left;
            else if (link.right - tooltip.width > view.left)
                // fit tooltip to left of link
                left = link.right - tooltip.width;
            // center tooltip in view
            else
                left = (view.right - view.left) / 2 - tooltip.width / 2;

            // vertical positioning
            if (top)
                // use explicit value
                top = top;
            else if (link.top - tooltip.height > view.top)
                // fit tooltip above link
                top = link.top - tooltip.height;
            else if (link.bottom + tooltip.height < view.bottom)
                // fit tooltip below link
                top = link.bottom;
            else {
                // center tooltip in view
                top = view.top + view.height / 2 - tooltip.height / 2;
                // nudge off of link to left/right if possible
                if (link.right + tooltip.width < view.right)
                    left = link.right;
                else if (link.left - tooltip.width > view.left)
                    left = link.left - tooltip.width;
            }

            tooltipElement.style.left = left + 'px';
            tooltipElement.style.top = top + 'px';
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get position of element relative to page
        function getRectInPage(element) {
            const rect = getRectInView(element);
            const body = getRectInView(document.body);

            const newRect = {};
            newRect.left = rect.left - body.left;
            newRect.top = rect.top - body.top;
            newRect.right = rect.right - body.left;
            newRect.bottom = rect.bottom - body.top;
            newRect.width = rect.width;
            newRect.height = rect.height;

            return newRect;
        }

        // (for interaction with accordion plugin)
        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // (for interaction with accordion plugin)
        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- jump to first plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds a button next to each reference entry,
        // figure, and table that jumps the page to the first occurrence of a
        // link to that item in the manuscript.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'jumpToFirst';

        // default plugin options
        const options = {
            // whether to add buttons next to reference entries
            references: 'true',
            // whether to add buttons next to figures
            figures: 'true',
            // whether to add buttons next to tables
            tables: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            if (options.references !== 'false')
                makeReferenceButtons();
            if (options.figures !== 'false')
                makeFigureButtons();
            if (options.tables !== 'false')
                makeTableButtons();
        }

        // when jump button clicked
        function onButtonClick() {
            const first = getFirstOccurrence(this.dataset.id);
            if (!first)
                return;

            // update url hash so navigating "back" in history will return
            // user to jump button
            window.location.hash = this.dataset.id;
            // scroll to link
            window.setTimeout(function() {
                goToElement(first, window.innerHeight * 0.5);
            }, 0);
        }

        // get first occurence of link to item in document
        function getFirstOccurrence(id) {
            let query = 'a';
            query += '[href="#' + id + '"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelector(query);
        }

        // add button next to each reference entry
        function makeReferenceButtons() {
            const references = document.querySelectorAll('div[id^="ref-"]');
            for (const reference of references) {
                // get reference id and element to add button to
                const id = reference.id;
                const container = reference.firstElementChild;
                const first = getFirstOccurrence(id);

                // if can't find link to reference, ignore
                if (!first)
                    continue;

                // make jump button
                let button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this reference in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.innerHTML = button.outerHTML + container.innerHTML;
                button = container.firstElementChild;
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeFigureButtons() {
            const figures = document.querySelectorAll('img[id^="fig:"]');
            for (const figure of figures) {
                // get figure id and element to add button to
                const id = figure.id;
                const container = figure.nextElementSibling;
                const first = getFirstOccurrence(id);

                // if can't find link to figure, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this figure in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeTableButtons() {
            const tables = document.querySelectorAll('a[name^="tbl:"]');
            for (const table of tables) {
                // get ref id and element to add button to
                const id = table.name;
                const container = table.nextElementSibling.querySelector(
                    'caption'
                );
                const first = getFirstOccurrence(id);

                // if can't find link to table, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this table in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
    <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
    <svg width="16" height="16" viewBox="0 0 320 512">
        <path
            fill="currentColor"
            d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
        ></path>
    </svg>
</template>
<!-- link highlight plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user hovers or
        // focuses a link, other links that have the same target will be
        // highlighted. It also makes it such that when clicking a link, the
        // target of the link (eg reference, figure, table) is briefly
        // highlighted.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'linkHighlight';

        // default plugin options
        const options = {
            // whether to also highlight links that go to external urls
            externalLinks: 'false',
            // whether user must click off to unhighlight instead of just
            // un-hovering
            clickUnhighlight: 'false',
            // whether to also highlight links that are unique
            highlightUnique: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach mouse and focus listeners to link
                link.addEventListener('mouseenter', onLinkFocus);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('mouseleave', onLinkUnhover);
            }

            // attach click and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('hashchange', onHashChange);

            // run hash change on window load in case user has navigated
            // directly to hash
            onHashChange();
        }

        // when link is focused (tabbed to) or hovered
        function onLinkFocus() {
            highlight(this);
        }

        // when link is unhovered
        function onLinkUnhover() {
            if (options.clickUnhighlight !== 'true')
                unhighlightAll();
        }

        // when the mouse is clicked anywhere in window
        function onClick(event) {
            unhighlightAll();
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                glowElement(target);
        }

        // get element that is target of link or url hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector(
                '[id="' + id + '"], [name="' + id + '"]'
            );
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (hash.indexOf('#fig:') === 0)
                target = target.parentNode;
            else if (hash.indexOf('#tbl:') === 0)
                target = target.nextElementSibling.querySelector('caption');

            return target;
        }

        // start glow sequence on an element
        function glowElement(element) {
            const startGlow = function() {
                onGlowEnd();
                element.dataset.glow = 'true';
                element.addEventListener('animationend', onGlowEnd);
            };
            const onGlowEnd = function() {
                element.removeAttribute('data-glow');
                element.removeEventListener('animationend', onGlowEnd);
            };
            startGlow();
        }

        // highlight link and all others with same target
        function highlight(link) {
            // force unhighlight all to start fresh
            unhighlightAll();

            // get links with same target
            if (!link)
                return;
            const sameLinks = getSameLinks(link);

            // if link unique and option is off, exit and don't highlight
            if (sameLinks.length <= 1 && options.highlightUnique !== 'true')
                return;

            // highlight all same links, and "select" (special highlight) this
            // one
            for (const sameLink of sameLinks) {
                if (sameLink === link)
                    sameLink.setAttribute('data-selected', 'true');
                else
                    sameLink.setAttribute('data-highlighted', 'true');
            }
        }

        // unhighlight all links
        function unhighlightAll() {
            const links = getLinks();
            for (const link of links) {
                link.setAttribute('data-selected', 'false');
                link.setAttribute('data-highlighted', 'false');
            }
        }

        // get links with same target
        function getSameLinks(link) {
            const results = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    results.push(otherLink);
            }
            return results;
        }

        // get all links of types we wish to handle
        function getLinks() {
            let query = 'a';
            if (options.externalLinks !== 'true')
                query += '[href^="#"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelectorAll(query);
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- table of contents plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin provides a "table of contents" (toc) panel on
        // the side of the document that allows the user to conveniently
        // navigate between sections of the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tableOfContents';

        // default plugin options
        const options = {
            // which types of elements to add links for, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3',
            // whether default behavior is to be closed ('false'), open
            // ('true'), or only open when screen wide enough to fit panel
            // ('auto'). note: still always starts closed when page loads.
            open: 'auto',
            // if list item is more than this many characters, text will be
            // truncated
            charLimit: '50',
            // whether or not to show bullets next to each toc item
            bullets: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // make toc panel and populate with entries (links to document
            // sections)
            const panel = makePanel();
            if (!panel)
                return;
            makeEntries(panel);
            document.body.insertBefore(panel, document.body.firstChild);

            closePanel();

            // attach click, scroll, and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('scroll', onScroll);
            window.addEventListener('hashchange', onScroll);
            window.addEventListener('keyup', onKeyUp);
            onScroll();

            // add class to push document body down out of way of toc button
            document.body.classList.add('toc_body_nudge');
        }

        // determine if screen wide enough to fit toc panel
        function isSmallScreen() {
            // in default theme:
            // 816px = 8.5in = width of "page" (<body>) element
            // 260px = min width of toc panel (*2 for both sides of <body>)
            return window.innerWidth < 816 + 260 * 2;
        }

        // open/close panel based on option and screen size
        function openOrClosePanel() {
            if (
                options.open === 'true' ||
                (options.open === 'auto' && !isSmallScreen())
            )
                openPanel();
            else
                closePanel();
        }

        // when mouse is clicked anywhere in window
        function onClick() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                openOrClosePanel();
        }

        // when window is scrolled or hash changed
        function onScroll() {
            highlightViewed();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            // close on esc
            if (event.key === 'Escape')
                closePanel();
        }

        // find entry of currently viewed document section in toc and highlight
        function highlightViewed() {
            const firstId = getFirstInView(options.typesQuery);

            // get toc entries (links), unhighlight all, then highlight viewed
            const list = document.getElementById('toc_list');
            if (!firstId || !list)
                return;
            const links = list.querySelectorAll('a');
            for (const link of links)
                link.dataset.viewing = 'false';
            const link = list.querySelector('a[href="#' + firstId + '"]');
            if (!link)
                return;
            link.dataset.viewing = 'true';
        }

        // get first or previous toc listed element in top half of view
        function getFirstInView(query) {
            // get all elements matching query and with id
            const elements = document.querySelectorAll(query);
            const elementsWithIds = [];
            for (const element of elements) {
                if (element.id)
                    elementsWithIds.push(element);
            }


            // get first or previous element in top half of view
            for (let i = 0; i < elementsWithIds.length; i++) {
                const element = elementsWithIds[i];
                const prevElement = elementsWithIds[Math.max(0, i - 1)];
                if (element.getBoundingClientRect().top >= 0) {
                    if (
                        element.getBoundingClientRect().top <
                        window.innerHeight / 2
                    )
                        return element.id;
                    else
                        return prevElement.id;
                }
            }
        }

        // make panel
        function makePanel() {
            // create panel
            const panel = document.createElement('div');
            panel.id = 'toc_panel';
            if (options.bullets === 'true')
                panel.dataset.bullets = 'true';

            // create header
            const header = document.createElement('div');
            header.id = 'toc_header';

            // create toc button
            const button = document.createElement('button');
            button.id = 'toc_button';
            button.innerHTML = document.querySelector(
                '.icon_th_list'
            ).innerHTML;
            button.classList.add('icon_button');

            // create header text
            const text = document.createElement('h3');
            text.innerHTML = 'Table of Contents';

            // create container for toc list
            const list = document.createElement('div');
            list.id = 'toc_list';

            // attach click listeners
            panel.addEventListener('click', onPanelClick);
            header.addEventListener('click', onHeaderClick);
            button.addEventListener('click', onButtonClick);

            // attach elements
            header.appendChild(button);
            header.appendChild(text);
            panel.appendChild(header);
            panel.appendChild(list);

            return panel;
        }

        // create toc entries (links) to each element of the specified types
        function makeEntries(panel) {
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements) {
                // do not add link if element doesn't have assigned id
                if (!element.id)
                    continue;

                // create link/list item
                const link = document.createElement('a');
                link.classList.add('toc_link');
                switch (element.tagName.toLowerCase()) {
                    case 'h1':
                        link.dataset.level = '1';
                        break;
                    case 'h2':
                        link.dataset.level = '2';
                        break;
                    case 'h3':
                        link.dataset.level = '3';
                        break;
                    case 'h4':
                        link.dataset.level = '4';
                        break;
                }
                link.title = element.innerText;
                let text = element.innerText;
                if (text.length > options.charLimit)
                    text = text.slice(0, options.charLimit) + '...';
                link.innerHTML = text;
                link.href = '#' + element.id;
                link.addEventListener('click', onLinkClick);

                // attach link
                panel.querySelector('#toc_list').appendChild(link);
            }
        }

        // when panel is clicked
        function onPanelClick(event) {
            // stop click from propagating to window/document and closing panel
            event.stopPropagation();
        }

        // when header itself is clicked
        function onHeaderClick(event) {
            togglePanel();
        }

        // when button is clicked
        function onButtonClick(event) {
            togglePanel();
            // stop header underneath button from also being clicked
            event.stopPropagation();
        }

        // when link is clicked
        function onLinkClick() {
            openOrClosePanel();
        }

        // open panel if closed, close if opened
        function togglePanel() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                closePanel();
            else
                openPanel();
        }

        // open panel
        function openPanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'true';
        }

        // close panel
        function closePanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'false';
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- th list icon -->

<template class="icon_th_list">
    <!-- modified from: https://fontawesome.com/icons/th-list -->
    <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
        <path
            fill="currentColor"
            d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- lightbox plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user clicks on an
        // image, the image fills the screen and the user can pan/drag/zoom
        // the image and navigate between other images in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'lightbox';

        // default plugin options
        const options = {
            // list of possible zoom/scale factors
            zoomSteps:
                '0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1,' +
                '1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8',
            // whether to fit image to view ('fit'), display at 100% and shrink
            // if necessary ('shrink'), or always display at 100% ('100')
            defaultZoom: 'fit',
            // whether to zoom in/out toward center of view ('true') or mouse
            // ('false')
            centerZoom: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <img> element
            const imgs = document.querySelectorAll('figure > img');
            let count = 1;
            for (const img of imgs) {
                img.classList.add('lightbox_document_img');
                img.dataset.number = count;
                img.dataset.total = imgs.length;
                img.addEventListener('click', openLightbox);
                count++;
            }

            // attach mouse and key listeners to window
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('keyup', onKeyUp);
        }

        // when mouse is moved anywhere in window
        function onWindowMouseMove(event) {
            window.mouseX = event.clientX;
            window.mouseY = event.clientY;
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'lightbox_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'lightbox_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeLightbox();
                    break;
            }
        }

        // open lightbox
        function openLightbox() {
            const lightbox = makeLightbox(this);
            if (!lightbox)
                return;

            blurBody(lightbox);
            document.body.appendChild(lightbox);
        }

        // make lightbox
        function makeLightbox(img) {
            // delete lightbox if it exists, start fresh
            closeLightbox();

            // create screen overlay containing lightbox
            const overlay = document.createElement('div');
            overlay.id = 'lightbox_overlay';

            // create image info boxes
            const numberInfo = document.createElement('div');
            const zoomInfo = document.createElement('div');
            numberInfo.id = 'lightbox_number_info';
            zoomInfo.id = 'lightbox_zoom_info';

            // create container for image
            const imageContainer = document.createElement('div');
            imageContainer.id = 'lightbox_image_container';
            const lightboxImg = makeLightboxImg(
                img,
                imageContainer,
                numberInfo,
                zoomInfo
            );
            imageContainer.appendChild(lightboxImg);

            // create bottom container for caption and navigation buttons
            const bottomContainer = document.createElement('div');
            bottomContainer.id = 'lightbox_bottom_container';
            const caption = makeCaption(img);
            const prevButton = makePrevButton(img);
            const nextButton = makeNextButton(img);
            bottomContainer.appendChild(prevButton);
            bottomContainer.appendChild(caption);
            bottomContainer.appendChild(nextButton);

            // attach top middle and bottom to overlay
            overlay.appendChild(numberInfo);
            overlay.appendChild(zoomInfo);
            overlay.appendChild(imageContainer);
            overlay.appendChild(bottomContainer);

            return overlay;
        }

        // make <img> object that is intuitively draggable and zoomable
        function makeLightboxImg(
            sourceImg,
            container,
            numberInfoBox,
            zoomInfoBox
        ) {
            // create copy of source <img>
            const img = sourceImg.cloneNode(true);
            img.classList.remove('lightbox_document_img');
            img.removeAttribute('id');
            img.removeAttribute('width');
            img.removeAttribute('height');
            img.style.position = 'unset';
            img.style.margin = '0';
            img.style.padding = '0';
            img.style.width = '';
            img.style.height = '';
            img.style.minWidth = '';
            img.style.minHeight = '';
            img.style.maxWidth = '';
            img.style.maxHeight = '';
            img.id = 'lightbox_img';

            // build sorted list of unique zoomSteps, always including a 100%
            let zoomSteps = [];
            const optionsZooms = options.zoomSteps.split(/[^0-9.]/);
            for (const optionZoom of optionsZooms) {
                const newZoom = parseFloat(optionZoom);
                if (newZoom && !zoomSteps.includes(newZoom))
                    zoomSteps.push(newZoom);
            }
            if (!zoomSteps.includes(1))
                zoomSteps.push(1);
            zoomSteps = zoomSteps.sort(function sortNumber(a, b) {
                return a - b;
            });

            // <img> object property variables
            let zoom = 1;
            let translateX = 0;
            let translateY = 0;
            let clickMouseX = undefined;
            let clickMouseY = undefined;
            let clickTranslateX = undefined;
            let clickTranslateY = undefined;

            updateNumberInfo();

            // update image numbers displayed in info box
            function updateNumberInfo() {
                numberInfoBox.innerHTML =
                    sourceImg.dataset.number + ' of ' + sourceImg.dataset.total;
            }

            // update zoom displayed in info box
            function updateZoomInfo() {
                let zoomInfo = zoom * 100;
                if (!Number.isInteger(zoomInfo))
                    zoomInfo = zoomInfo.toFixed(2);
                zoomInfoBox.innerHTML = zoomInfo + '%';
            }

            // move to closest zoom step above current zoom
            const zoomIn = function() {
                for (const zoomStep of zoomSteps) {
                    if (zoomStep > zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                updateTransform();
            };

            // move to closest zoom step above current zoom
            const zoomOut = function() {
                zoomSteps.reverse();
                for (const zoomStep of zoomSteps) {
                    if (zoomStep < zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                zoomSteps.reverse();

                updateTransform();
            };

            // update display of <img> based on scale/translate properties
            const updateTransform = function() {
                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                // get new width/height after scale
                const rect = img.getBoundingClientRect();
                // limit translate
                translateX = Math.max(translateX, -rect.width / 2);
                translateX = Math.min(translateX, rect.width / 2);
                translateY = Math.max(translateY, -rect.height / 2);
                translateY = Math.min(translateY, rect.height / 2);

                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                updateZoomInfo();
            };

            // fit <img> to container
            const fit = function() {
                // no x/y offset, 100% zoom by default
                translateX = 0;
                translateY = 0;
                zoom = 1;

                // widths of <img> and container
                const imgWidth = img.naturalWidth;
                const imgHeight = img.naturalHeight;
                const containerWidth = parseFloat(
                    window.getComputedStyle(container).width
                );
                const containerHeight = parseFloat(
                    window.getComputedStyle(container).height
                );

                // how much zooming is needed to fit <img> to container
                const xRatio = imgWidth / containerWidth;
                const yRatio = imgHeight / containerHeight;
                const maxRatio = Math.max(xRatio, yRatio);
                const newZoom = 1 / maxRatio;

                // fit <img> to container according to option
                if (options.defaultZoom === 'shrink') {
                    if (maxRatio > 1)
                        zoom = newZoom;
                } else if (options.defaultZoom === 'fit')
                    zoom = newZoom;

                updateTransform();
            };

            // when mouse wheel is rolled anywhere in container
            const onContainerWheel = function(event) {
                if (!event)
                    return;

                // let ctrl + mouse wheel to zoom behave as normal
                if (event.ctrlKey)
                    return;

                // prevent normal scroll behavior
                event.preventDefault();
                event.stopPropagation();

                // point around which to scale img
                const viewRect = container.getBoundingClientRect();
                const viewX = (viewRect.left + viewRect.right) / 2;
                const viewY = (viewRect.top + viewRect.bottom) / 2;
                const originX = options.centerZoom === 'true' ? viewX : mouseX;
                const originY = options.centerZoom === 'true' ? viewY : mouseY;

                // get point on image under origin
                const oldRect = img.getBoundingClientRect();
                const oldPercentX = (originX - oldRect.left) / oldRect.width;
                const oldPercentY = (originY - oldRect.top) / oldRect.height;

                // increment/decrement zoom
                if (event.deltaY < 0)
                    zoomIn();
                if (event.deltaY > 0)
                    zoomOut();

                // get offset between previous image point and origin
                const newRect = img.getBoundingClientRect();
                const offsetX =
                    originX - (newRect.left + newRect.width * oldPercentX);
                const offsetY =
                    originY - (newRect.top + newRect.height * oldPercentY);

                // translate image to keep image point under origin
                translateX += offsetX;
                translateY += offsetY;

                // perform translate
                updateTransform();
            };

            // when container is clicked
            function onContainerClick(event) {
                // if container itself is target of click, and not other
                // element above it
                if (event.target === this)
                    closeLightbox();
            }

            // when mouse button is pressed on image
            const onImageMouseDown = function(event) {
                // store original mouse position relative to image
                clickMouseX = window.mouseX;
                clickMouseY = window.mouseY;
                clickTranslateX = translateX;
                clickTranslateY = translateY;
                event.stopPropagation();
                event.preventDefault();
            };

            // when mouse button is released anywhere in window
            const onWindowMouseUp = function(event) {
                // reset original mouse position
                clickMouseX = undefined;
                clickMouseY = undefined;
                clickTranslateX = undefined;
                clickTranslateY = undefined;

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mouseup', onWindowMouseUp);
            };

            // when mouse is moved anywhere in window
            const onWindowMouseMove = function(event) {
                if (
                    clickMouseX === undefined ||
                    clickMouseY === undefined ||
                    clickTranslateX === undefined ||
                    clickTranslateY === undefined
                )
                    return;

                // offset image based on original and current mouse position
                translateX = clickTranslateX + window.mouseX - clickMouseX;
                translateY = clickTranslateY + window.mouseY - clickMouseY;
                updateTransform();
                event.preventDefault();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mousemove', onWindowMouseMove);
            };

            // when window is resized
            const onWindowResize = function(event) {
                fit();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('resize', onWindowResize);
            };

            // attach the necessary event listeners
            img.addEventListener('dblclick', fit);
            img.addEventListener('mousedown', onImageMouseDown);
            container.addEventListener('wheel', onContainerWheel);
            container.addEventListener('mousedown', onContainerClick);
            container.addEventListener('touchstart', onContainerClick);
            window.addEventListener('mouseup', onWindowMouseUp);
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('resize', onWindowResize);

            // run fit() after lightbox atttached to document and <img> Loaded
            // so needed container and img dimensions available
            img.addEventListener('load', fit);

            return img;
        }

        // make caption
        function makeCaption(img) {
            const caption = document.createElement('div');
            caption.id = 'lightbox_caption';
            const captionSource = img.nextElementSibling;
            if (captionSource.tagName.toLowerCase() === 'figcaption') {
                const captionCopy = makeCopy(captionSource);
                caption.innerHTML = captionCopy.innerHTML;
            }

            caption.addEventListener('touchstart', function(event) {
                event.stopPropagation();
            });

            return caption;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // make button to jump to previous image in document
        function makePrevButton(img) {
            const prevButton = document.createElement('button');
            prevButton.id = 'lightbox_prev_button';
            prevButton.title = 'Jump to the previous image in the document [←]';
            prevButton.classList.add('icon_button', 'lightbox_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;

            // attach click listeners to button
            prevButton.addEventListener('click', function() {
                getPrevImg(img).click();
            });

            return prevButton;
        }

        // make button to jump to next image in document
        function makeNextButton(img) {
            const nextButton = document.createElement('button');
            nextButton.id = 'lightbox_next_button';
            nextButton.title = 'Jump to the next image in the document [→]';
            nextButton.classList.add('icon_button', 'lightbox_button');
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;

            // attach click listeners to button
            nextButton.addEventListener('click', function() {
                getNextImg(img).click();
            });

            return nextButton;
        }

        // get previous image in document
        function getPrevImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if < 1
            if (index - 1 >= 0)
                index--;
            else
                index = imgs.length - 1;
            return imgs[index];
        }

        // get next image in document
        function getNextImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if > total
            if (index + 1 <= imgs.length - 1)
                index++;
            else
                index = 0;
            return imgs[index];
        }

        // close lightbox
        function closeLightbox() {
            focusBody();

            const lightbox = document.getElementById('lightbox_overlay');
            if (lightbox)
                lightbox.remove();
        }

        // make all elements behind lightbox non-focusable
        function blurBody(overlay) {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.tabIndex = -1;
            document.body.classList.add('body_no_scroll');
        }

        // make all elements focusable again
        function focusBody() {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.removeAttribute('tabIndex');
            document.body.classList.remove('body_no_scroll');
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- math plugin configuration -->

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        "CommonHTML": { linebreaks: { automatic: true } },
        "HTML-CSS": { linebreaks: { automatic: true } },
        "SVG": { linebreaks: { automatic: true } }
    });
</script>

<!-- math plugin -->

<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML'>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'MathJax' allows the proper rendering of
    // math/equations written in LaTeX.

    // https://www.mathjax.org/
</script>
<!-- annotations plugin configuration -->

<script>
    window.hypothesisConfig = function() {
        return {
            branding: {
                accentColor: '#2196f3',
                appBackgroundColor: '#f8f8f8',
                ctaBackgroundColor: '#f8f8f8',
                ctaTextColor: '#000000',
                selectionFontFamily: 'Open Sans, Helvetica, sans serif',
                annotationFontFamily: 'Open Sans, Helvetica, sans serif'
            }
        };
    };
</script>

<!-- annotations plugin -->

<script src='https://hypothes.is/embed.js'>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'Hypothesis' allows public annotation of the
    // manuscript.

    // https://web.hypothes.is/
</script>
<!-- analytics plugin -->

<!-- copy and paste code from Google Analytics or similar service here -->
</body>
</html>
