<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="David N. Nicholson" />
  <meta name="author" content="Daniel S. Himmelstein" />
  <meta name="author" content="Casey S. Greene" />
  <meta name="dcterms.date" content="2020-01-23" />
  <meta name="keywords" content="machine learning, weak supervision, natural language processing, heterogenous netowrks, text mining" />
  <title>Reusing label functions to extract multiple types of relationships from biomedical abstracts at scale</title>
  <style>
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Reusing label functions to extract multiple types of relationships from biomedical abstracts at scale" />
  <meta name="citation_title" content="Reusing label functions to extract multiple types of relationships from biomedical abstracts at scale" />
  <meta property="og:title" content="Reusing label functions to extract multiple types of relationships from biomedical abstracts at scale" />
  <meta property="twitter:title" content="Reusing label functions to extract multiple types of relationships from biomedical abstracts at scale" />
  <meta name="dc.date" content="2020-01-23" />
  <meta name="citation_publication_date" content="2020-01-23" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="David N. Nicholson" />
  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania" />
  <meta name="citation_author_orcid" content="0000-0003-0002-5761" />
  <meta name="twitter:creator" content="@None" />
  <meta name="citation_author" content="Daniel S. Himmelstein" />
  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania" />
  <meta name="citation_author_orcid" content="0000-0002-3012-7446" />
  <meta name="twitter:creator" content="@dhimmel" />
  <meta name="citation_author" content="Casey S. Greene" />
  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania" />
  <meta name="citation_author_orcid" content="0000-0001-8713-9213" />
  <meta name="twitter:creator" content="@GreeneScientist" />
  <link rel="canonical" href="https://greenelab.github.io/text_mined_hetnet_manuscript/" />
  <meta property="og:url" content="https://greenelab.github.io/text_mined_hetnet_manuscript/" />
  <meta property="twitter:url" content="https://greenelab.github.io/text_mined_hetnet_manuscript/" />
  <meta name="citation_fulltext_html_url" content="https://greenelab.github.io/text_mined_hetnet_manuscript/" />
  <meta name="citation_pdf_url" content="https://greenelab.github.io/text_mined_hetnet_manuscript/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://greenelab.github.io/text_mined_hetnet_manuscript/manuscript.pdf" /> <link rel="alternate" type="text/html" href="https://greenelab.github.io/text_mined_hetnet_manuscript/v/4ff5960460ed6c74368aceca1a178413b49d9ad1/" />
  <meta name="manubot_html_url_versioned" content="https://greenelab.github.io/text_mined_hetnet_manuscript/v/4ff5960460ed6c74368aceca1a178413b49d9ad1/" />
  <meta name="manubot_pdf_url_versioned" content="https://greenelab.github.io/text_mined_hetnet_manuscript/v/4ff5960460ed6c74368aceca1a178413b49d9ad1/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <meta property="og:image" content="https://github.com/greenelab/text_mined_hetnet_manuscript/raw/4ff5960460ed6c74368aceca1a178413b49d9ad1/thumbnail.png" />
  <meta property="twitter:image" content="https://github.com/greenelab/text_mined_hetnet_manuscript/raw/4ff5960460ed6c74368aceca1a178413b49d9ad1/thumbnail.png" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" /> <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Reusing label functions to extract multiple types of relationships from biomedical abstracts at scale</h1>
</header>
<p><em>A DOI-citable version of this manuscript is available at <a href="https://doi.org/10.1101/730085" class="uri">https://doi.org/10.1101/730085</a></em>.</p>
<p><small><em> This manuscript (<a href="https://greenelab.github.io/text_mined_hetnet_manuscript/v/4ff5960460ed6c74368aceca1a178413b49d9ad1/">permalink</a>) was automatically generated from <a href="https://github.com/greenelab/text_mined_hetnet_manuscript/tree/4ff5960460ed6c74368aceca1a178413b49d9ad1">greenelab/text_mined_hetnet_manuscript@4ff5960</a> on January 23, 2020. </em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>David N. Nicholson</strong><br> <img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" /> <a href="https://orcid.org/0000-0003-0002-5761">0000-0003-0002-5761</a> · <img src="images/github.svg" class="inline_icon" alt="GitHub icon" /> <a href="https://github.com/danich1">danich1</a><br> <small> Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania · Funded by GBMF4552 </small></p></li>
<li><p><strong>Daniel S. Himmelstein</strong><br> <img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" /> <a href="https://orcid.org/0000-0002-3012-7446">0000-0002-3012-7446</a> · <img src="images/github.svg" class="inline_icon" alt="GitHub icon" /> <a href="https://github.com/dhimmel">dhimmel</a> · <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" /> <a href="https://twitter.com/dhimmel">dhimmel</a><br> <small> Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania · Funded by GBMF4552 </small></p></li>
<li><p><strong>Casey S. Greene</strong><br> <img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" /> <a href="https://orcid.org/0000-0001-8713-9213">0000-0001-8713-9213</a> · <img src="images/github.svg" class="inline_icon" alt="GitHub icon" /> <a href="https://github.com/cgreene">cgreene</a> · <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" /> <a href="https://twitter.com/GreeneScientist">GreeneScientist</a><br> <small> Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania · Funded by GBMF4552 and R01 HG010067 </small></p></li>
</ul>
<h2 id="abstract" class="page_break_before">Abstract</h2>
<p>Knowledge bases support multiple research efforts such as providing contextual information for biomedical entities, constructing networks, and supporting the interpretation of high-throughput analyses. Some knowledge bases are automatically constructed, but most are populated via some form of manual curation. Manual curation is time consuming and difficult to scale in the context of an increasing publication rate. A recently described “data programming” paradigm seeks to circumvent this arduous process by combining distant supervision with simple rules and heuristics written as labeling functions that can be automatically applied to inputs. Unfortunately writing useful label functions requires substantial error analysis and is a nontrivial task: in early efforts to use data programming we found that producing each label function could take a few days. Producing a biomedical knowledge base with multiple node and edge types could take hundreds or possibly thousands of label functions. In this paper we sought to evaluate the extent to which label functions could be re-used across edge types. We used a subset of Hetionet v1 that centered on disease, compound, and gene nodes to evaluate this approach. We compared a baseline distant supervision model with the same distant supervision resources added to edge-type-specific label functions, edge-type-mismatch label functions, and all label functions. We confirmed that adding additional edge-type-specific label functions improves performance. We also found that adding one or a few edge-type-mismatch label functions nearly always improved performance. Adding a large number of edge-type-mismatch label functions produce variable performance that depends on the edge type being predicted and the label function’s edge type source. Lastly, we show that this approach, even on this subgraph of Hetionet, could add new edges to Hetionet v1 with high confidence. We expect that practical use of this strategy would include additional filtering and scoring methods which would further enhance precision.</p>
<h2 id="introduction">Introduction</h2>
<p>Knowledge bases are important resources that hold complex structured and unstructed information. These resources have been used in important tasks such as network analysis for drug repurposing discovery <span class="citation" data-cites="u8pIAt5j bPvC638e O21tn8vf">[<a href="#ref-u8pIAt5j" role="doc-biblioref">1</a>,<a href="#ref-bPvC638e" role="doc-biblioref">2</a>,<a href="#ref-O21tn8vf" role="doc-biblioref">3</a>]</span> or as a source of training labels for text mining systems <span class="citation" data-cites="EHeTvZht CVHSURuI HS4ARwmZ">[<a href="#ref-EHeTvZht" role="doc-biblioref">4</a>,<a href="#ref-CVHSURuI" role="doc-biblioref">5</a>,<a href="#ref-HS4ARwmZ" role="doc-biblioref">6</a>]</span>. Populating knowledge bases often requires highly-trained scientists to read biomedical literature and summarize the results <span class="citation" data-cites="N1Ai0gaI">[<a href="#ref-N1Ai0gaI" role="doc-biblioref">7</a>]</span>. This manual curation process requires a significant amount of effort and time: in 2007 researchers estimated that filling in the missing annotations would require approximately 8.4 years <span class="citation" data-cites="UdzvLgBM">[<a href="#ref-UdzvLgBM" role="doc-biblioref">8</a>]</span>. The rate of publications has continued to increase exponentially <span class="citation" data-cites="1DBISRlwN">[<a href="#ref-1DBISRlwN" role="doc-biblioref">9</a>]</span>. This has been recognized as a considerable challenge, which can lead to gaps in knowledge bases <span class="citation" data-cites="UdzvLgBM">[<a href="#ref-UdzvLgBM" role="doc-biblioref">8</a>]</span>.<br />
Relationship extraction has been studied as a solution towards handling this problem <span class="citation" data-cites="N1Ai0gaI">[<a href="#ref-N1Ai0gaI" role="doc-biblioref">7</a>]</span>. This process consists of creating a machine learning system to automatically scan and extract relationships from textual sources. Machine learning methods often leverage a large corpus of well-labeled training data, which still requires manual curation. Distant supervision is one technique to sidestep the requirement of well-annotated sentences: with distant supervision one makes the assumption that all sentences containing an entity pair found in a selected database provide evidence for a relationship <span class="citation" data-cites="EHeTvZht">[<a href="#ref-EHeTvZht" role="doc-biblioref">4</a>]</span>. Distant supervision provides many labeled examples; however it is accompanied by a decrease in the quality of the labels.<br />
Ratner et al. <span class="citation" data-cites="5Il3kN32">[<a href="#ref-5Il3kN32" role="doc-biblioref">10</a>]</span> recently introduced “data programming” as a solution. Data programming combines distant supervision with the automated labeling of text using hand-written label functions. The distant supervision sources and label functions are integrated using a noise aware generative model that is used to produce training labels. Combining distant supervision with label functions can dramatically reduce the time required to acquire sufficient training data. However, constructing a knowledge base of heterogeneous relationships through this framework still requires tens of hand-written label functions for each relationship type. Writing useful label functions requires significant error analysis, which can be a time-consuming process.</p>
<p>In this paper, we aim to address the question: to what extent can label functions be re-used across different relationship types? We hypothesized that sentences describing one relationship type may share information in the form of keywords or sentence structure with sentences that indicate other relationship types. We designed a series of experiments to determine the extent to which label function re-use enhanced performance over distant supervision alone. We examined relationships that indicated similar types of physical interactions (i.e., gene-binds-gene and compound-binds-gene) as well as different types (i.e., disease-associates-gene and compound-treats-disease). The re-use of label functions could dramatically reduce the number required to generate and update a heterogeneous knowledge graph.</p>
<h3 id="related-work">Related Work</h3>
<p>Relationship extraction is the process of detecting and classifying semantic relationships from a collection of text. This process can be broken down into three different categories: (1) the use of natural language processing techniques such as manually crafted rules and the identification of key text patterns for relationship extraction, (2) the use of unsupervised methods via co-occurrence scores or clustering, and (3) supervised or semi-supervised machine learning using annotated datasets for the classification of documents or sentences. In this section, we discuss selected efforts for each type of edge that we include in this project.</p>
<h4 id="disease-gene-associations">Disease-Gene Associations</h4>
<p>Efforts to extract Disease-associates-Gene (DaG) relationships have often used manually crafted rules or unsupervised methods. One study used hand crafted rules based on a sentence’s grammatical structure, represented as dependency trees, to extract DaG relationships <span class="citation" data-cites="NLxmpSdj">[<a href="#ref-NLxmpSdj" role="doc-biblioref">11</a>]</span>. Some of these rules inspired certain DaG text pattern label functions in our work. Another study used co-occurrence frequencies within abstracts and sentences to score the likelihood of association between disease and gene pairs <span class="citation" data-cites="5gG8hwv7">[<a href="#ref-5gG8hwv7" role="doc-biblioref">12</a>]</span>. The results of this study were incorporated into Hetionet v1 <span class="citation" data-cites="O21tn8vf">[<a href="#ref-O21tn8vf" role="doc-biblioref">3</a>]</span>, so this served as one of our distant supervision label functions. Another approach built off of the above work by incorporating a supervised classifier, trained via distant supervision, into a scoring scheme <span class="citation" data-cites="IGXdryzB">[<a href="#ref-IGXdryzB" role="doc-biblioref">13</a>]</span>. Each sentence containing a disease and gene mention is scored using a logistic regression model and combined using the same co-occurrence approach used in Pletscher-Frankild et al. <span class="citation" data-cites="5gG8hwv7">[<a href="#ref-5gG8hwv7" role="doc-biblioref">12</a>]</span>. We compared our results to this approach to measure how well our overall method performs relative to other methods. Besides the mentioned three studies, researchers have used co-occurrences for extraction alone <span class="citation" data-cites="19zkt9R1G WDNuFZ4j DGlWGDEt">[<a href="#ref-19zkt9R1G" role="doc-biblioref">14</a>,<a href="#ref-WDNuFZ4j" role="doc-biblioref">15</a>,<a href="#ref-DGlWGDEt" role="doc-biblioref">16</a>]</span> or in combination with other features to recover DaG relationships <span class="citation" data-cites="CxErbNTp">[<a href="#ref-CxErbNTp" role="doc-biblioref">17</a>]</span>. One recent effort relied on a bi-clustering approach to detect DaG-relevant sentences from Pubmed abstracts <span class="citation" data-cites="CSiMoOrI">[<a href="#ref-CSiMoOrI" role="doc-biblioref">18</a>]</span> with clustering of dependency paths grouping similar sentences together. The results of this work supply our domain heuristic label functions. These approaches do not rely on a well-annotated training performance and tend to provide excellent recall, though the precision is often worse than with supervised methods <span class="citation" data-cites="199TFjkrC 1ZjlFRHa">[<a href="#ref-199TFjkrC" role="doc-biblioref">19</a>,<a href="#ref-1ZjlFRHa" role="doc-biblioref">20</a>]</span>.</p>
<p>Hand-crafted high-quality datasets <span class="citation" data-cites="hbAqN08A Y2DcwTrA luGt8luc 1Du6MinB8">[<a href="#ref-hbAqN08A" role="doc-biblioref">21</a>,<a href="#ref-Y2DcwTrA" role="doc-biblioref">22</a>,<a href="#ref-luGt8luc" role="doc-biblioref">23</a>,<a href="#ref-1Du6MinB8" role="doc-biblioref">24</a>]</span> often serve as a gold standard for training, tuning, and testing supervised machine learning methods in this setting. Support vector machines have been repeatedly used to detect DaG relationships <span class="citation" data-cites="hbAqN08A 3j1T67vB GeCe9qfW">[<a href="#ref-hbAqN08A" role="doc-biblioref">21</a>,<a href="#ref-3j1T67vB" role="doc-biblioref">25</a>,<a href="#ref-GeCe9qfW" role="doc-biblioref">26</a>]</span>. These models perform well in large feature spaces, but are slow to train as the number of data points becomes large. Recently, some studies have used deep neural network models. One used a pre-trained recurrent neural network <span class="citation" data-cites="riimmjYr">[<a href="#ref-riimmjYr" role="doc-biblioref">27</a>]</span>, and another used distant supervision <span class="citation" data-cites="k7ZUI6FL">[<a href="#ref-k7ZUI6FL" role="doc-biblioref">28</a>]</span>. Due to the success of these two models, we decided to use a deep neural network as our discriminative model.</p>
<h4 id="compound-treats-disease">Compound Treats Disease</h4>
<p>The goal of extracting Compound-treats-Disease (CtD) edges is to identify sentences that mention current drug treatments or propose new uses for existing drugs. One study combined an inference model from previously established drug-gene and gene-disease relationships to infer novel drug-disease interactions via co-occurrences <span class="citation" data-cites="ETC6lm7S">[<a href="#ref-ETC6lm7S" role="doc-biblioref">29</a>]</span>. A similar approach has also been applied to CtD extraction <span class="citation" data-cites="AdKPf5EO">[<a href="#ref-AdKPf5EO" role="doc-biblioref">30</a>]</span>. Manually-curated rules have also been applied to PubMed abstracts to address this task <span class="citation" data-cites="1avvFjJ9">[<a href="#ref-1avvFjJ9" role="doc-biblioref">31</a>]</span>. The rules were based on identifying key phrases and wordings related to using drugs to treat a disease, and we used these patterns as inspirations for some of our CtD label functions. Lastly, one study used a bi-clustering approach to identify sentences relevant to CtD edges <span class="citation" data-cites="CSiMoOrI">[<a href="#ref-CSiMoOrI" role="doc-biblioref">18</a>]</span>. As with DaG edges, we use the results from this study to provide what we term as domain heuristic label functions.</p>
<p>Recent work with supervised machine learning methods has often focused on compounds that induce a disease: an important question for toxicology and the subject of the BioCreative V dataset <span class="citation" data-cites="6wNuLZWb">[<a href="#ref-6wNuLZWb" role="doc-biblioref">32</a>]</span>. We don’t consider environmental toxicants in our work, as our source databases for distant supervision are primarily centered around FDA-approved therapies.</p>
<h4 id="compound-binds-gene">Compound Binds Gene</h4>
<p>The BioCreative VI track 5 task focused on classifying compound-protein interactions and has led to a great deal of work on the topic <span class="citation" data-cites="16As8893j">[<a href="#ref-16As8893j" role="doc-biblioref">33</a>]</span>. The equivalent edge in our networks is Compound-binds-Gene (CbG). Curators manually annotated 2,432 PubMed abstracts for five different compound protein interactions (agonist, antagonist, inhibitor, activator and substrate/product production) as part of the BioCreative task. The best performers on this task achieved an F1 score of 64.10% <span class="citation" data-cites="16As8893j">[<a href="#ref-16As8893j" role="doc-biblioref">33</a>]</span>. Numerous additional groups have now used the publicly available dataset, that resulted from this competition, to train supervised machine learning methods <span class="citation" data-cites="OnvaFHG9 i7KpvzCo 5LOkzCNK riimmjYr 5LOkzCNK 1H34cFSl8 16MGWGDUB 1HjIKY59u WP5p3RT3">[<a href="#ref-riimmjYr" role="doc-biblioref">27</a>,<a href="#ref-OnvaFHG9" role="doc-biblioref">34</a>,<a href="#ref-i7KpvzCo" role="doc-biblioref">35</a>,<a href="#ref-5LOkzCNK" role="doc-biblioref">36</a>,<a href="#ref-5LOkzCNK" role="doc-biblioref">36</a>,<a href="#ref-1H34cFSl8" role="doc-biblioref">37</a>,<a href="#ref-16MGWGDUB" role="doc-biblioref">38</a>,<a href="#ref-1HjIKY59u" role="doc-biblioref">39</a>,<a href="#ref-WP5p3RT3" role="doc-biblioref">40</a>]</span> and semi-supervised machine learning methods <span class="citation" data-cites="P2pnebCX">[<a href="#ref-P2pnebCX" role="doc-biblioref">41</a>]</span>. These approaches depend on well-annotated training datasets, which creates a bottleneck. In addition to supervised and semi-supervised machine learning methods, hand crafted rules <span class="citation" data-cites="107WYOcxW">[<a href="#ref-107WYOcxW" role="doc-biblioref">42</a>]</span> and bi-clustering of dependency trees <span class="citation" data-cites="CSiMoOrI">[<a href="#ref-CSiMoOrI" role="doc-biblioref">18</a>]</span> have been used. We use the results from the bi-clustering study to provide a subset of the CbG label functions in this work.</p>
<h4 id="gene-gene-interactions">Gene-Gene Interactions</h4>
<p>Akin to the DaG edge type, many efforts to extract Gene-interacts-Gene (GiG) relationships used co-occurrence approaches. This edge type is more frequently referred to as a protein-protein interaction. Even approaches as simple as calculating Z-scores from PubMed abstract co-occurrences can be informative <span class="citation" data-cites="q9Fhy8eq">[<a href="#ref-q9Fhy8eq" role="doc-biblioref">43</a>]</span>, and there are numerous studies using co-occurrences <span class="citation" data-cites="yGMDz6lK w32u0Rj9 8GVs1dBG DGlWGDEt">[<a href="#ref-DGlWGDEt" role="doc-biblioref">16</a>,<a href="#ref-yGMDz6lK" role="doc-biblioref">44</a>,<a href="#ref-w32u0Rj9" role="doc-biblioref">45</a>,<a href="#ref-8GVs1dBG" role="doc-biblioref">46</a>]</span>. However, more sophisticated strategies such as distant supervision appear to improve performance <span class="citation" data-cites="IGXdryzB">[<a href="#ref-IGXdryzB" role="doc-biblioref">13</a>]</span>. Similarly to the other edge types, the bi-clustering approach over dependency trees has also been applied to this edge type <span class="citation" data-cites="CSiMoOrI">[<a href="#ref-CSiMoOrI" role="doc-biblioref">18</a>]</span>. This manuscript provides a set of label functions for our work.</p>
<p>Most supervised classifiers used publicly available datasets for evaluation <span class="citation" data-cites="YWh6tPj DWpAeBxB szMMEMdC L9IIm3Zd 115pgEuOr">[<a href="#ref-YWh6tPj" role="doc-biblioref">47</a>,<a href="#ref-DWpAeBxB" role="doc-biblioref">48</a>,<a href="#ref-szMMEMdC" role="doc-biblioref">49</a>,<a href="#ref-L9IIm3Zd" role="doc-biblioref">50</a>,<a href="#ref-115pgEuOr" role="doc-biblioref">51</a>]</span>. These datasets are used equally among studies, but can generate noticeable differences in terms of performance <span class="citation" data-cites="DR8XM4Ff">[<a href="#ref-DR8XM4Ff" role="doc-biblioref">52</a>]</span>. Support vector machines were a common approach to extract GiG edges <span class="citation" data-cites="iiQkIqUX 1B0lnkj35">[<a href="#ref-iiQkIqUX" role="doc-biblioref">53</a>,<a href="#ref-1B0lnkj35" role="doc-biblioref">54</a>]</span>. However, with the growing popularity of deep learning numerous deep neural network architectures have been applied <span class="citation" data-cites="ibJfUvEe 1H4LpFrU0 bLKJwjMD P2pnebCX">[<a href="#ref-P2pnebCX" role="doc-biblioref">41</a>,<a href="#ref-ibJfUvEe" role="doc-biblioref">55</a>,<a href="#ref-1H4LpFrU0" role="doc-biblioref">56</a>,<a href="#ref-bLKJwjMD" role="doc-biblioref">57</a>]</span>. Distant supervision has also been used in this domain <span class="citation" data-cites="WYud0jQT">[<a href="#ref-WYud0jQT" role="doc-biblioref">58</a>]</span>, and in fact this effort was one of the motivating rationales for our work.</p>
<style> 
span.gene_color { color:#02b3e4 } 
span.disease_color { color:#875442 } 
span.compound_color { color:#e91e63 }
 </style>
<h2 id="materials-and-methods">Materials and Methods</h2>
<h3 id="hetionet">Hetionet</h3>
<div id="fig:hetionet" class="fignos">
<figure>
<img src="images/figures/hetionet/metagraph_highlighted_edges.png" alt="" /><figcaption><span>Figure 1:</span> A metagraph (schema) of Hetionet where biomedical entities are represented as nodes and the relationships between them are represented as edges. We examined performance on the highlighted subgraph; however, the long-term vision is to capture edges for the entire graph.</figcaption>
</figure>
</div>
<p>Hetionet <span class="citation" data-cites="O21tn8vf">[<a href="#ref-O21tn8vf" role="doc-biblioref">3</a>]</span> is a large heterogenous network that contains pharmacological and biological information. This network depicts information in the form of nodes and edges of different types: nodes that represent biological and pharmacological entities and edges which represent relationships between entities. Hetionet v1.0 contains 47,031 nodes with 11 different data types and 2,250,197 edges that represent 24 different relationship types (Figure <a href="#fig:hetionet">1</a>). Edges in Hetionet were obtained from open databases, such as the GWAS Catalog <span class="citation" data-cites="16cIDAXhG">[<a href="#ref-16cIDAXhG" role="doc-biblioref">59</a>]</span> and DrugBank <span class="citation" data-cites="1FI8iuYiQ">[<a href="#ref-1FI8iuYiQ" role="doc-biblioref">60</a>]</span>. For this project, we analyzed performance over a subset of the Hetionet relationship types: disease associates with a gene (DaG), compound binds to a gene (CbG), gene interacts with gene (GiG) and compound treating a disease (CtD).</p>
<h3 id="dataset">Dataset</h3>
<p>We used PubTator <span class="citation" data-cites="13vw5RIy4">[<a href="#ref-13vw5RIy4" role="doc-biblioref">61</a>]</span> as input to our analysis. PubTator provides MEDLINE abstracts that have been annotated with well-established entity recognition tools including DNorm <span class="citation" data-cites="vtuZ3Wx7">[<a href="#ref-vtuZ3Wx7" role="doc-biblioref">62</a>]</span> for disease mentions, GeneTUKit <span class="citation" data-cites="4S2HMNpa">[<a href="#ref-4S2HMNpa" role="doc-biblioref">63</a>]</span> for gene mentions, Gnorm <span class="citation" data-cites="1AkC7QdyP">[<a href="#ref-1AkC7QdyP" role="doc-biblioref">64</a>]</span> for gene normalizations and a dictionary based search system for compound mentions <span class="citation" data-cites="r501gnuM">[<a href="#ref-r501gnuM" role="doc-biblioref">65</a>]</span>. We downloaded PubTator on June 30, 2017, at which point it contained 10,775,748 abstracts. Then we filtered out mention tags that were not contained in hetionet. We used the Stanford CoreNLP parser <span class="citation" data-cites="RQkLuc5t">[<a href="#ref-RQkLuc5t" role="doc-biblioref">66</a>]</span> to tag parts of speech and generate dependency trees. We extracted sentences with two or more mentions, termed candidate sentences. Each candidate sentence was stratified by co-mention pair to produce a training set, tuning set and a testing set (shown in Table <a href="#tbl:candidate-sentences">1</a>). Each unique co-mention pair is sorted into four categories: (1) in hetionet and has sentences, (2) in hetionet and doesn’t have sentences, (3) not in hetionet and does have sentences and (4) not in hetionet and doesn’t have sentences. Within these four categories each pair is randomly assigned their own individual partition rank (continuous number between 0 and 1). Any rank lower than 0.7 is sorted into the training set, while any rank greater than 0.7 and lower than 0.9 is assigned to the tuning set. The rest of the pairs with a rank greater than or equal to 0.9 is assigned to the test set. Sentences that contain more than one co-mention pair are treated as multiple individual candidates. We hand labeled five hundred to a thousand candidate sentences of each relationship type to obtain a ground truth set (Table <a href="#tbl:candidate-sentences">1</a>)<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p>
<div id="tbl:candidate-sentences" class="tablenos">
<table>
<caption><span>Table 1:</span> Statistics of Candidate Sentences. We sorted each candidate sentence into a training, tuning and testing set. Numbers in parentheses show the number of positives and negatives that resulted from the hand-labeling process. </caption>
<thead>
<tr class="header">
<th style="text-align: left;">Relationship</th>
<th style="text-align: center;">Train</th>
<th style="text-align: center;">Tune</th>
<th style="text-align: center;">Test</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Disease Associates Gene</td>
<td style="text-align: center;">2.35 M</td>
<td style="text-align: center;">31K (397+, 603-)</td>
<td style="text-align: center;">313K (351+, 649-)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Compound Binds Gene</td>
<td style="text-align: center;">1.7M</td>
<td style="text-align: center;">468K (37+, 463-)</td>
<td style="text-align: center;">227k (31+, 469-)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Compound Treats Disease</td>
<td style="text-align: center;">1.013M</td>
<td style="text-align: center;">96K (96+, 404-)</td>
<td style="text-align: center;">32K (112+, 388-)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Gene Interacts Gene</td>
<td style="text-align: center;">12.6M</td>
<td style="text-align: center;">1.056M (60+, 440-)</td>
<td style="text-align: center;">257K (76+, 424-)</td>
</tr>
</tbody>
</table>
</div>
<h3 id="label-functions-for-annotating-sentences">Label Functions for Annotating Sentences</h3>
<p>The challenge of having too few ground truth annotations is common to many natural language processing settings, even when unannotated text is abundant. Data programming circumvents this issue by quickly annotating large datasets by using multiple noisy signals emitted by label functions <span class="citation" data-cites="5Il3kN32">[<a href="#ref-5Il3kN32" role="doc-biblioref">10</a>]</span>. Label functions are simple pythonic functions that emit: a positive label (1), a negative label (-1) or abstain from emitting a label (0). We combine these functions using a generative model to output a single annotation, which is a consensus probability score bounded between 0 (low chance of mentioning a relationship) and 1 (high chance of mentioning a relationship). We used these annotations to train a discriminator model that makes the final classification step.</p>
<h4 id="label-function-categories">Label Function Categories</h4>
<p>Label functions can be constructed in a multitude of ways; however, many label functions share similar characteristics with one another.<br />
We group these characteristics into the following categories: databases, text patterns and domain heuristics. Most of our label functions fall into the text pattern category, while the others were distributed across the database and domain heuristic categories (Table <a href="#tbl:label-functions">2</a>). We describe each category and provide an example using the candidate sentence: “<span class="gene_color">PTK6</span> may be a novel therapeutic target for <span class="disease_color">pancreatic cancer</span>.”.</p>
<p><strong>Databases</strong>: These label functions incorporate existing databases to generate a signal, as seen in distant supervision <span class="citation" data-cites="EHeTvZht">[<a href="#ref-EHeTvZht" role="doc-biblioref">4</a>]</span>. These functions detect if a candidate sentence’s co-mention pair is present in a given database. If the pair is present, our label function emits a positive label and abstains otherwise. If the pair is not present in any existing database, a separate label function emits a negative label. We used a separate label function to prevent a label imbalance problem that we encountered during development: emitting positives and negatives from the same label function causes downstream classifiers to generate almost exclusively negative predictions.</p>
<p><span class="math display">\[ \Lambda_{DB}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 1 &amp; (\color{#875442}{D}, \color{#02b3e4}{G}) \in DB \\
0 &amp; otherwise \\
\end{cases} \]</span></p>
<p><span class="math display">\[ \Lambda_{\neg DB}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 -1 &amp; (\color{#875442}{D}, \color{#02b3e4}{G}) \notin DB \\
0 &amp; otherwise \\
\end{cases} \]</span></p>
<p><strong>Domain Heuristics</strong>: These label functions used results from published text-based analyses to generate a signal. We used dependency path cluster themes generated by Percha et al. <span class="citation" data-cites="CSiMoOrI">[<a href="#ref-CSiMoOrI" role="doc-biblioref">18</a>]</span>. If a candidate sentence’s dependency path belonged to a previously generated cluster, then the label function emitted a positive label and abstained otherwise.</p>
<p><span class="math display">\[
\Lambda_{DH}(\color{#875442}{D}, \color{#02b3e4}{G}) = \begin{cases}
    1 &amp; Candidate \&gt; Sentence \in Cluster \&gt; Theme\\
    0 &amp; otherwise \\
    \end{cases}
\]</span></p>
<p><strong>Text Patterns</strong>: These label functions are designed to use keywords and sentence context to generate a signal. For example, a label function could focus on the number of words between two mentions or focus on the grammatical structure of a sentence. These functions emit a positive or negative label depending on the context.</p>
<p><span class="math display">\[ \Lambda_{TP}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 1 &amp; &quot;target&quot; \&gt; \in Candidate \&gt; Sentence \\
 0 &amp; otherwise \\
\end{cases} \]</span></p>
<p><span class="math display">\[ \Lambda_{TP}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 -1 &amp;   &quot;VB&quot; \&gt; \notin pos\_tags(Candidate \&gt; Sentence) \\
 0 &amp; otherwise \\
\end{cases} \]</span></p>
<p>Each text pattern label function was constructed by manual examination of sentences within the training set. For example, in the candidate sentence above one would extract the keywords “novel therapeutic target” and incorporate them in a text pattern label function. After initial construction, we tested and augmented the label function using sentences in the tune set. We repeated the above process for each label function in our repertoire.</p>
<div id="tbl:label-functions" class="tablenos">
<table>
<caption><span>Table 2:</span> The distribution of each label function per relationship. </caption>
<thead>
<tr class="header">
<th>Relationship</th>
<th style="text-align: center;">Databases (DB)</th>
<th style="text-align: center;">Text Patterns (TP)</th>
<th style="text-align: center;">Domain Heuristics (DH)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>DaG</td>
<td style="text-align: center;">7</td>
<td style="text-align: center;">20</td>
<td style="text-align: center;">10</td>
</tr>
<tr class="even">
<td>CtD</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">15</td>
<td style="text-align: center;">7</td>
</tr>
<tr class="odd">
<td>CbG</td>
<td style="text-align: center;">9</td>
<td style="text-align: center;">13</td>
<td style="text-align: center;">7</td>
</tr>
<tr class="even">
<td>GiG</td>
<td style="text-align: center;">9</td>
<td style="text-align: center;">20</td>
<td style="text-align: center;">8</td>
</tr>
</tbody>
</table>
</div>
<h3 id="training-models">Training Models</h3>
<h4 id="generative-model">Generative Model</h4>
<p>The generative model is a core part of this automatic annotation framework. It integrates multiple signals emitted by label functions and assigns a training class to each candidate sentence. This model assigns training classes by estimating the joint probability distribution of the latent true class (<span class="math inline">\(Y\)</span>) and label function signals (<span class="math inline">\(\Lambda\)</span>), (<span class="math inline">\(P_{\theta}(\Lambda, Y)\)</span>). Assuming each label function is conditionally independent, the joint distribution is defined as follows:</p>
<p><span class="math display">\[
P_{\theta}(\Lambda, Y) = \frac{\exp(\sum_{i=1}^{m} \theta^{T}F_{i}(\Lambda, y))}
{\sum_{\Lambda&#39;}\sum_{y&#39;} \exp(\sum_{i=1}^{m} \theta^{T}F_{i}(\Lambda&#39;, y&#39;))}
\]</span></p>
<p>where <span class="math inline">\(m\)</span> is the number of candidate sentences, <span class="math inline">\(F\)</span> is the vector of summary statistics and <span class="math inline">\(\theta\)</span> is a vector of weights for each summary statistic. The summary statistics used by the generative model are as follows:</p>
<p><span class="math display">\[F^{Lab}_{i,j}(\Lambda, Y) = \unicode{x1D7D9}\{\Lambda_{i,j} \neq 0\}\]</span> <span class="math display">\[F^{Acc}_{i,j}(\Lambda, Y) = \unicode{x1D7D9}\{\Lambda_{i,j} = y_{i,j}\}\]</span></p>
<p><em>Lab</em> is the label function’s propensity (the frequency of a label function emitting a signal). <em>Acc</em> is the individual label function’s accuracy given the training class. This model optimizes the weights (<span class="math inline">\(\theta\)</span>) by minimizing the negative log likelihood:</p>
<p><span class="math display">\[\hat{\theta} = argmin_{\theta} -\sum_{\Lambda} \sum_{Y} log P_{\theta}(\Lambda, Y)\]</span></p>
<p>In the framework we used predictions from the generative model, <span class="math inline">\(\hat{Y} = P_{\hat{\theta}}(Y \mid \Lambda)\)</span>, as training classes for our dataset <span class="citation" data-cites="vzoBuh4l 9Jo1af7Z">[<a href="#ref-vzoBuh4l" role="doc-biblioref">67</a>,<a href="#ref-9Jo1af7Z" role="doc-biblioref">68</a>]</span>.</p>
<h3 id="experimental-design">Experimental Design</h3>
<p>Being able to re-use label functions across edge types would substantially reduce the number of label functions required to extract multiple relationships from biomedical literature. We first established a baseline by training a generative model using only distant supervision label functions designed for the target edge type. For example, in the Gene interacts Gene (GiG) edge type we used label functions that returned a <strong>1</strong> if the pair of genes were included in the Human Interaction database <span class="citation" data-cites="LCyCrr7W">[<a href="#ref-LCyCrr7W" role="doc-biblioref">69</a>]</span>, the iRefIndex database <span class="citation" data-cites="gtV3bOpd">[<a href="#ref-gtV3bOpd" role="doc-biblioref">70</a>]</span> or in the Incomplete Interactome database <span class="citation" data-cites="2jkcXYxN">[<a href="#ref-2jkcXYxN" role="doc-biblioref">71</a>]</span>. Then we compared the baseline model with models that also included text and domain-heuristic label functions. Using a sampling with replacement approach, we sampled these text and domain-heuristic label functions separately within edge types, across edge types, and from a pool of all label functions. We compared within-edge-type performance to across-edge-type and all-edge-type performance. For each edge type we sampled a fixed number of label functions consisting of five evenly spaced numbers between one and the total number of possible label functions. We repeated this sampling process 50 times for each point. We evaluated both generative and discriminative (training and downstream analyses are described in the <a href="#discriminative-model">supplemental methods section</a>) models at each point, and report performance of each in terms of the area under the receiver operating characteristic curve (AUROC) and the area under the precision-recall curve (AUPR).</p>
<h2 id="results">Results</h2>
<h3 id="generative-model-using-randomly-sampled-label-functions">Generative Model Using Randomly Sampled Label Functions</h3>
<div id="fig:auroc_gen_model_performance" class="fignos">
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/ee638b4e45717a86f54a2744a813baaa90bc6b84/figures/label_sampling_experiment/transfer_test_set_auroc.png" alt="" /><figcaption><span>Figure 2:</span> Grid of AUROC scores for each generative model trained on randomly sampled label functions. The rows depict the relationship each model is trying to predict and the columns are the edge type specific sources from which each label function is sampled. The right most column consists of pooling every relationship specific label function and proceeding as above.</figcaption>
</figure>
</div>
<p>We added randomly sampled label functions to a baseline for each edge type to evaluate the feasibility of label function re-use. Our baseline model consisted of a generative model trained with only edge-specific distant supervision label functions. We reported the results in AUROC and AUPR (Figure <a href="#fig:auroc_gen_model_performance">2</a> and Supplemental Figure <a href="#fig:aupr_gen_model_performance">6</a>).<br />
The on-diagonal plots of figure <a href="#fig:auroc_gen_model_performance">2</a> and supplemental figure <a href="#fig:aupr_gen_model_performance">6</a> show increasing performance when edge-specific label functions are added on top of the edge-specific baselines. The CtD edge type is a quintessential example of this trend. The baseline model starts off with an AUROC score of 52% and an AUPRC of 28%, which increase to 76% and 49% respectively as more CtD label functions are included. DaG edges have a similar trend: performance starting off with an AUROC of 56% and AUPR of 41% then increases to 62% and 45% respectively. Both the CbG and GiG edges have an increasing trend but plateau after a few label functions are added.</p>
<p>The off-diagonals in figure <a href="#fig:auroc_gen_model_performance">2</a> and supplemental figure <a href="#fig:aupr_gen_model_performance">6</a> show how performance varies when label functions from one edge type are added to a different edge type’s baseline. In certain cases (apparent for DaG), performance increases regardless of the edge type used for label functions. In other cases (apparent with CtD), one label function appears to improve performance; however, adding more label functions does not improve performance (AUROC) or decreases it (AUPR). In certain cases, the source of the label functions appears to be important: the performance of CbG edges decrease when using label functions from the DaG and CtD categories.</p>
<p>Our initial hypothesis was based on the idea that certain edge types capture similar physical relationships and that these cases would be particularly amenable for label function transfer. For example, CbG and GiG both describe physical interactions. We observed that performance increased as assessed by both AUROC and AUPR when using label functions from the GiG edge type to predict CbG edges. A similar trend was observed when predicting the GiG edge; however, the performance differences were small for this edge type making the importance difficult to assess.<br />
The last column shows increasing performance (AUROC and AUPR) for both DaG and CtD when sampling from all label functions. CbG and GiG also had increased performance when one random label function was sampled, but performance decreased drastically as more label functions were added. It is possible that a small number of irrelevant label functions are able to overwhelm the distant supervision label functions in these cases (see Figure <a href="#fig:auroc_random_label_function_performance">3</a> and Supplemental Figure <a href="#fig:aupr_random_label_function_performance">7</a>).</p>
<h3 id="random-label-function-generative-model-analysis">Random Label Function Generative Model Analysis</h3>
<div id="fig:auroc_random_label_function_performance" class="fignos">
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/ee638b4e45717a86f54a2744a813baaa90bc6b84/figures/gen_model_error_analysis/transfer_test_set_auroc.png" alt="" /><figcaption><span>Figure 3:</span> A grid of AUROC (A) scores for each edge type. Each plot consists of adding a single label function on top of the baseline model. This label function emits a positive (shown in blue) or negative (shown in orange) label at specified frequencies, and performance at zero is equivalent to not having a randomly emitting label function. The error bars represent 95% confidence intervals for AUROC or AUPR (y-axis) at each emission frequency.</figcaption>
</figure>
</div>
<p>We observed that including one label function of a mismatched type to distant supervision often improved performance, so we evaluated the effects of adding a random label function in the same setting. We found that usually adding random noise did not improve performance (Figure <a href="#fig:auroc_random_label_function_performance">3</a> and Supplemental Figure <a href="#fig:aupr_random_label_function_performance">7</a>). For the CbG edge type we did observe slightly increased performance via AUPR (Supplemental Figure <a href="#fig:aupr_random_label_function_performance">7</a>). However, performance changes in general were smaller than those observed with mismatched label types.</p>
<h3 id="discriminative-model-calibration">Discriminative Model Calibration</h3>
<div id="fig:discriminative_model_calibration" class="fignos">
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/86037d185a299a1f6dd4dd68605073849c72af6f/figures/model_calibration_experiment/model_calibration.png" alt="" /><figcaption><span>Figure 4:</span> Deep learning models are overconfident in their predictions and need to be calibrated after training. These are calibration plots for the discrimintative model. The green line represents the predictions before calibration and the blue line shows predictions after calibration. Data points that lie closer to diagonal line show better model calibration, while data points far from the diagonal show poor performance. A perfectly calibrated model would align straight along the diagonal line.</figcaption>
</figure>
</div>
<p>Even deep learning models with good AUROC and AUPR statistics can be subject to poor calibration. Typically, these models are overconfident in their predictions <span class="citation" data-cites="QJ6hYH8N rLVjMJ5l">[<a href="#ref-QJ6hYH8N" role="doc-biblioref">72</a>,<a href="#ref-rLVjMJ5l" role="doc-biblioref">73</a>]</span>. We attempted to use temperature scaling to fix the calibration of the best performing discriminative models (Figure <a href="#fig:discriminative_model_calibration">4</a>). Before calibration (green lines), our models were aligned with the ideal calibration only when predicting low probability scores (close to 0.25). Applying the temperature scaling calibration algorithm (blue lines) did not substantially improve the calibration of the model in most cases. The exception to this pattern is the Disease associates Gene (DaG) model where high confidence scores are shown to be better calibrated. Overall, calbrating deep learning models is a nontrivial task that requires more complex approaches to accomplish.</p>
<h2 id="discussion">Discussion</h2>
<p>We tested the feasibility of re-using label functions to extract relationships from literature. Through our sampling experiment, we found that adding relevant label functions increases prediction performance (shown in the on-diagonals of Figures <a href="#fig:auroc_gen_model_performance">2</a> and Supplemental Figure <a href="#fig:aupr_gen_model_performance">6</a>). We found that label functions designed from relatively related edge types can increase performance (seen when GiG label functions predicts CbG and vice versa). We noticed that one edge type (DaG) is agnostic to label function source (Figure <a href="#fig:auroc_gen_model_performance">2</a> and Supplemental Figure <a href="#fig:aupr_gen_model_performance">6</a>). Performance routinely increases when adding a single mismatched label function to our baseline model (the generative model trained only on distant supervision label functions). These results led us to hypothesize that adding a small amount of noise aided the model, but our experiment with a random label function reveals that this was not the case (Figures <a href="#fig:auroc_random_label_function_performance">3</a> and <a href="#fig:aupr_random_label_function_performance">7</a>). Based on these results one question still remains: why does performance drastically increase when adding a single label function to our distant supervision baseline?</p>
<p>The discriminative model didn’t work as intended. The majority of the time the discriminative model underperformed the generative model (Supplemental Figures <a href="#fig:auroc_discriminative_model_performance">8</a> and <a href="#fig:aupr_discriminative_model_performance">9</a>). Potential reasons for this are the discriminative model overfitting to the generative model’s predictions and a negative class bias in some of our datasets (Table <a href="#tbl:candidate-sentences">1</a>). The challenges with the discriminative model are likely to have led to issues in our downstream analyses: poor model calibration (Supplemental Figure <a href="#fig:discriminative_model_calibration">4</a>) and poor recall in detecting existing Hetionet edges (Supplemental Figure <a href="#fig:hetionet_reconstruction">11</a>). Despite the above complications, our model had similar performance with a published baseline model (Supplemental Figure <a href="#fig:cocoscore_comparison">10</a>). This implies that with better tuning the discriminative model has the potential to perform better than the baseline model.</p>
<h2 id="conclusion-and-future-direction">Conclusion and Future Direction</h2>
<p>Filling out knowledge bases via manual curation can be an arduous and erroneous task <span class="citation" data-cites="UdzvLgBM">[<a href="#ref-UdzvLgBM" role="doc-biblioref">8</a>]</span>. As the rate of publications increases manual curation becomes an infeasible approach. Data programming, a paradigm that uses label functions as a means to speed up the annotation process, can be used as a solution for this problem. A problem with this paradigm is that creating a useful label function takes a significant amount of time. We tested the feasibility of reusing label functions as a way to speed up the label function creation process. We conclude that label function re-use across edge types can increase performance when there are certain constraints on the number of functions re-used. More sophisticated methods of reuse may be able to capture many of the advantages and avoid many of the drawbacks. Adding more relevant label functions can increase overall performance. The discriminative model, under this paradigm, has a tendency to overfit to predictions of the generative model. We recommend implementing regularization techniques such as drop out and weight decay to combat this issue.</p>
<p>This work sets up the foundation for creating a common framework that mines text to create edges. Within this framework we would continuously ingest new knowledge as novel findings are published, while providing a single confidence score for an edge by consolidating sentence scores. Different from existing hetnets like Hetionet where text-derived edges generally cannot be exactly attributed to excerpts from literature <span class="citation" data-cites="O21tn8vf L2B5V7XC">[<a href="#ref-O21tn8vf" role="doc-biblioref">3</a>,<a href="#ref-L2B5V7XC" role="doc-biblioref">74</a>]</span>, our approach would annotate each edge with its source sentences. In addition, edges generated with this approach would be unencumbered from upstream licensing or copyright restrictions, enabling openly licensed hetnets at a scale not previously possible <span class="citation" data-cites="4G0GW8oe 137tbemL9 1GwdMLPbV">[<a href="#ref-4G0GW8oe" role="doc-biblioref">75</a>,<a href="#ref-137tbemL9" role="doc-biblioref">76</a>,<a href="#ref-1GwdMLPbV" role="doc-biblioref">77</a>]</span>. Accordingly, we plan to use this framework to create a robust multi-edge extractor via multitask learning <span class="citation" data-cites="9Jo1af7Z">[<a href="#ref-9Jo1af7Z" role="doc-biblioref">68</a>]</span> to construct continuously updating literature-derived hetnets.</p>
<h2 id="supplemental-information">Supplemental Information</h2>
<p>This manuscript and supplemental information are available at <a href="https://greenelab.github.io/text_mined_hetnet_manuscript/" class="uri">https://greenelab.github.io/text_mined_hetnet_manuscript/</a>. Source code for this work is available under open licenses at: <a href="https://github.com/greenelab/snorkeling/" class="uri">https://github.com/greenelab/snorkeling/</a>.</p>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>The authors would like to thank Christopher Ré’s group at Stanford Univeristy, especially Alex Ratner and Steven Bach, for their assistance with this project. We also want to thank Graciela Gonzalez-Hernandez for her advice and input with this project. This work was support by <a href="https://www.moore.org/grant-detail?grantId=GBMF4552">Grant GBMF4552</a> from the Gordon Betty Moore Foundation.</p>
<h2 id="references" class="page_break_before">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" role="doc-bibliography">
<div id="ref-u8pIAt5j">
<p>1. <strong>Graph Theory Enables Drug Repurposing – How a Mathematical Model Can Drive the Discovery of Hidden Mechanisms of Action</strong><br />
Ruggero Gramatica, T. Di Matteo, Stefano Giorgetti, Massimo Barbiani, Dorian Bevec, Tomaso Aste<br />
<em>PLoS ONE</em> (2014-01-09) <a href="https://doi.org/gf45zp">https://doi.org/gf45zp</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0084912">10.1371/journal.pone.0084912</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24416311">24416311</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3886994">PMC3886994</a></p>
</div>
<div id="ref-bPvC638e">
<p>2. <strong>Drug repurposing through joint learning on knowledge graphs and literature</strong><br />
Mona Alshahrani, Robert Hoehndorf<br />
<em>Cold Spring Harbor Laboratory</em> (2018-08-06) <a href="https://doi.org/gf45zk">https://doi.org/gf45zk</a><br />
DOI: <a href="https://doi.org/10.1101/385617">10.1101/385617</a></p>
</div>
<div id="ref-O21tn8vf">
<p>3. <strong>Systematic integration of biomedical knowledge prioritizes drugs for repurposing</strong><br />
Daniel Scott Himmelstein, Antoine Lizee, Christine Hessler, Leo Brueggeman, Sabrina L Chen, Dexter Hadley, Ari Green, Pouya Khankhanian, Sergio E Baranzini<br />
<em>eLife</em> (2017-09-22) <a href="https://doi.org/cdfk">https://doi.org/cdfk</a><br />
DOI: <a href="https://doi.org/10.7554/elife.26726">10.7554/elife.26726</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28936969">28936969</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5640425">PMC5640425</a></p>
</div>
<div id="ref-EHeTvZht">
<p>4. <strong>Distant supervision for relation extraction without labeled data</strong><br />
Mike Mintz, Steven Bills, Rion Snow, Dan Jurafsky<br />
<em>Proceedings of the Joint Conference of the 47th Annual Meeting of the ACL and the 4th International Joint Conference on Natural Language Processing of the AFNLP: Volume 2 - ACL-IJCNLP ’09</em> (2009) <a href="https://doi.org/fg9q43">https://doi.org/fg9q43</a><br />
DOI: <a href="https://doi.org/10.3115/1690219.1690287">10.3115/1690219.1690287</a></p>
</div>
<div id="ref-CVHSURuI">
<p>5. <strong>CoCoScore: Context-aware co-occurrence scoring for text mining applications using distant supervision</strong><br />
Alexander Junge, Lars Juhl Jensen<br />
<em>Cold Spring Harbor Laboratory</em> (2018-10-16) <a href="https://doi.org/gf45zm">https://doi.org/gf45zm</a><br />
DOI: <a href="https://doi.org/10.1101/444398">10.1101/444398</a></p>
</div>
<div id="ref-HS4ARwmZ">
<p>6. <strong>Knowledge-guided convolutional networks for chemical-disease relation extraction</strong><br />
Huiwei Zhou, Chengkun Lang, Zhuang Liu, Shixian Ning, Yingyu Lin, Lei Du<br />
<em>BMC Bioinformatics</em> (2019-05-21) <a href="https://doi.org/gf45zn">https://doi.org/gf45zn</a><br />
DOI: <a href="https://doi.org/10.1186/s12859-019-2873-7">10.1186/s12859-019-2873-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31113357">31113357</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6528333">PMC6528333</a></p>
</div>
<div id="ref-N1Ai0gaI">
<p>7. <strong>Facts from text: can text mining help to scale-up high-quality manual curation of gene products with ontologies?</strong><br />
R. Winnenburg, T. Wachter, C. Plake, A. Doms, M. Schroeder<br />
<em>Briefings in Bioinformatics</em> (2008-07-11) <a href="https://doi.org/bfsnwg">https://doi.org/bfsnwg</a><br />
DOI: <a href="https://doi.org/10.1093/bib/bbn043">10.1093/bib/bbn043</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19060303">19060303</a></p>
</div>
<div id="ref-UdzvLgBM">
<p>8. <strong>Manual curation is not sufficient for annotation of genomic databases</strong><br />
William A. Baumgartner Jr, K. Bretonnel Cohen, Lynne M. Fox, George Acquaah-Mensah, Lawrence Hunter<br />
<em>Bioinformatics</em> (2007-07-01) <a href="https://doi.org/dtck86">https://doi.org/dtck86</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btm229">10.1093/bioinformatics/btm229</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17646325">17646325</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2516305">PMC2516305</a></p>
</div>
<div id="ref-1DBISRlwN">
<p>9. <strong>Growth rates of modern science: A bibliometric analysis based on the number of publications and cited references</strong><br />
Lutz Bornmann, Rüdiger Mutz<br />
<em>Journal of the Association for Information Science and Technology</em> (2015-04-29) <a href="https://doi.org/gfj5zc">https://doi.org/gfj5zc</a><br />
DOI: <a href="https://doi.org/10.1002/asi.23329">10.1002/asi.23329</a></p>
</div>
<div id="ref-5Il3kN32">
<p>10. <strong>Data Programming: Creating Large Training Sets, Quickly</strong><br />
Alexander Ratner, Christopher De Sa, Sen Wu, Daniel Selsam, Christopher Ré<br />
<em>arXiv</em> (2016-05-25) <a href="https://arxiv.org/abs/1605.07723v3">https://arxiv.org/abs/1605.07723v3</a></p>
</div>
<div id="ref-NLxmpSdj">
<p>11. <strong>PKDE4J: Entity and relation extraction for public knowledge discovery.</strong><br />
Min Song, Won Chul Kim, Dahee Lee, Go Eun Heo, Keun Young Kang<br />
<em>Journal of biomedical informatics</em> (2015-08-12) <a href="https://www.ncbi.nlm.nih.gov/pubmed/26277115">https://www.ncbi.nlm.nih.gov/pubmed/26277115</a><br />
DOI: <a href="https://doi.org/10.1016/j.jbi.2015.08.008">10.1016/j.jbi.2015.08.008</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26277115">26277115</a></p>
</div>
<div id="ref-5gG8hwv7">
<p>12. <strong>DISEASES: Text mining and data integration of disease–gene associations</strong><br />
Sune Pletscher-Frankild, Albert Pallejà, Kalliopi Tsafou, Janos X. Binder, Lars Juhl Jensen<br />
<em>Methods</em> (2015-03) <a href="https://doi.org/f3mn6s">https://doi.org/f3mn6s</a><br />
DOI: <a href="https://doi.org/10.1016/j.ymeth.2014.11.020">10.1016/j.ymeth.2014.11.020</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25484339">25484339</a></p>
</div>
<div id="ref-IGXdryzB">
<p>13. <strong>CoCoScore: context-aware co-occurrence scoring for text mining applications using distant supervision</strong><br />
Alexander Junge, Lars Juhl Jensen<br />
<em>Bioinformatics</em> (2019-06-14) <a href="https://doi.org/gf4789">https://doi.org/gf4789</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btz490">10.1093/bioinformatics/btz490</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31199464">31199464</a></p>
</div>
<div id="ref-19zkt9R1G">
<p>14. <strong>LGscore: A method to identify disease-related genes using biological literature and Google data</strong><br />
Jeongwoo Kim, Hyunjin Kim, Youngmi Yoon, Sanghyun Park<br />
<em>Journal of Biomedical Informatics</em> (2015-04) <a href="https://doi.org/f7bj9c">https://doi.org/f7bj9c</a><br />
DOI: <a href="https://doi.org/10.1016/j.jbi.2015.01.003">10.1016/j.jbi.2015.01.003</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25617670">25617670</a></p>
</div>
<div id="ref-WDNuFZ4j">
<p>15. <strong>PolySearch2: a significantly improved text-mining system for discovering associations between human diseases, genes, drugs, metabolites, toxins and more</strong><br />
Yifeng Liu, Yongjie Liang, David Wishart<br />
<em>Nucleic Acids Research</em> (2015-04-29) <a href="https://doi.org/f7nzn5">https://doi.org/f7nzn5</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkv383">10.1093/nar/gkv383</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25925572">25925572</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4489268">PMC4489268</a></p>
</div>
<div id="ref-DGlWGDEt">
<p>16. <strong>A comprehensive and quantitative comparison of text-mining in 15 million full-text articles versus their corresponding abstracts</strong><br />
David Westergaard, Hans-Henrik Stærfeldt, Christian Tønsberg, Lars Juhl Jensen, Søren Brunak<br />
<em>PLOS Computational Biology</em> (2018-02-15) <a href="https://doi.org/gcx747">https://doi.org/gcx747</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pcbi.1005962">10.1371/journal.pcbi.1005962</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29447159">29447159</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5831415">PMC5831415</a></p>
</div>
<div id="ref-CxErbNTp">
<p>17. <strong>The research on gene-disease association based on text-mining of PubMed</strong><br />
Jie Zhou, Bo-quan Fu<br />
<em>BMC Bioinformatics</em> (2018-02-07) <a href="https://doi.org/gf479k">https://doi.org/gf479k</a><br />
DOI: <a href="https://doi.org/10.1186/s12859-018-2048-y">10.1186/s12859-018-2048-y</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29415654">29415654</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5804013">PMC5804013</a></p>
</div>
<div id="ref-CSiMoOrI">
<p>18. <strong>A global network of biomedical relationships derived from text</strong><br />
Bethany Percha, Russ B Altman<br />
<em>Bioinformatics</em> (2018-02-27) <a href="https://doi.org/gc3ndk">https://doi.org/gc3ndk</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/bty114">10.1093/bioinformatics/bty114</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29490008">29490008</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6061699">PMC6061699</a></p>
</div>
<div id="ref-199TFjkrC">
<p>19. <strong>Literature mining for the biologist: from information retrieval to biological discovery</strong><br />
Lars Juhl Jensen, Jasmin Saric, Peer Bork<br />
<em>Nature Reviews Genetics</em> (2006-02) <a href="https://doi.org/bgq7q9">https://doi.org/bgq7q9</a><br />
DOI: <a href="https://doi.org/10.1038/nrg1768">10.1038/nrg1768</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/16418747">16418747</a></p>
</div>
<div id="ref-1ZjlFRHa">
<p>20. <strong>Application of text mining in the biomedical domain</strong><br />
Wilco W. M. Fleuren, Wynand Alkema<br />
<em>Methods</em> (2015-03) <a href="https://doi.org/f64p6n">https://doi.org/f64p6n</a><br />
DOI: <a href="https://doi.org/10.1016/j.ymeth.2015.01.015">10.1016/j.ymeth.2015.01.015</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25641519">25641519</a></p>
</div>
<div id="ref-hbAqN08A">
<p>21. <strong>Extraction of relations between genes and diseases from text and large-scale data analysis: implications for translational research</strong><br />
Àlex Bravo, Janet Piñero, Núria Queralt-Rosinach, Michael Rautschka, Laura I Furlong<br />
<em>BMC Bioinformatics</em> (2015-02-21) <a href="https://doi.org/f7kn8s">https://doi.org/f7kn8s</a><br />
DOI: <a href="https://doi.org/10.1186/s12859-015-0472-9">10.1186/s12859-015-0472-9</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25886734">25886734</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4466840">PMC4466840</a></p>
</div>
<div id="ref-Y2DcwTrA">
<p>22. <strong>The EU-ADR corpus: Annotated drugs, diseases, targets, and their relationships</strong><br />
Erik M. van Mulligen, Annie Fourrier-Reglat, David Gurwitz, Mariam Molokhia, Ainhoa Nieto, Gianluca Trifiro, Jan A. Kors, Laura I. Furlong<br />
<em>Journal of Biomedical Informatics</em> (2012-10) <a href="https://doi.org/f36vn6">https://doi.org/f36vn6</a><br />
DOI: <a href="https://doi.org/10.1016/j.jbi.2012.04.004">10.1016/j.jbi.2012.04.004</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22554700">22554700</a></p>
</div>
<div id="ref-luGt8luc">
<p>23. <strong>CoMAGC: a corpus with multi-faceted annotations of gene-cancer relations</strong><br />
Hee-Jin Lee, Sang-Hyung Shim, Mi-Ryoung Song, Hyunju Lee, Jong C Park<br />
<em>BMC Bioinformatics</em> (2013) <a href="https://doi.org/gb8v5s">https://doi.org/gb8v5s</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-14-323">10.1186/1471-2105-14-323</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24225062">24225062</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3833657">PMC3833657</a></p>
</div>
<div id="ref-1Du6MinB8">
<p>24. <strong>Concept annotation in the CRAFT corpus</strong><br />
Michael Bada, Miriam Eckert, Donald Evans, Kristin Garcia, Krista Shipley, Dmitry Sitnikov, William A Baumgartner, K Bretonnel Cohen, Karin Verspoor, Judith A Blake, Lawrence E Hunter<br />
<em>BMC Bioinformatics</em> (2012-07-09) <a href="https://doi.org/gb8vdr">https://doi.org/gb8vdr</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-13-161">10.1186/1471-2105-13-161</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22776079">22776079</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3476437">PMC3476437</a></p>
</div>
<div id="ref-3j1T67vB">
<p>25. <strong>DTMiner: identification of potential disease targets through biomedical literature mining</strong><br />
Dong Xu, Meizhuo Zhang, Yanping Xie, Fan Wang, Ming Chen, Kenny Q. Zhu, Jia Wei<br />
<em>Bioinformatics</em> (2016-08-09) <a href="https://doi.org/f9nw36">https://doi.org/f9nw36</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btw503">10.1093/bioinformatics/btw503</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27506226">27506226</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5181534">PMC5181534</a></p>
</div>
<div id="ref-GeCe9qfW">
<p>26. <strong>Automatic extraction of gene-disease associations from literature using joint ensemble learning</strong><br />
Balu Bhasuran, Jeyakumar Natarajan<br />
<em>PLOS ONE</em> (2018-07-26) <a href="https://doi.org/gdx63f">https://doi.org/gdx63f</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0200699">10.1371/journal.pone.0200699</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30048465">30048465</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6061985">PMC6061985</a></p>
</div>
<div id="ref-riimmjYr">
<p>27. <strong>BioBERT: a pre-trained biomedical language representation model for biomedical text mining</strong><br />
Jinhyuk Lee, Wonjin Yoon, Sungdong Kim, Donghyeon Kim, Sunkyu Kim, Chan Ho So, Jaewoo Kang<br />
<em>arXiv</em> (2019-01-25) <a href="https://arxiv.org/abs/1901.08746v4">https://arxiv.org/abs/1901.08746v4</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btz682">10.1093/bioinformatics/btz682</a></p>
</div>
<div id="ref-k7ZUI6FL">
<p>28. <strong>Distant Supervision for Large-Scale Extraction of Gene–Disease Associations from Literature Using DeepDive</strong><br />
Balu Bhasuran, Jeyakumar Natarajan<br />
<em>International Conference on Innovative Computing and Communications</em> (2018-11-20) <a href="https://doi.org/gf5hfv">https://doi.org/gf5hfv</a><br />
DOI: <a href="https://doi.org/10.1007/978-981-13-2354-6_39">10.1007/978-981-13-2354-6_39</a></p>
</div>
<div id="ref-ETC6lm7S">
<p>29. <strong>A new method for prioritizing drug repositioning candidates extracted by literature-based discovery</strong><br />
Majid Rastegar-Mojarad, Ravikumar Komandur Elayavilli, Dingcheng Li, Rashmi Prasad, Hongfang Liu<br />
<em>2015 IEEE International Conference on Bioinformatics and Biomedicine (BIBM)</em> (2015-11) <a href="https://doi.org/gf479j">https://doi.org/gf479j</a><br />
DOI: <a href="https://doi.org/10.1109/bibm.2015.7359766">10.1109/bibm.2015.7359766</a></p>
</div>
<div id="ref-AdKPf5EO">
<p>30. <strong>Literature Mining for the Discovery of Hidden Connections between Drugs, Genes and Diseases</strong><br />
Raoul Frijters, Marianne van Vugt, Ruben Smeets, René van Schaik, Jacob de Vlieg, Wynand Alkema<br />
<em>PLoS Computational Biology</em> (2010-09-23) <a href="https://doi.org/bhrw7x">https://doi.org/bhrw7x</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pcbi.1000943">10.1371/journal.pcbi.1000943</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/20885778">20885778</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2944780">PMC2944780</a></p>
</div>
<div id="ref-1avvFjJ9">
<p>31. <strong>Large-scale extraction of accurate drug-disease treatment pairs from biomedical literature for drug repurposing</strong><br />
Rong Xu, QuanQiu Wang<br />
<em>BMC Bioinformatics</em> (2013-06-06) <a href="https://doi.org/gb8v3k">https://doi.org/gb8v3k</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-14-181">10.1186/1471-2105-14-181</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23742147">23742147</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3702428">PMC3702428</a></p>
</div>
<div id="ref-6wNuLZWb">
<p>32. <strong>BioCreative V CDR task corpus: a resource for chemical disease relation extraction</strong><br />
Jiao Li, Yueping Sun, Robin J. Johnson, Daniela Sciaky, Chih-Hsuan Wei, Robert Leaman, Allan Peter Davis, Carolyn J. Mattingly, Thomas C. Wiegers, Zhiyong Lu<br />
<em>Database</em> (2016) <a href="https://doi.org/gf5hfw">https://doi.org/gf5hfw</a><br />
DOI: <a href="https://doi.org/10.1093/database/baw068">10.1093/database/baw068</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27161011">27161011</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4860626">PMC4860626</a></p>
</div>
<div id="ref-16As8893j">
<p>33. <strong>Overview of the biocreative vi chemical-protein interaction track</strong><br />
Martin Krallinger, Obdulia Rabal, Saber A Akhondi, others<br />
<em>Proceedings of the sixth biocreative challenge evaluation workshop</em> (2017) <a href="https://www.semanticscholar.org/paper/Overview-of-the-BioCreative-VI-chemical-protein-Krallinger-Rabal/eed781f498b563df5a9e8a241c67d63dd1d92ad5">https://www.semanticscholar.org/paper/Overview-of-the-BioCreative-VI-chemical-protein-Krallinger-Rabal/eed781f498b563df5a9e8a241c67d63dd1d92ad5</a></p>
</div>
<div id="ref-OnvaFHG9">
<p>34. <strong>LimTox: a web tool for applied text mining of adverse event and toxicity associations of compounds, drugs and genes</strong><br />
Andres Cañada, Salvador Capella-Gutierrez, Obdulia Rabal, Julen Oyarzabal, Alfonso Valencia, Martin Krallinger<br />
<em>Nucleic Acids Research</em> (2017-05-22) <a href="https://doi.org/gf479h">https://doi.org/gf479h</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkx462">10.1093/nar/gkx462</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28531339">28531339</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5570141">PMC5570141</a></p>
</div>
<div id="ref-i7KpvzCo">
<p>35. <strong>LPTK: a linguistic pattern-aware dependency tree kernel approach for the BioCreative VI CHEMPROT task</strong><br />
Neha Warikoo, Yung-Chun Chang, Wen-Lian Hsu<br />
<em>Database</em> (2018-01-01) <a href="https://doi.org/gfhjr6">https://doi.org/gfhjr6</a><br />
DOI: <a href="https://doi.org/10.1093/database/bay108">10.1093/database/bay108</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30346607">30346607</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6196310">PMC6196310</a></p>
</div>
<div id="ref-5LOkzCNK">
<p>36. <strong>Extracting chemical–protein relations with ensembles of SVM and deep learning models</strong><br />
Yifan Peng, Anthony Rios, Ramakanth Kavuluru, Zhiyong Lu<br />
<em>Database</em> (2018-01-01) <a href="https://doi.org/gf479f">https://doi.org/gf479f</a><br />
DOI: <a href="https://doi.org/10.1093/database/bay073">10.1093/database/bay073</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30020437">30020437</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6051439">PMC6051439</a></p>
</div>
<div id="ref-1H34cFSl8">
<p>37. <strong>Extracting chemical–protein interactions from literature using sentence structure analysis and feature engineering</strong><br />
Pei-Yau Lung, Zhe He, Tingting Zhao, Disa Yu, Jinfeng Zhang<br />
<em>Database</em> (2019-01-01) <a href="https://doi.org/gf479g">https://doi.org/gf479g</a><br />
DOI: <a href="https://doi.org/10.1093/database/bay138">10.1093/database/bay138</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30624652">30624652</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6323317">PMC6323317</a></p>
</div>
<div id="ref-16MGWGDUB">
<p>38. <strong>Improving the learning of chemical-protein interactions from literature using transfer learning and specialized word embeddings</strong><br />
P Corbett, J Boyle<br />
<em>Database</em> (2018-01-01) <a href="https://doi.org/gf479d">https://doi.org/gf479d</a><br />
DOI: <a href="https://doi.org/10.1093/database/bay066">10.1093/database/bay066</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30010749">30010749</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6044291">PMC6044291</a></p>
</div>
<div id="ref-1HjIKY59u">
<p>39. <strong>Extracting chemical–protein relations using attention-based neural networks</strong><br />
Sijia Liu, Feichen Shen, Ravikumar Komandur Elayavilli, Yanshan Wang, Majid Rastegar-Mojarad, Vipin Chaudhary, Hongfang Liu<br />
<em>Database</em> (2018-01-01) <a href="https://doi.org/gfdz8d">https://doi.org/gfdz8d</a><br />
DOI: <a href="https://doi.org/10.1093/database/bay102">10.1093/database/bay102</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30295724">30295724</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6174551">PMC6174551</a></p>
</div>
<div id="ref-WP5p3RT3">
<p>40. <strong>Chemical–gene relation extraction using recursive neural network</strong><br />
Sangrak Lim, Jaewoo Kang<br />
<em>Database</em> (2018-01-01) <a href="https://doi.org/gdss6f">https://doi.org/gdss6f</a><br />
DOI: <a href="https://doi.org/10.1093/database/bay060">10.1093/database/bay060</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29961818">29961818</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6014134">PMC6014134</a></p>
</div>
<div id="ref-P2pnebCX">
<p>41. <strong>Exploring Semi-supervised Variational Autoencoders for Biomedical Relation Extraction</strong><br />
Yijia Zhang, Zhiyong Lu<br />
<em>arXiv</em> (2019-01-18) <a href="https://arxiv.org/abs/1901.06103v1">https://arxiv.org/abs/1901.06103v1</a></p>
</div>
<div id="ref-107WYOcxW">
<p>42. <strong>Pharmspresso: a text mining tool for extraction of pharmacogenomic concepts and relationships from full text</strong><br />
Yael Garten, Russ B Altman<br />
<em>BMC Bioinformatics</em> (2009-02) <a href="https://doi.org/df75hq">https://doi.org/df75hq</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-10-s2-s6">10.1186/1471-2105-10-s2-s6</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19208194">19208194</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2646239">PMC2646239</a></p>
</div>
<div id="ref-q9Fhy8eq">
<p>43. <strong>STRING v10: protein–protein interaction networks, integrated over the tree of life</strong><br />
Damian Szklarczyk, Andrea Franceschini, Stefan Wyder, Kristoffer Forslund, Davide Heller, Jaime Huerta-Cepas, Milan Simonovic, Alexander Roth, Alberto Santos, Kalliopi P. Tsafou, … Christian von Mering<br />
<em>Nucleic Acids Research</em> (2014-10-28) <a href="https://doi.org/f64rfn">https://doi.org/f64rfn</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gku1003">10.1093/nar/gku1003</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25352553">25352553</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4383874">PMC4383874</a></p>
</div>
<div id="ref-yGMDz6lK">
<p>44. <strong>PPInterFinder—a mining tool for extracting causal relations on human proteins from literature</strong><br />
Kalpana Raja, Suresh Subramani, Jeyakumar Natarajan<br />
<em>Database</em> (2013-01-01) <a href="https://doi.org/gf479b">https://doi.org/gf479b</a><br />
DOI: <a href="https://doi.org/10.1093/database/bas052">10.1093/database/bas052</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23325628">23325628</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3548331">PMC3548331</a></p>
</div>
<div id="ref-w32u0Rj9">
<p>45. <strong>HPIminer: A text mining system for building and visualizing human protein interaction networks and pathways</strong><br />
Suresh Subramani, Raja Kalpana, Pankaj Moses Monickaraj, Jeyakumar Natarajan<br />
<em>Journal of Biomedical Informatics</em> (2015-04) <a href="https://doi.org/f7bgnr">https://doi.org/f7bgnr</a><br />
DOI: <a href="https://doi.org/10.1016/j.jbi.2015.01.006">10.1016/j.jbi.2015.01.006</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25659452">25659452</a></p>
</div>
<div id="ref-8GVs1dBG">
<p>46. <strong>Analyzing a co-occurrence gene-interaction network to identify disease-gene association</strong><br />
Amira Al-Aamri, Kamal Taha, Yousof Al-Hammadi, Maher Maalouf, Dirar Homouz<br />
<em>BMC Bioinformatics</em> (2019-02-08) <a href="https://doi.org/gf49nm">https://doi.org/gf49nm</a><br />
DOI: <a href="https://doi.org/10.1186/s12859-019-2634-7">10.1186/s12859-019-2634-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30736752">30736752</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6368766">PMC6368766</a></p>
</div>
<div id="ref-YWh6tPj">
<p>47. <strong>Comparative experiments on learning information extractors for proteins and their interactions</strong><br />
Razvan Bunescu, Ruifang Ge, Rohit J. Kate, Edward M. Marcotte, Raymond J. Mooney, Arun K. Ramani, Yuk Wah Wong<br />
<em>Artificial Intelligence in Medicine</em> (2005-02) <a href="https://doi.org/dhztpn">https://doi.org/dhztpn</a><br />
DOI: <a href="https://doi.org/10.1016/j.artmed.2004.07.016">10.1016/j.artmed.2004.07.016</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/15811782">15811782</a></p>
</div>
<div id="ref-DWpAeBxB">
<p>48. <strong>BioInfer: a corpus for information extraction in the biomedical domain</strong><br />
Sampo Pyysalo, Filip Ginter, Juho Heimonen, Jari Björne, Jorma Boberg, Jouni Järvinen, Tapio Salakoski<br />
<em>BMC Bioinformatics</em> (2007-02-09) <a href="https://doi.org/b7bhhc">https://doi.org/b7bhhc</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-8-50">10.1186/1471-2105-8-50</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17291334">17291334</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1808065">PMC1808065</a></p>
</div>
<div id="ref-szMMEMdC">
<p>49. <strong>Learning language in logic - genic interaction extraction challenge</strong><br />
C. Nédellec<br />
<em>Proceedings of the learning language in logic 2005 workshop at the international conference on machine learning</em> (2005)</p>
</div>
<div id="ref-L9IIm3Zd">
<p>50. <strong>RelEx–Relation extraction using dependency parse trees</strong><br />
K. Fundel, R. Kuffner, R. Zimmer<br />
<em>Bioinformatics</em> (2006-12-01) <a href="https://doi.org/cz7q4d">https://doi.org/cz7q4d</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btl616">10.1093/bioinformatics/btl616</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17142812">17142812</a></p>
</div>
<div id="ref-115pgEuOr">
<p>51. <strong>Mining medline: Abstracts, sentences, or phrases?</strong><br />
Jing Ding, Daniel Berleant, Dan Nettleton, Eve Syrkin Wurtele<br />
<em>Pacific symposium on biocomputing</em> (2002) <a href="http://helix-web.stanford.edu/psb02/ding.pdf">http://helix-web.stanford.edu/psb02/ding.pdf</a></p>
</div>
<div id="ref-DR8XM4Ff">
<p>52. <strong>Comparative analysis of five protein-protein interaction corpora</strong><br />
Sampo Pyysalo, Antti Airola, Juho Heimonen, Jari Björne, Filip Ginter, Tapio Salakoski<br />
<em>BMC Bioinformatics</em> (2008-04) <a href="https://doi.org/fh3df7">https://doi.org/fh3df7</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-9-s3-s6">10.1186/1471-2105-9-s3-s6</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/18426551">18426551</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2349296">PMC2349296</a></p>
</div>
<div id="ref-iiQkIqUX">
<p>53. <strong>Exploiting graph kernels for high performance biomedical relation extraction</strong><br />
Nagesh C. Panyam, Karin Verspoor, Trevor Cohn, Kotagiri Ramamohanarao<br />
<em>Journal of Biomedical Semantics</em> (2018-01-30) <a href="https://doi.org/gf49nn">https://doi.org/gf49nn</a><br />
DOI: <a href="https://doi.org/10.1186/s13326-017-0168-3">10.1186/s13326-017-0168-3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29382397">29382397</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5791373">PMC5791373</a></p>
</div>
<div id="ref-1B0lnkj35">
<p>54. <strong>Text Mining for Protein Docking</strong><br />
Varsha D. Badal, Petras J. Kundrotas, Ilya A. Vakser<br />
<em>PLOS Computational Biology</em> (2015-12-09) <a href="https://doi.org/gcvj3b">https://doi.org/gcvj3b</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pcbi.1004630">10.1371/journal.pcbi.1004630</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26650466">26650466</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4674139">PMC4674139</a></p>
</div>
<div id="ref-ibJfUvEe">
<p>55. <strong>Feature assisted stacked attentive shortest dependency path based Bi-LSTM model for protein–protein interaction</strong><br />
Shweta Yadav, Asif Ekbal, Sriparna Saha, Ankit Kumar, Pushpak Bhattacharyya<br />
<em>Knowledge-Based Systems</em> (2019-02) <a href="https://doi.org/gf4788">https://doi.org/gf4788</a><br />
DOI: <a href="https://doi.org/10.1016/j.knosys.2018.11.020">10.1016/j.knosys.2018.11.020</a></p>
</div>
<div id="ref-1H4LpFrU0">
<p>56. <strong>Extraction of protein–protein interactions (PPIs) from the literature by deep convolutional neural networks with various feature embeddings</strong><br />
Sung-Pil Choi<br />
<em>Journal of Information Science</em> (2016-11-01) <a href="https://doi.org/gcv8bn">https://doi.org/gcv8bn</a><br />
DOI: <a href="https://doi.org/10.1177/0165551516673485">10.1177/0165551516673485</a></p>
</div>
<div id="ref-bLKJwjMD">
<p>57. <strong>Deep learning for extracting protein-protein interactions from biomedical literature</strong><br />
Yifan Peng, Zhiyong Lu<br />
<em>arXiv</em> (2017-06-05) <a href="https://arxiv.org/abs/1706.01556v2">https://arxiv.org/abs/1706.01556v2</a></p>
</div>
<div id="ref-WYud0jQT">
<p>58. <strong>Large-scale extraction of gene interactions from full-text literature using DeepDive</strong><br />
Emily K. Mallory, Ce Zhang, Christopher Ré, Russ B. Altman<br />
<em>Bioinformatics</em> (2015-09-03) <a href="https://doi.org/gb5g7b">https://doi.org/gb5g7b</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btv476">10.1093/bioinformatics/btv476</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26338771">26338771</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4681986">PMC4681986</a></p>
</div>
<div id="ref-16cIDAXhG">
<p>59. <strong>The new NHGRI-EBI Catalog of published genome-wide association studies (GWAS Catalog)</strong><br />
Jacqueline MacArthur, Emily Bowler, Maria Cerezo, Laurent Gil, Peggy Hall, Emma Hastings, Heather Junkins, Aoife McMahon, Annalisa Milano, Joannella Morales, … Helen Parkinson<br />
<em>Nucleic Acids Research</em> (2016-11-29) <a href="https://doi.org/f9v7cp">https://doi.org/f9v7cp</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkw1133">10.1093/nar/gkw1133</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27899670">27899670</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5210590">PMC5210590</a></p>
</div>
<div id="ref-1FI8iuYiQ">
<p>60. <strong>DrugBank 5.0: a major update to the DrugBank database for 2018</strong><br />
David S Wishart, Yannick D Feunang, An C Guo, Elvis J Lo, Ana Marcu, Jason R Grant, Tanvir Sajed, Daniel Johnson, Carin Li, Zinat Sayeeda, … Michael Wilson<br />
<em>Nucleic Acids Research</em> (2017-11-08) <a href="https://doi.org/gcwtzk">https://doi.org/gcwtzk</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkx1037">10.1093/nar/gkx1037</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29126136">29126136</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5753335">PMC5753335</a></p>
</div>
<div id="ref-13vw5RIy4">
<p>61. <strong>PubTator: a web-based text mining tool for assisting biocuration</strong><br />
Chih-Hsuan Wei, Hung-Yu Kao, Zhiyong Lu<br />
<em>Nucleic Acids Research</em> (2013-05-22) <a href="https://doi.org/f475th">https://doi.org/f475th</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkt441">10.1093/nar/gkt441</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23703206">23703206</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3692066">PMC3692066</a></p>
</div>
<div id="ref-vtuZ3Wx7">
<p>62. <strong>DNorm: disease name normalization with pairwise learning to rank</strong><br />
R. Leaman, R. Islamaj Dogan, Z. Lu<br />
<em>Bioinformatics</em> (2013-08-21) <a href="https://doi.org/f5gj9n">https://doi.org/f5gj9n</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btt474">10.1093/bioinformatics/btt474</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23969135">23969135</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3810844">PMC3810844</a></p>
</div>
<div id="ref-4S2HMNpa">
<p>63. <strong>GeneTUKit: a software for document-level gene normalization</strong><br />
M. Huang, J. Liu, X. Zhu<br />
<em>Bioinformatics</em> (2011-02-08) <a href="https://doi.org/dng2cb">https://doi.org/dng2cb</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btr042">10.1093/bioinformatics/btr042</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21303863">21303863</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3065680">PMC3065680</a></p>
</div>
<div id="ref-1AkC7QdyP">
<p>64. <strong>Cross-species gene normalization by species inference</strong><br />
Chih-Hsuan Wei, Hung-Yu Kao<br />
<em>BMC Bioinformatics</em> (2011-10-03) <a href="https://doi.org/dnmvds">https://doi.org/dnmvds</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-12-s8-s5">10.1186/1471-2105-12-s8-s5</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22151999">22151999</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3269940">PMC3269940</a></p>
</div>
<div id="ref-r501gnuM">
<p>65. <strong>Collaborative biocuration–text-mining development task for document prioritization for curation</strong><br />
T. C. Wiegers, A. P. Davis, C. J. Mattingly<br />
<em>Database</em> (2012-11-22) <a href="https://doi.org/gbb3zw">https://doi.org/gbb3zw</a><br />
DOI: <a href="https://doi.org/10.1093/database/bas037">10.1093/database/bas037</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23180769">23180769</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3504477">PMC3504477</a></p>
</div>
<div id="ref-RQkLuc5t">
<p>66. <strong>The Stanford CoreNLP Natural Language Processing Toolkit</strong><br />
Christopher Manning, Mihai Surdeanu, John Bauer, Jenny Finkel, Steven Bethard, David McClosky<br />
<em>Proceedings of 52nd Annual Meeting of the Association for Computational Linguistics: System Demonstrations</em> (2014) <a href="https://doi.org/gf3xhp">https://doi.org/gf3xhp</a><br />
DOI: <a href="https://doi.org/10.3115/v1/p14-5010">10.3115/v1/p14-5010</a></p>
</div>
<div id="ref-vzoBuh4l">
<p>67. <strong>Snorkel</strong><br />
Alexander Ratner, Stephen H. Bach, Henry Ehrenberg, Jason Fries, Sen Wu, Christopher Ré<br />
<em>Proceedings of the VLDB Endowment</em> (2017-11-01) <a href="https://doi.org/ch44">https://doi.org/ch44</a><br />
DOI: <a href="https://doi.org/10.14778/3157794.3157797">10.14778/3157794.3157797</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29770249">29770249</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5951191">PMC5951191</a></p>
</div>
<div id="ref-9Jo1af7Z">
<p>68. <strong>Snorkel MeTaL</strong><br />
Alex Ratner, Braden Hancock, Jared Dunnmon, Roger Goldman, Christopher Ré<br />
<em>Proceedings of the Second Workshop on Data Management for End-To-End Machine Learning - DEEM’18</em> (2018) <a href="https://doi.org/gf3xk7">https://doi.org/gf3xk7</a><br />
DOI: <a href="https://doi.org/10.1145/3209889.3209898">10.1145/3209889.3209898</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30931438">30931438</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6436830">PMC6436830</a></p>
</div>
<div id="ref-LCyCrr7W">
<p>69. <strong>A Proteome-Scale Map of the Human Interactome Network</strong><br />
Thomas Rolland, Murat Taşan, Benoit Charloteaux, Samuel J. Pevzner, Quan Zhong, Nidhi Sahni, Song Yi, Irma Lemmens, Celia Fontanillo, Roberto Mosca, … Marc Vidal<br />
<em>Cell</em> (2014-11) <a href="https://doi.org/f3mn6x">https://doi.org/f3mn6x</a><br />
DOI: <a href="https://doi.org/10.1016/j.cell.2014.10.050">10.1016/j.cell.2014.10.050</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25416956">25416956</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4266588">PMC4266588</a></p>
</div>
<div id="ref-gtV3bOpd">
<p>70. <strong>iRefIndex: A consolidated protein interaction database with provenance</strong><br />
Sabry Razick, George Magklaras, Ian M Donaldson<br />
<em>BMC Bioinformatics</em> (2008) <a href="https://doi.org/b99bjj">https://doi.org/b99bjj</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-9-405">10.1186/1471-2105-9-405</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/18823568">18823568</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2573892">PMC2573892</a></p>
</div>
<div id="ref-2jkcXYxN">
<p>71. <strong>Uncovering disease-disease relationships through the incomplete interactome</strong><br />
J. Menche, A. Sharma, M. Kitsak, S. D. Ghiassian, M. Vidal, J. Loscalzo, A.-L. Barabasi<br />
<em>Science</em> (2015-02-19) <a href="https://doi.org/f3mn6z">https://doi.org/f3mn6z</a><br />
DOI: <a href="https://doi.org/10.1126/science.1257601">10.1126/science.1257601</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25700523">25700523</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4435741">PMC4435741</a></p>
</div>
<div id="ref-QJ6hYH8N">
<p>72. <strong>On Calibration of Modern Neural Networks</strong><br />
Chuan Guo, Geoff Pleiss, Yu Sun, Kilian Q. Weinberger<br />
<em>arXiv</em> (2017-06-14) <a href="https://arxiv.org/abs/1706.04599v2">https://arxiv.org/abs/1706.04599v2</a></p>
</div>
<div id="ref-rLVjMJ5l">
<p>73. <strong>Accurate Uncertainties for Deep Learning Using Calibrated Regression</strong><br />
Volodymyr Kuleshov, Nathan Fenner, Stefano Ermon<br />
<em>arXiv</em> (2018-07-01) <a href="https://arxiv.org/abs/1807.00263v1">https://arxiv.org/abs/1807.00263v1</a></p>
</div>
<div id="ref-L2B5V7XC">
<p>74. <strong>Mining knowledge from MEDLINE articles and their indexed MeSH terms</strong><br />
Daniel Himmelstein, Alex Pankov<br />
<em>ThinkLab</em> (2015-05-10) <a href="https://doi.org/f3mqwp">https://doi.org/f3mqwp</a><br />
DOI: <a href="https://doi.org/10.15363/thinklab.d67">10.15363/thinklab.d67</a></p>
</div>
<div id="ref-4G0GW8oe">
<p>75. <strong>Integrating resources with disparate licensing into an open network</strong><br />
Daniel Himmelstein, Lars Juhl Jensen, MacKenzie Smith, Katie Fortney, Caty Chung<br />
<em>ThinkLab</em> (2015-08-28) <a href="https://doi.org/bfmk">https://doi.org/bfmk</a><br />
DOI: <a href="https://doi.org/10.15363/thinklab.d107">10.15363/thinklab.d107</a></p>
</div>
<div id="ref-137tbemL9">
<p>76. <strong>Legal confusion threatens to slow data science</strong><br />
Simon Oxenham<br />
<em>Nature</em> (2016-08) <a href="https://doi.org/bndt">https://doi.org/bndt</a><br />
DOI: <a href="https://doi.org/10.1038/536016a">10.1038/536016a</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27488781">27488781</a></p>
</div>
<div id="ref-1GwdMLPbV">
<p>77. <strong>An analysis and metric of reusable data licensing practices for biomedical resources</strong><br />
Seth Carbon, Robin Champieux, Julie A. McMurry, Lilly Winfree, Letisha R. Wyatt, Melissa A. Haendel<br />
<em>PLOS ONE</em> (2019-03-27) <a href="https://doi.org/gf5m8v">https://doi.org/gf5m8v</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0213090">10.1371/journal.pone.0213090</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30917137">30917137</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6436688">PMC6436688</a></p>
</div>
<div id="ref-fs8rAHoJ">
<p>78. <strong>A Sensitivity Analysis of (and Practitioners’ Guide to) Convolutional Neural Networks for Sentence Classification</strong><br />
Ye Zhang, Byron Wallace<br />
<em>arXiv</em> (2015-10-13) <a href="https://arxiv.org/abs/1510.03820v4">https://arxiv.org/abs/1510.03820v4</a></p>
</div>
<div id="ref-c6d3lKFX">
<p>79. <strong>Adam: A Method for Stochastic Optimization</strong><br />
Diederik P. Kingma, Jimmy Ba<br />
<em>arXiv</em> (2014-12-22) <a href="https://arxiv.org/abs/1412.6980v9">https://arxiv.org/abs/1412.6980v9</a></p>
</div>
<div id="ref-u5iJzbp9">
<p>80. <strong>Distributed Representations of Words and Phrases and their Compositionality</strong><br />
Tomas Mikolov, Ilya Sutskever, Kai Chen, Greg Corrado, Jeffrey Dean<br />
<em>arXiv</em> (2013-10-16) <a href="https://arxiv.org/abs/1310.4546v1">https://arxiv.org/abs/1310.4546v1</a></p>
</div>
<div id="ref-qUpCDz2v">
<p>81. <strong>Enriching Word Vectors with Subword Information</strong><br />
Piotr Bojanowski, Edouard Grave, Armand Joulin, Tomas Mikolov<br />
<em>arXiv</em> (2016-07-15) <a href="https://arxiv.org/abs/1607.04606v2">https://arxiv.org/abs/1607.04606v2</a></p>
</div>
<div id="ref-1GhHIDxuW">
<p>82. <strong>Efficient Estimation of Word Representations in Vector Space</strong><br />
Tomas Mikolov, Kai Chen, Greg Corrado, Jeffrey Dean<br />
<em>arXiv</em> (2013-01-16) <a href="https://arxiv.org/abs/1301.3781v3">https://arxiv.org/abs/1301.3781v3</a></p>
</div>
</div>
<h2 id="supplemental-methods" class="page_break_before">Supplemental Methods</h2>
<h3 id="adding-random-noise-to-generative-model">Adding Random Noise to Generative Model</h3>
<p>We discovered in the course of this work that adding a single label function from a mismatched type would often improve the performance of the generative model (see Results). We designed an experiment to test whether adding a noisy label function also increased performance. This label function emitted a positive or negative label at varying frequencies, which were evenly spaced from zero to one. Zero was the same as distant supervision and one meant that all sentences were randomly labeled. We trained the generative model with these label functions added and reported results in terms of AUROC and AUPR.</p>
<h3 id="discriminative-model">Discriminative Model</h3>
<p>The discriminative model is a neural network, which we train to predict labels from the generative model. The expectation is that the discriminative model can learn more complete features of the text than the label functions used in the generative model. We used a convolutional neural network with multiple filters as our discriminative model. This network uses multiple filters with fixed widths of 300 dimensions and a fixed height of 7 (Figure <a href="#fig:convolutional_network">5</a>), because this height provided the best performance in terms of relationship classification <span class="citation" data-cites="fs8rAHoJ">[<a href="#ref-fs8rAHoJ" role="doc-biblioref">78</a>]</span>. We trained this model for 20 epochs using the adam optimizer <span class="citation" data-cites="c6d3lKFX">[<a href="#ref-c6d3lKFX" role="doc-biblioref">79</a>]</span> with pytorch’s default parameter settings and a learning rate of 0.001. We added a L2 penalty on the network weights to prevent overfitting. Lastly, we added a dropout layer (p=0.25) between the fully connected layer and the softmax layer.</p>
<div id="fig:convolutional_network" class="fignos">
<figure>
<img src="images/figures/convolutional_neural_network/convolutional_neural_nework.png" alt="" /><figcaption><span>Figure 5:</span> The architecture of the discriminative model was a convolutional neural network. We performed a convolution step using multiple filters. The filters generated a feature map that was sent into a maximum pooling layer that was designed to extract the largest feature in each map. The extracted features were concatenated into a singular vector that was passed into a fully connected network. The fully connected network had 300 neurons for the first layer, 100 neurons for the second layer and 50 neurons for the last layer. The last step from the fully connected network was to generate predictions using a softmax layer.</figcaption>
</figure>
</div>
<h4 id="word-embeddings">Word Embeddings</h4>
<p>Word embeddings are representations that map individual words to real valued vectors of user-specified dimensions. These embeddings have been shown to capture the semantic and syntactic information between words <span class="citation" data-cites="u5iJzbp9">[<a href="#ref-u5iJzbp9" role="doc-biblioref">80</a>]</span>. We trained Facebook’s fastText <span class="citation" data-cites="qUpCDz2v">[<a href="#ref-qUpCDz2v" role="doc-biblioref">81</a>]</span> using all candidate sentences for each individual relationship pair to generate word embeddings. fastText uses a skipgram model <span class="citation" data-cites="1GhHIDxuW">[<a href="#ref-1GhHIDxuW" role="doc-biblioref">82</a>]</span> that aims to predict the surrounding context for a candidate word and pairs the model with a novel scoring function that treats each word as a bag of character n-grams. We trained this model for 20 epochs using a window size of 2 and generated 300-dimensional word embeddings. We use the optimized word embeddings to train a discriminative model.</p>
<h4 id="calibration-of-the-discriminative-model">Calibration of the Discriminative Model</h4>
<p>Often many tasks require a machine learning model to output reliable probability predictions. A model is well calibrated if the probabilities emitted from the model match the observed probabilities: a well-calibrated model that assigns a class label with 80% probability should have that class appear 80% of the time. Deep neural network models can often be poorly calibrated <span class="citation" data-cites="QJ6hYH8N rLVjMJ5l">[<a href="#ref-QJ6hYH8N" role="doc-biblioref">72</a>,<a href="#ref-rLVjMJ5l" role="doc-biblioref">73</a>]</span>. These models are usually over-confident in their predictions. As a result, we calibrated our convolutional neural network using temperature scaling. Temperature scaling uses a parameter T to scale each value of the logit vector (z) before being passed into the softmax (SM) function.</p>
<p><span class="math display">\[\sigma_{SM}(\frac{z_{i}}{T}) = \frac{\exp(\frac{z_{i}}{T})}{\sum_{i}\exp(\frac{z_{i}}{T})}\]</span></p>
<p>We found the optimal T by minimizing the negative log likelihood (NLL) of a held out validation set. The benefit of using this method is that the model becomes more reliable and the accuracy of the model doesn’t change <span class="citation" data-cites="QJ6hYH8N">[<a href="#ref-QJ6hYH8N" role="doc-biblioref">72</a>]</span>.</p>
<h2 id="supplemental-tables-and-figures">Supplemental Tables and Figures</h2>
<h3 id="generative-model-aupr-performance">Generative Model AUPR Performance</h3>
<div id="fig:aupr_gen_model_performance" class="fignos">
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/ee638b4e45717a86f54a2744a813baaa90bc6b84/figures/label_sampling_experiment/transfer_test_set_auprc.png" alt="" /><figcaption><span>Figure 6:</span> Grid of AUPR scores for each generative model trained on randomly sampled label functions. The rows depict the relationship each model is trying to predict and the columns are the edge type specific sources from which each label function is sampled. For example, the top-left most square depicts the generative model predicting DaG sentences, while randomly sampling label functions designed to predict the DaG relationship. The square towards the right depicts the generative model predicting DaG sentences, while randomly sampling label functions designed to predict the CtD relationship. This pattern continues filling out the rest of the grid. The right most column consists of pooling every relationship specific label function and proceeding as above.</figcaption>
</figure>
</div>
<h3 id="random-label-function-generative-model-analysis-1">Random Label Function Generative Model Analysis</h3>
<div id="fig:aupr_random_label_function_performance" class="fignos">
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/ee638b4e45717a86f54a2744a813baaa90bc6b84/figures/gen_model_error_analysis/transfer_test_set_auprc.png" alt="" /><figcaption><span>Figure 7:</span> A grid of AUROC (A) scores for each edge type. Each plot consists of adding a single label function on top of the baseline model. This label function emits a positive (shown in blue) or negative (shown in orange) label at specified frequencies, and performance at zero is equivalent to not having a randomly emitting label function. The error bars represent 95% confidence intervals for AUROC or AUPR (y-axis) at each emission frequency.</figcaption>
</figure>
</div>
<h3 id="discriminative-model-performance">Discriminative Model Performance</h3>
<div id="fig:auroc_discriminative_model_performance" class="fignos">
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/ee638b4e45717a86f54a2744a813baaa90bc6b84/figures/label_sampling_experiment/disc_performance_test_set_auroc.png" alt="" /><figcaption><span>Figure 8:</span> Grid of AUROC scores for each discriminative model trained using generated labels from the generative models. The rows depict the edge type each model is trying to predict and the columns are the edge type specific sources from which each label function was sampled. For example, the top-left most square depicts the discriminator model predicting DaG sentences, while randomly sampling label functions designed to predict the DaG relationship. The error bars over the points represents the standard deviation between sampled runs. The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the CtD relationship. This pattern continues filling out the rest of the grid. The right most column consists of pooling every relationship specific label function and proceeding as above.</figcaption>
</figure>
</div>
<p>In this framework we used a generative model trained over label functions to produce probabilistic training labels for each sentence. Then we trained a discriminative model, which has full access to a representation of the text of the sentence, to predict the generated labels. The discriminative model is a convolutional neural network trained over word embeddings (See Methods). We report the results of the discriminative model using AUROC and AUPR (Figures <a href="#fig:auroc_discriminative_model_performance">8</a> and <a href="#fig:aupr_discriminative_model_performance">9</a>).</p>
<p>We found that the discriminative model under-performed the generative model in most cases. Only for the CtD edge does the discriminative model appear to provide performance above the generative model and that increased performance is only with a modest number of label functions. With the full set of label functions, performance of both models remain similar. The one or a few mismatched label functions (off-diagonal) improving generative model performance trend is retained despite the limited performance of the discriminative model.</p>
<div id="fig:aupr_discriminative_model_performance" class="fignos">
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/ee638b4e45717a86f54a2744a813baaa90bc6b84/figures/label_sampling_experiment/disc_performance_test_set_auprc.png" alt="" /><figcaption><span>Figure 9:</span> Grid of AUPR scores for each discriminative model trained using generated labels from the generative models. The rows depict the edge type each model is trying to predict and the columns are the edge type specific sources from which each label function was sampled. For example, the top-left most square depicts the discriminator model predicting DaG sentences, while randomly sampling label functions designed to predict the DaG relationship. The error bars over the points represents the standard deviation between sampled runs. The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the CtD relationship. This pattern continues filling out the rest of the grid. The right most column consists of pooling every relationship specific label function and proceeding as above.</figcaption>
</figure>
</div>
<h4 id="model-calibration-tables">Model Calibration Tables</h4>
<div id="tbl:dg_top_ten_table" class="tablenos">
<table>
<caption><span>Table 3:</span> Contains the top ten Disease-associates-Gene confidence scores before and after model calbration. Disease mentions are highlighted in <span class="disease_color">brown</span> and Gene mentions are highlighted in <span class="gene_color">blue</span>. </caption>
<colgroup>
<col style="width: 18%" />
<col style="width: 9%" />
<col style="width: 42%" />
<col style="width: 15%" />
<col style="width: 14%" />
</colgroup>
<thead>
<tr class="header">
<th>Disease Name</th>
<th>Gene Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>prostate cancer</td>
<td>DKK1</td>
<td>conclusion : high <span class="gene_color">dkk-1</span> serum levels are associated with a poor survival in patients with <span class="disease_color">prostate cancer</span> .</td>
<td>0.999</td>
<td>0.916</td>
</tr>
<tr class="even">
<td>breast cancer</td>
<td>ERBB2</td>
<td>conclusion : <span class="gene_color">her-2 / neu</span> overexpression in primary <span class="disease_color">breast carcinoma</span> is correlated with patients ’ age ( under age 50 ) and calcifications at mammography .</td>
<td>0.998</td>
<td>0.906</td>
</tr>
<tr class="odd">
<td>breast cancer</td>
<td>ERBB2</td>
<td>the results of multiple linear regression analysis , with her2 as the dependent variable , showed that family history of <span class="disease_color">breast cancer</span> was significantly associated with elevated <span class="gene_color">her2</span> levels in the tumors ( p = 0.0038 ) , after controlling for the effects of age , tumor estrogen receptor , and dna index .</td>
<td>0.998</td>
<td>0.904</td>
</tr>
<tr class="even">
<td>colon cancer</td>
<td>SP3</td>
<td>ba also decreased expression of sp1 , <span class="gene_color">sp3</span> and sp4 transcription factors which are overexpressed in <span class="disease_color">colon cancer</span> cells and decreased levels of several sp-regulated genes including survivin , vascular endothelial growth factor , p65 sub-unit of nfkb , epidermal growth factor receptor , cyclin d1 , and pituitary tumor transforming gene-1 .</td>
<td>0.998</td>
<td>0.902</td>
</tr>
<tr class="odd">
<td>breast cancer</td>
<td>ERBB2</td>
<td>in <span class="disease_color">breast cancer</span> , overexpression of <span class="gene_color">her2</span> is associated with an aggressive tumor phenotype and poor prognosis .</td>
<td>0.998</td>
<td>0.898</td>
</tr>
<tr class="even">
<td>breast cancer</td>
<td>BCL2</td>
<td>in clinical <span class="disease_color">breast cancer</span> samples , high <span class="gene_color">bcl2</span> expression was associated with poor prognosis .</td>
<td>0.997</td>
<td>0.886</td>
</tr>
<tr class="odd">
<td>adrenal gland cancer</td>
<td>TP53</td>
<td>the mechanisms of adrenal tumorigenesis remain poorly established ; the r337h germline mutation in the <span class="gene_color">p53</span> gene has previously been associated with <span class="disease_color">acts</span> in brazilian children .</td>
<td>0.996</td>
<td>0.883</td>
</tr>
<tr class="even">
<td>prostate cancer</td>
<td>AR</td>
<td>the <span class="gene_color">androgen receptor</span> was expressed in all primary and metastatic <span class="disease_color">prostate cancer</span> tissues and no mutations were identified .</td>
<td>0.996</td>
<td>0.881</td>
</tr>
<tr class="odd">
<td>urinary bladder cancer</td>
<td>PIK3CA</td>
<td>conclusions : increased levels of fgfr3 and <span class="gene_color">pik3ca</span> mutated dna in urine and plasma are indicative of later progression and metastasis in <span class="disease_color">bladder cancer</span> .</td>
<td>0.995</td>
<td>0.866</td>
</tr>
<tr class="even">
<td>ovarian cancer</td>
<td>EPAS1</td>
<td>the log-rank test showed that nuclear positive immunostaining for hif-1alpha ( p = .002 ) and cytoplasmic positive immunostaining for <span class="gene_color">hif-2alpha</span> ( p = .0112 ) in tumor cells are associated with poor prognosis of patients with <span class="disease_color">ovarian carcinoma</span> .</td>
<td>0.994</td>
<td>0.86</td>
</tr>
</tbody>
</table>
</div>
<div id="tbl:dg_bottom_ten_table" class="tablenos">
<table>
<caption><span>Table 4:</span> Contains the bottom ten Disease-associates-Gene confidence scores before and after model calbration. Disease mentions are highlighted in <span class="disease_color">brown</span> and Gene mentions are highlighted in <span class="gene_color">blue</span>. </caption>
<colgroup>
<col style="width: 27%" />
<col style="width: 11%" />
<col style="width: 28%" />
<col style="width: 16%" />
<col style="width: 16%" />
</colgroup>
<thead>
<tr class="header">
<th>Disease Name</th>
<th>Gene Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>endogenous depression</td>
<td>EP300</td>
<td>from a clinical point of view , <span class="gene_color">p300</span> amplitude should be considered as a psychophysiological index of suicidal risk in major <span class="disease_color">depressive disorder</span> .</td>
<td>0.202</td>
<td>0.379</td>
</tr>
<tr class="even">
<td>Alzheimer’s disease</td>
<td>PDK1</td>
<td><span class="gene_color">from prion diseases to <span class="disease_color">alzheimer ’s disease</span> : a common therapeutic target , [pdk1 ]</span> .</td>
<td>0.2</td>
<td>0.378</td>
</tr>
<tr class="odd">
<td>endogenous depression</td>
<td>HTR1A</td>
<td>gepirone , a selective serotonin ( <span class="gene_color">5ht1a )</span> partial agonist in the treatment of <span class="disease_color">major depression</span> .</td>
<td>0.199</td>
<td>0.378</td>
</tr>
<tr class="even">
<td>Gilles de la Tourette syndrome</td>
<td>FGF9</td>
<td>there were no differences in gender distribution , age at tic onset or <span class="disease_color">td</span> diagnosis , tic severity , proportion with current diagnoses of ocd/oc behavior or attention deficit hyperactivity disorder ( adhd ) , cbcl internalizing , externalizing , or total problems scores , ygtss scores , or <span class="gene_color">gaf</span> scores .</td>
<td>0.185</td>
<td>0.37</td>
</tr>
<tr class="odd">
<td>hematologic cancer</td>
<td>MLANA</td>
<td>methods : the sln sections ( n = 214 ) were assessed by qrt assay for 4 established messenger rna biomarkers : <span class="gene_color">mart-1</span> , mage-a3 , <span class="disease_color">galnac-t</span> , and pax3 .</td>
<td>0.18</td>
<td>0.368</td>
</tr>
<tr class="even">
<td>endogenous depression</td>
<td>MAOA</td>
<td>alpha 2-adrenoceptor responsivity in <span class="disease_color">depression</span> : effect of chronic treatment with moclobemide , a selective <span class="gene_color">mao-a-inhibitor</span> , versus maprotiline .</td>
<td>0.179</td>
<td>0.367</td>
</tr>
<tr class="odd">
<td>chronic kidney failure</td>
<td>B2M</td>
<td>to evaluate comparative <span class="gene_color">beta 2-m</span> removal we studied six stable <span class="disease_color">end-stage renal failure</span> patients during high-flux 3-h haemodialysis , haemodia-filtration , and haemofiltration , using acrylonitrile , cellulose triacetate , polyamide and polysulphone capillary devices .</td>
<td>0.178</td>
<td>0.366</td>
</tr>
<tr class="even">
<td>hematologic cancer</td>
<td>C7</td>
<td>serum antibody responses to four haemophilus influenzae type b capsular polysaccharide-protein conjugate vaccines ( prp-d , hboc , <span class="gene_color">c7p</span> , and <span class="disease_color">prp-t )</span> were studied and compared in 175 infants , 85 adults and 140 2-year-old children .</td>
<td>0.174</td>
<td>0.364</td>
</tr>
<tr class="odd">
<td>hypertension</td>
<td>AVP</td>
<td>portohepatic pressures , hepatic function , and blood gases in the combination of nitroglycerin and <span class="gene_color">vasopressin</span> : search for additive effects in <span class="disease_color">cirrhotic portal hypertension</span> .</td>
<td>0.168</td>
<td>0.361</td>
</tr>
<tr class="even">
<td>endogenous depression</td>
<td>GAD1</td>
<td>within-individual deflections in gad , physical , and social symptoms predicted later deflections in <span class="disease_color">depressive symptoms</span> , and deflections in depressive symptoms predicted later deflections in <span class="gene_color">gad</span> and separation anxiety symptoms .</td>
<td>0.149</td>
<td>0.349</td>
</tr>
</tbody>
</table>
</div>
<div id="tbl:cd_top_ten_table" class="tablenos">
<table>
<caption><span>Table 5:</span> Contains the top ten Compound-treats-Disease confidence scores after model calbration. Disease mentions are highlighted in <span class="disease_color">brown</span> and Compound mentions are highlighted in <span class="compound_color">red</span>. </caption>
<colgroup>
<col style="width: 14%" />
<col style="width: 14%" />
<col style="width: 41%" />
<col style="width: 14%" />
<col style="width: 13%" />
</colgroup>
<thead>
<tr class="header">
<th>Compound Name</th>
<th>Disease Name</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Prazosin</td>
<td>hypertension</td>
<td>experience with <span class="compound_color">prazosin</span> in the treatment of <span class="disease_color">hypertension</span> .</td>
<td>0.997</td>
<td>0.961</td>
</tr>
<tr class="even">
<td>Methyldopa</td>
<td>hypertension</td>
<td>oxprenolol plus cyclopenthiazide-kcl versus <span class="compound_color">methyldopa</span> in the treatment of <span class="disease_color">hypertension</span> .</td>
<td>0.997</td>
<td>0.961</td>
</tr>
<tr class="odd">
<td>Methyldopa</td>
<td>hypertension</td>
<td>atenolol and <span class="compound_color">methyldopa</span> in the treatment of <span class="disease_color">hypertension</span> .</td>
<td>0.996</td>
<td>0.957</td>
</tr>
<tr class="even">
<td>Prednisone</td>
<td>asthma</td>
<td><span class="compound_color">prednisone</span> and beclomethasone for treatment of <span class="disease_color">asthma</span> .</td>
<td>0.995</td>
<td>0.953</td>
</tr>
<tr class="odd">
<td>Sulfasalazine</td>
<td>ulcerative colitis</td>
<td><span class="compound_color">sulphasalazine</span> , used in the treatment of <span class="disease_color">ulcerative colitis</span> , is cleaved in the colon by the metabolic action of colonic bacteria on the diazo bond to release 5-aminosalicylic acid ( 5-asa ) and sulpharidine .</td>
<td>0.994</td>
<td>0.949</td>
</tr>
<tr class="even">
<td>Prazosin</td>
<td>hypertension</td>
<td>letter : <span class="compound_color">prazosin</span> in treatment of <span class="disease_color">hypertension</span> .</td>
<td>0.994</td>
<td>0.949</td>
</tr>
<tr class="odd">
<td>Methylprednisolone</td>
<td>asthma</td>
<td>use of tao without <span class="compound_color">methylprednisolone</span> in the treatment of severe <span class="disease_color">asthma</span> .</td>
<td>0.994</td>
<td>0.948</td>
</tr>
<tr class="even">
<td>Budesonide</td>
<td>asthma</td>
<td>thus , a regimen of <span class="compound_color">budesonide</span> treatment that consistently attenuates bronchial responsiveness in <span class="disease_color">asthmatic</span> subjects had no effect in these men ; larger and longer trials will be required to establish whether a subgroup of smokers shows a favorable response .</td>
<td>0.994</td>
<td>0.946</td>
</tr>
<tr class="odd">
<td>Methyldopa</td>
<td>hypertension</td>
<td>pressor and chronotropic responses to bilateral carotid occlusion ( bco ) and tyramine were also markedly reduced following treatment with <span class="compound_color">methyldopa</span> , which is consistent with the clinical findings that chronic methyldopa treatment in <span class="disease_color">hypertensive</span> patients impairs cardiovascular reflexes .</td>
<td>0.994</td>
<td>0.946</td>
</tr>
<tr class="even">
<td>Fluphenazine</td>
<td>schizophrenia</td>
<td>low dose <span class="compound_color">fluphenazine decanoate</span> in maintenance treatment of <span class="disease_color">schizophrenia</span> .</td>
<td>0.994</td>
<td>0.946</td>
</tr>
</tbody>
</table>
</div>
<div id="tbl:cd_bottom_ten_table" class="tablenos">
<table>
<caption><span>Table 6:</span> Contains the bottom ten Compound-treats-Disease confidence scores before and after model calbration. Disease mentions are highlighted in <span class="disease_color">brown</span> and Compound mentions are highlighted in <span class="compound_color">red</span>. </caption>
<colgroup>
<col style="width: 12%" />
<col style="width: 21%" />
<col style="width: 35%" />
<col style="width: 15%" />
<col style="width: 14%" />
</colgroup>
<thead>
<tr class="header">
<th>Compound Name</th>
<th>Disease Name</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Indomethacin</td>
<td>hypertension</td>
<td>effects of <span class="compound_color">indomethacin</span> in rabbit <span class="disease_color">renovascular hypertension</span> .</td>
<td>0.033</td>
<td>0.13</td>
</tr>
<tr class="even">
<td>Alprazolam</td>
<td>panic disorder</td>
<td>according to logistic regression analysis , the relationships between plasma <span class="compound_color">alprazolam</span> concentration and response , as reflected by number of <span class="disease_color">panic attacks</span> reported , phobia ratings , physicians ’ and patients ’ ratings of global improvement , and the emergence of side effects , were significant .</td>
<td>0.03</td>
<td>0.124</td>
</tr>
<tr class="odd">
<td>Mestranol</td>
<td>polycystic ovary syndrome</td>
<td>the binding capacity of plasma testosterone-estradiol-binding globulin ( tebg ) and testosterone ( t ) levels were measured in four women with proved <span class="disease_color">polycystic ovaries</span> and three women with a clinical diagnosis of polycystic ovarian disease before , during , and after administration of norethindrone , 2 mg. , and <span class="compound_color">mestranol</span> , 0.1 mg .</td>
<td>0.03</td>
<td>0.123</td>
</tr>
<tr class="even">
<td>Creatine</td>
<td>coronary artery disease</td>
<td>during successful and uncomplicated angioplasty ( ptca ) , we studied the effect of a short lasting <span class="disease_color">myocardial ischemia</span> on plasma creatine kinase , creatine kinase mb-activity , and <span class="compound_color">creatine</span> kinase mm-isoforms ( mm1 , mm2 , mm3 ) in 23 patients .</td>
<td>0.028</td>
<td>0.12</td>
</tr>
<tr class="odd">
<td>Creatine</td>
<td>coronary artery disease</td>
<td>in 141 patients with <span class="disease_color">acute myocardial infarction</span> , <span class="compound_color">creatine</span> phosphokinase isoenzyme ( cpk-mb ) was determined by the activation method with dithiothreitol ( rao et al. : clin .</td>
<td>0.027</td>
<td>0.117</td>
</tr>
<tr class="even">
<td>Morphine</td>
<td>brain cancer</td>
<td>the tissue to serum ratio of <span class="compound_color">morphine</span> in the <span class="disease_color">hypothalamus</span> , hippocampus , striatum , midbrain and cortex were also smaller in morphine tolerant than in non-tolerant rats .</td>
<td>0.026</td>
<td>0.115</td>
</tr>
<tr class="odd">
<td>Glutathione</td>
<td>anemia</td>
<td>our results suggest that an association between <span class="compound_color">gsh</span> px <span class="disease_color">deficiency and hemolytic anemia</span> need not represent a cause-and-effect relationship .</td>
<td>0.026</td>
<td>0.114</td>
</tr>
<tr class="even">
<td>Dinoprostone</td>
<td>stomach cancer</td>
<td>prostaglandin e2 ( <span class="compound_color">pge2 )</span> - and 6-keto-pgf1 alpha-like immunoactivity was measured in incubates of <span class="disease_color">forestomach and gastric corpus mucosa</span> in ( a ) unoperated rats , ( b ) rats with sham-operation of the kidneys and ( c ) rats with bilateral nephrectomy .</td>
<td>0.023</td>
<td>0.107</td>
</tr>
<tr class="odd">
<td>Creatine</td>
<td>coronary artery disease</td>
<td>the value of the electrocardiogram in assessing infarct size was studied using serial estimates of the mb isomer of <span class="compound_color">creatine</span> kinase ( ck mb ) in plasma , serial 35 lead praecordial maps in 28 patients with <span class="disease_color">anterior myocardial infarction</span> , and serial 12 lead electrocardiograms in 17 patients with inferior myocardial infarction .</td>
<td>0.022</td>
<td>0.105</td>
</tr>
<tr class="even">
<td>Sulfamethazine</td>
<td>multiple sclerosis</td>
<td>quantitation and confirmation of <span class="compound_color">sulfamethazine</span> residues in swine muscle and liver by lc and <span class="disease_color">gc/ms</span> .</td>
<td>0.017</td>
<td>0.093</td>
</tr>
</tbody>
</table>
</div>
<div id="tbl:cg_top_ten_table" class="tablenos">
<table>
<caption><span>Table 7:</span> Contains the top ten Compound-binds-Gene confidence scores before and after model calbration. Gene mentions are highlighted in <span class="gene_color">blue</span> and Compound mentions are highlighted in <span class="compound_color">red</span>. </caption>
<colgroup>
<col style="width: 29%" />
<col style="width: 11%" />
<col style="width: 22%" />
<col style="width: 18%" />
<col style="width: 17%" />
</colgroup>
<thead>
<tr class="header">
<th>Compound Name</th>
<th>Gene Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Cyclic Adenosine Monophosphate</td>
<td>B3GNT2</td>
<td>in sk-n-mc human neuroblastoma cells , the <span class="compound_color">camp</span> response to 10 nm isoproterenol ( iso ) is mediated primarily by <span class="gene_color">beta 1-adrenergic</span> receptors .</td>
<td>0.903</td>
<td>0.93</td>
</tr>
<tr class="even">
<td>Indomethacin</td>
<td>AGT</td>
<td><span class="compound_color">indomethacin</span> , a potent inhibitor of prostaglandin synthesis , is known to increase the maternal blood pressure response to <span class="gene_color">angiotensin ii</span> infusion .</td>
<td>0.894</td>
<td>0.922</td>
</tr>
<tr class="odd">
<td>Tretinoin</td>
<td>RXRA</td>
<td>the vitamin a derivative <span class="compound_color">retinoic acid</span> exerts its effects on transcription through two distinct classes of nuclear receptors , the retinoic acid receptor ( rar ) and the <span class="gene_color">retinoid x receptor</span> ( rxr ) .</td>
<td>0.882</td>
<td>0.912</td>
</tr>
<tr class="even">
<td>Tretinoin</td>
<td>RXRA</td>
<td>the vitamin a derivative retinoic acid exerts its effects on transcription through two distinct classes of nuclear receptors , the <span class="compound_color">retinoic acid</span> receptor ( rar ) and the <span class="gene_color">retinoid x receptor</span> ( rxr ) .</td>
<td>0.872</td>
<td>0.903</td>
</tr>
<tr class="odd">
<td>D-Tyrosine</td>
<td>CSF1</td>
<td>however , the extent of gap <span class="compound_color">tyrosine</span> phosphorylation induced by <span class="gene_color">csf-1</span> was approximately 10 % of that induced by pdgf-bb in the nih3t3 fibroblasts .</td>
<td>0.851</td>
<td>0.883</td>
</tr>
<tr class="even">
<td>D-Glutamic Acid</td>
<td>GLB1</td>
<td>thus , the negatively charged side chain of <span class="compound_color">glu-461</span> is important for divalent cation binding to <span class="gene_color">beta-galactosidase</span> .</td>
<td>0.849</td>
<td>0.882</td>
</tr>
<tr class="odd">
<td>D-Tyrosine</td>
<td>CD4</td>
<td>second , we use the same system to provide evidence that the physical association of <span class="gene_color">cd4</span> with the tcr is required for effective <span class="compound_color">tyrosine</span> phosphorylation of the tcr zeta-chain subunit , presumably reflecting delivery of p56lck ( lck ) to the tcr .</td>
<td>0.825</td>
<td>0.859</td>
</tr>
<tr class="even">
<td>Calcium Chloride</td>
<td>TNC</td>
<td>the possibility that the enhanced length dependence of <span class="compound_color">ca2</span> + sensitivity after cardiac tnc reconstitution was attributable to reduced <span class="gene_color">tnc</span> binding was excluded when the length dependence of partially extracted fast fibres was reduced to one-half the normal value after a 50 % deletion of the native tnc .</td>
<td>0.821</td>
<td>0.855</td>
</tr>
<tr class="odd">
<td>Metoprolol</td>
<td>KCNMB2</td>
<td>studies in difi cells of the displacement of specific 125i-cyp binding by nonselective ( propranolol ) , beta 1-selective ( <span class="compound_color">metoprolol</span> and atenolol ) , and beta 2-selective ( ici 118-551 ) antagonists revealed only a single class of <span class="gene_color">beta 2-adrenergic</span> receptors .</td>
<td>0.82</td>
<td>0.854</td>
</tr>
<tr class="even">
<td>D-Tyrosine</td>
<td>PLCG1</td>
<td>epidermal growth factor ( egf ) or platelet-derived growth factor binding to their receptor on fibroblasts induces tyrosine phosphorylation of plc gamma 1 and stable association of <span class="gene_color">plc gamma 1</span> with the receptor protein <span class="compound_color">tyrosine</span> kinase .</td>
<td>0.818</td>
<td>0.851</td>
</tr>
</tbody>
</table>
</div>
<div id="tbl:cg_bottom_ten_table" class="tablenos">
<table>
<caption><span>Table 8:</span> Contains the bottom ten Compound-binds-Gene confidence scores before and after model calbration. Gene mentions are highlighted in <span class="gene_color">blue</span> and Compound mentions are highlighted in <span class="compound_color">red</span>. </caption>
<colgroup>
<col style="width: 13%" />
<col style="width: 9%" />
<col style="width: 48%" />
<col style="width: 14%" />
<col style="width: 13%" />
</colgroup>
<thead>
<tr class="header">
<th>Compound Name</th>
<th>Gene Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Deferoxamine</td>
<td>TF</td>
<td>the mechanisms of fe uptake have been characterised using 59fe complexes of citrate , nitrilotriacetate , <span class="compound_color">desferrioxamine</span> , and 59fe added to eagle ’s minimum essential medium ( mem ) and compared with human transferrin ( <span class="gene_color">tf )</span> labelled with 59fe and iodine-125 .</td>
<td>0.02</td>
<td>0.011</td>
</tr>
<tr class="even">
<td>Hydrocortisone</td>
<td>GH1</td>
<td>group iv patients had normal basal levels of lh and normal lh , <span class="gene_color">gh</span> and <span class="compound_color">cortisol</span> responses .</td>
<td>0.02</td>
<td>0.011</td>
</tr>
<tr class="odd">
<td>Carbachol</td>
<td>INS</td>
<td>at the same concentration , however , iapp significantly ( p less than 0.05 ) inhibited <span class="compound_color">carbachol-stimulated</span> ( 10 ( -7 ) m ) release of insulin by 30 % , and cgrp significantly inhibited carbachol-stimulated release of <span class="gene_color">insulin</span> by 33 % when compared with the control group .</td>
<td>0.02</td>
<td>0.011</td>
</tr>
<tr class="even">
<td>Adenosine</td>
<td>ME2</td>
<td>at physiological concentrations , atp , adp , and <span class="compound_color">amp</span> all inhibit the enzyme from atriplex spongiosa and panicum miliaceum ( <span class="gene_color">nad-me-type</span> plants ) , with atp the most inhibitory species .</td>
<td>0.019</td>
<td>0.01</td>
</tr>
<tr class="odd">
<td>Naloxone</td>
<td>POMC</td>
<td>specifically , opioids , including 2-n-pentyloxy-2-phenyl-4-methyl-morpholine , <span class="compound_color">naloxone</span> , and <span class="gene_color">beta-endorphin</span> , have been shown to interact with il-2 receptors ( 134 ) and regulate production of il-1 and il-2 ( 48-50 , 135 ) .</td>
<td>0.018</td>
<td>0.01</td>
</tr>
<tr class="even">
<td>Cortisone acetate</td>
<td>POMC</td>
<td>sarcoidosis therapy with <span class="compound_color">cortisone</span> and <span class="gene_color">acth –</span> the role of acth therapy .</td>
<td>0.017</td>
<td>0.009</td>
</tr>
<tr class="odd">
<td>Epinephrine</td>
<td>INS</td>
<td>thermogenic effect of thyroid hormones : interactions with <span class="compound_color">epinephrine</span> and <span class="gene_color">insulin</span> .</td>
<td>0.017</td>
<td>0.009</td>
</tr>
<tr class="even">
<td>Aldosterone</td>
<td>KNG1</td>
<td>important vasoconstrictor , fluid - and sodium-retaining factors are the <span class="compound_color">renin-angiotensin-aldosterone</span> system , sympathetic nerve activity , and vasopressin ; vasodilator , volume , and sodium-eliminating factors are atrial natriuretic peptide , vasodilator prostaglandins like prostacyclin and prostaglandin e2 , dopamine , <span class="gene_color">bradykinin</span> , and possibly , endothelial derived relaxing factor ( edrf ) .</td>
<td>0.016</td>
<td>0.008</td>
</tr>
<tr class="odd">
<td>D-Leucine</td>
<td>POMC</td>
<td>cross-reactivities of <span class="compound_color">leucine-enkephalin</span> and <span class="gene_color">beta-endorphin</span> with the eia were less than 0.1 % , while that with gly-gly-phe-met and oxidized gly-gly-phe-met were 2.5 % and 10.2 % , respectively .</td>
<td>0.011</td>
<td>0.005</td>
</tr>
<tr class="even">
<td>Estriol</td>
<td>LGALS1</td>
<td>[ diagnostic value of serial determination of <span class="compound_color">estriol</span> and <span class="gene_color">hpl</span> in plasma and of total estrogens in 24-h-urine compared to single values for diagnosis of fetal danger ] .</td>
<td>0.01</td>
<td>0.005</td>
</tr>
</tbody>
</table>
</div>
<div id="tbl:gg_top_ten_table" class="tablenos">
<table>
<caption><span>Table 9:</span> Contains the top ten Gene-interacts-Gene confidence scores before and after model calbration. Both gene mentions highlighted in <span class="gene_color">blue</span>. </caption>
<colgroup>
<col style="width: 18%" />
<col style="width: 18%" />
<col style="width: 12%" />
<col style="width: 25%" />
<col style="width: 24%" />
</colgroup>
<thead>
<tr class="header">
<th>Gene1 Symbol</th>
<th>Gene2 Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ESR1</td>
<td>HSP90AA1</td>
<td>previous studies have suggested that the 90-kda heat shock protein ( <span class="gene_color">hsp90 )</span> interacts with the <span class="gene_color">er</span> , thus stabilizing the receptor in an inactive state .</td>
<td>0.812</td>
<td>0.864</td>
</tr>
<tr class="even">
<td>TP53</td>
<td>TP73</td>
<td>cyclin g interacts with p53 as well as <span class="gene_color">p73</span> , and its binding to <span class="gene_color">p53</span> or p73 presumably mediates downregulation of p53 and p73 .</td>
<td>0.785</td>
<td>0.837</td>
</tr>
<tr class="odd">
<td>TP53</td>
<td>AKT1</td>
<td>treatment of c81 cells with ly294002 resulted in an increase in the <span class="gene_color">p53-responsive</span> gene mdm2 , suggesting a role for <span class="gene_color">akt</span> in the tax-mediated regulation of p53 transcriptional activity .</td>
<td>0.773</td>
<td>0.825</td>
</tr>
<tr class="even">
<td>ABCB1</td>
<td>NR1I3</td>
<td>valproic acid induces cyp3a4 and <span class="gene_color">mdr1</span> gene expression by activation of <span class="gene_color">constitutive androstane receptor</span> and pregnane x receptor pathways .</td>
<td>0.762</td>
<td>0.813</td>
</tr>
<tr class="odd">
<td>PTH2R</td>
<td>PTH2</td>
<td>thus , the juxtamembrane receptor domain specifies the signaling and binding selectivity of <span class="gene_color">tip39</span> for the <span class="gene_color">pth2 receptor</span> over the pth1 receptor .</td>
<td>0.761</td>
<td>0.812</td>
</tr>
<tr class="even">
<td>CCND1</td>
<td>ABL1</td>
<td>synergy with <span class="gene_color">v-abl</span> depended on a motif in <span class="gene_color">cyclin d1</span> that mediates its binding to the retinoblastoma protein , suggesting that abl oncogenes in part mediate their mitogenic effects via a retinoblastoma protein-dependent pathway .</td>
<td>0.757</td>
<td>0.808</td>
</tr>
<tr class="odd">
<td>CTNND1</td>
<td>CDH1</td>
<td>these complexes are formed independently of ddr1 activation and of beta-catenin and <span class="gene_color">p120-catenin</span> binding to <span class="gene_color">e-cadherin</span> ; they are ubiquitous in epithelial cells .</td>
<td>0.748</td>
<td>0.798</td>
</tr>
<tr class="even">
<td>CSF1</td>
<td>CSF1R</td>
<td>this is in agreement with current thought that the <span class="gene_color">c-fms</span> proto-oncogene product functions as the <span class="gene_color">csf-1</span> receptor specific to this pathway .</td>
<td>0.745</td>
<td>0.795</td>
</tr>
<tr class="odd">
<td>EZR</td>
<td>CFTR</td>
<td>without <span class="gene_color">ezrin</span> binding , the cytoplasmic tail of <span class="gene_color">cftr</span> only interacts strongly with the first amino-terminal pdz domain to form a 1:1 c-cftr .</td>
<td>0.732</td>
<td>0.78</td>
</tr>
<tr class="even">
<td>SRC</td>
<td>PIK3CG</td>
<td>we have demonstrated that the sh2 ( <span class="gene_color">src</span> homology 2 ) domains of the 85 kda subunit of pi-3k are sufficient to mediate binding of the <span class="gene_color">pi-3k</span> complex to tyrosine phosphorylated , but not non-phosphorylated il-2r beta , suggesting that tyrosine phosphorylation is an integral component of the activation of pi-3k by the il-2r .</td>
<td>0.731</td>
<td>0.78</td>
</tr>
</tbody>
</table>
</div>
<div id="tbl:gg_bottom_ten_table" class="tablenos">
<table>
<caption><span>Table 10:</span> Contains the bottom ten Gene-interacts-Gene confidence scores before and after model calbration. Both gene mentions highlighted in <span class="gene_color">blue</span>. </caption>
<colgroup>
<col style="width: 18%" />
<col style="width: 18%" />
<col style="width: 12%" />
<col style="width: 25%" />
<col style="width: 24%" />
</colgroup>
<thead>
<tr class="header">
<th>Gene1 Symbol</th>
<th>Gene2 Symbol</th>
<th>Text</th>
<th>Before Calibration</th>
<th>After Calibration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>AGTR1</td>
<td>ACE</td>
<td>result ( s ) : the luteal tissue is the major site of ang ii , <span class="gene_color">ace</span> , <span class="gene_color">at1r</span> , and vegf , with highest staining intensity found during the midluteal phase and at pregnancy .</td>
<td>0.009</td>
<td>0.003</td>
</tr>
<tr class="even">
<td>ABCE1</td>
<td>ABCF2</td>
<td>in relation to normal melanocytes , abcb3 , abcb6 , abcc2 , abcc4 , <span class="gene_color">abce1</span> and <span class="gene_color">abcf2</span> were significantly increased in melanoma cell lines , whereas abca7 , abca12 , abcb2 , abcb4 , abcb5 and abcd1 showed lower expression levels .</td>
<td>0.008</td>
<td>0.002</td>
</tr>
<tr class="odd">
<td>IL4</td>
<td>IFNG</td>
<td>in contrast , il-13ralpha2 mrna expression was up-regulated by <span class="gene_color">ifn-gamma</span> plus <span class="gene_color">il-4</span> .</td>
<td>0.007</td>
<td>0.002</td>
</tr>
<tr class="even">
<td>FCAR</td>
<td>CD79A</td>
<td>we report here the presence of circulating soluble fcalphar ( <span class="gene_color">cd89 )</span> - <span class="gene_color">iga</span> complexes in patients with igan .</td>
<td>0.007</td>
<td>0.002</td>
</tr>
<tr class="odd">
<td>IL4</td>
<td>VCAM1</td>
<td>similarly , <span class="gene_color">il-4</span> induced <span class="gene_color">vcam-1</span> expression and augmented tnf-alpha-induced expression on huvec but did not affect vcam-1 expression on hdmec .</td>
<td>0.007</td>
<td>0.002</td>
</tr>
<tr class="even">
<td>IL2</td>
<td>IFNG</td>
<td>prostaglandin e2 at priming of naive cd4 + t cells inhibits acquisition of ability to produce <span class="gene_color">ifn-gamma</span> and <span class="gene_color">il-2</span> , but not il-4 and il-5 .</td>
<td>0.006</td>
<td>0.002</td>
</tr>
<tr class="odd">
<td>IL2</td>
<td>FOXP3</td>
<td>il-1b promotes tgf-b1 and <span class="gene_color">il-2</span> dependent <span class="gene_color">foxp3</span> expression in regulatory t cells .</td>
<td>0.006</td>
<td>0.002</td>
</tr>
<tr class="even">
<td>IL2</td>
<td>IFNG</td>
<td>the detailed distribution of lymphokine-producing cells showed that <span class="gene_color">il-2</span> and <span class="gene_color">ifn-gamma-producing</span> cells were located mainly in the follicular areas .</td>
<td>0.005</td>
<td>0.001</td>
</tr>
<tr class="odd">
<td>IFNG</td>
<td>IL10</td>
<td>results : we found weak mrna expression of interleukin-4 ( il-4 ) and il-5 , and strong expression of il-6 , <span class="gene_color">il-10</span> and <span class="gene_color">ifn-gamma</span> before therapy .</td>
<td>0.005</td>
<td>0.001</td>
</tr>
<tr class="even">
<td>PIK3R1</td>
<td>PTEN</td>
<td>both <span class="gene_color">pten</span> ( <span class="gene_color">pi3k</span> antagonist ) and pp2 ( unspecific phosphatase ) were down-regulated .</td>
<td>0.005</td>
<td>0.001</td>
</tr>
</tbody>
</table>
</div>
<h3 id="baseline-comparison">Baseline Comparison</h3>
<div id="fig:cocoscore_comparison" class="fignos">
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/0149086785b19f9429c92565d650e9d049c136ff/figures/literature_models/model_comparison.png" alt="" /><figcaption><span>Figure 10:</span> Comparion between our model and CoCoScore model <span class="citation" data-cites="IGXdryzB">[<a href="#ref-IGXdryzB" role="doc-biblioref">13</a>]</span>. We report both model’s performance in terms of AUROC and AUPR. Our model achieves comparable performance against CoCoScore in terms of AUROC. As for AUPR, CoCoScore consistently outperforms our model except for CtD.</figcaption>
</figure>
</div>
<p>Once our discriminator model is calibrated, we grouped sentences based on mention pair (edges). We assigned each edge the maximum score over all grouped sentences and compared our model’s ability to predict pairs in our test set to a previously published baseline model <span class="citation" data-cites="IGXdryzB">[<a href="#ref-IGXdryzB" role="doc-biblioref">13</a>]</span>. Performance is reported in terms of AUROC and AUPR (Figure <a href="#fig:cocoscore_comparison">10</a>). Across edge types our model shows comparable performance against the baseline in terms of AUROC. Regarding AUPR, our model shows hindered performance against the baseline. The exception for both cases is CtD where our model performs better than the baseline.</p>
<h3 id="reconstructing-hetionet">Reconstructing Hetionet</h3>
<div id="fig:hetionet_reconstruction" class="fignos">
<figure>
<img src="https://raw.githubusercontent.com/danich1/snorkeling/0149086785b19f9429c92565d650e9d049c136ff/figures/edge_prediction_experiment/edges_added.png" alt="" /><figcaption><span>Figure 11:</span> A scatter plot showing the number of edges (log scale) we can add or recall at specified precision levels. The blue depicts edges existing in hetionet and the orange depicts how many novel edges can be added.</figcaption>
</figure>
</div>
<p>We evaluated how many edges we can recall/add to Hetionet v1 (Supplemental Figure <a href="#fig:hetionet_reconstruction">11</a> and Table <a href="#tbl:edge_prediction_tbl">11</a>). In our evaluation we used edges assigned to our test set. Overall, we can recall a small amount of edges at high precision thresholds. A key example is CbG and GiG where we recalled only one exisiting edge at 100% precision. Despite the low recall, we are still able to add novel edges to DaG and CtD while retaining modest precision.</p>
<div id="tbl:edge_prediction_tbl" class="tablenos">
<table>
<caption><span>Table 11:</span> Contains the top ten predictions for each edge type. Highlighted words represent entities mentioned within the given sentence. </caption>
<colgroup>
<col style="width: 8%" />
<col style="width: 13%" />
<col style="width: 14%" />
<col style="width: 13%" />
<col style="width: 13%" />
<col style="width: 12%" />
<col style="width: 24%" />
</colgroup>
<thead>
<tr class="header">
<th>Edge Type</th>
<th>Source Node</th>
<th>Target Node</th>
<th>Gen Model Prediction</th>
<th>Disc Model Prediction</th>
<th>Number of Sentences</th>
<th>Text</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>lung cancer</td>
<td>VEGFA</td>
<td>1.000</td>
<td>0.912</td>
<td>3293</td>
<td>conclusion : the plasma <span class="gene_color">vegf</span> level is increased in <span class="disease_color">nsclc</span> patients with approximate1y one fourth to have cancer cells in the peripheral blood.</td>
</tr>
<tr class="even">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>hematologic cancer</td>
<td>TP53</td>
<td>1.000</td>
<td>0.905</td>
<td>8660</td>
<td>mutations of the <span class="gene_color">p53</span> gene were found in four cases of <span class="disease_color">cml</span> in blastic crisis ( bc ).</td>
</tr>
<tr class="odd">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>obesity</td>
<td>MC4R</td>
<td>1.000</td>
<td>0.901</td>
<td>1493</td>
<td>several mutations in the <span class="gene_color">melanocortin 4 receptor</span> gene are associated with <span class="disease_color">obesity</span> in chinese children and adolescents.</td>
</tr>
<tr class="even">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>Alzheimer’s disease</td>
<td>VLDLR</td>
<td>1.000</td>
<td>0.886</td>
<td>86</td>
<td>the 5-repeat allele in the <span class="gene_color">very-low-density lipoprotein receptor</span> gene polymorphism is not increased in sporadic <span class="disease_color">alzheimer ’s disease</span> in japanese.</td>
</tr>
<tr class="odd">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>lung cancer</td>
<td>XRCC1</td>
<td>1.000</td>
<td>0.885</td>
<td>662</td>
<td>results : <span class="gene_color">xrcc1</span> gene polymorphism is associated with increased risk of <span class="disease_color">lung cancer</span> when the arg/arg genotype was used as the reference group.</td>
</tr>
<tr class="even">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>prostate cancer</td>
<td>ESR1</td>
<td>1.000</td>
<td>0.883</td>
<td>500</td>
<td>conclusion : these results suggest that variants of the ggga polymorphism from the <span class="gene_color">estrogen receptor alpha</span> gene may be associated with an increased risk of developing <span class="disease_color">prostate cancer</span>.</td>
</tr>
<tr class="odd">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>breast cancer</td>
<td>REG1A</td>
<td>1.000</td>
<td>0.878</td>
<td>37</td>
<td>conclusion : high levels of <span class="gene_color">reg1a</span> expression within tumors are an independent predictor of poor prognosis in patients with <span class="disease_color">breast cancer</span>.</td>
</tr>
<tr class="even">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>breast cancer</td>
<td>INSR</td>
<td>1.000</td>
<td>0.877</td>
<td>200</td>
<td>we have previously reported that <span class="gene_color">insulin receptor</span> expression is increased in human <span class="disease_color">breast cancer</span> specimens ( v. papa et al. , j. clin.</td>
</tr>
<tr class="odd">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>rheumatoid arthritis</td>
<td>AR</td>
<td>1.000</td>
<td>0.877</td>
<td>53</td>
<td>conclusion : our results suggest no correlation between cag repeat polymorphism in the <span class="gene_color">ar</span> gene and response to treatment with lef in women with <span class="disease_color">ra</span>.</td>
</tr>
<tr class="even">
<td><span class="disease_color">D</span>a<span class="gene_color">G</span></td>
<td>coronary artery disease</td>
<td>CTLA4</td>
<td>1.000</td>
<td>0.875</td>
<td>12</td>
<td>conclusion : the g/g genotype polymorphism of the <span class="gene_color">ctla-4</span> gene is associated with increased risk of <span class="disease_color">ami</span>.</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Zonisamide</td>
<td>epilepsy syndrome</td>
<td>1.000</td>
<td>0.943</td>
<td>1011</td>
<td>adjunctive <span class="compound_color">zonisamide</span> therapy in the long-term treatment of children with <span class="disease_color">partial epilepsy</span> : results of an open-label extension study of a phase iii , randomized , double-blind , placebo-controlled trial.</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Metformin</td>
<td>polycystic ovary syndrome</td>
<td>1.000</td>
<td>0.942</td>
<td>3217</td>
<td>in the present study , 23 <span class="disease_color">pcos</span> subjects [ mean ( + / - se ) body mass index 30.0 + / -1.1 kg/m2 ] were randomly assigned to double-blind treatment with <span class="compound_color">metformin</span> ( 500 mg tid ) or placebo for 6 months , while maintaining their usual eating habits.</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Piroxicam</td>
<td>rheumatoid arthritis</td>
<td>1.000</td>
<td>0.928</td>
<td>184</td>
<td>methods : a double-blind , randomized , crossover trial in 49 patients with active <span class="disease_color">ra</span> compared 6 weeks of treatment with tenidap ( 120 mg/day ) versus 6 weeks of treatment with <span class="compound_color">piroxicam</span> ( 20 mg/day ).</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Irinotecan</td>
<td>stomach cancer</td>
<td>1.000</td>
<td>0.918</td>
<td>968</td>
<td>randomized phase ii trial of first-line treatment with tailored <span class="compound_color">irinotecan</span> and s-1 therapy versus s-1 monotherapy for advanced or recurrent <span class="disease_color">gastric carcinoma</span> ( jfmc31-0301 ).</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Treprostinil</td>
<td>hypertension</td>
<td>1.000</td>
<td>0.913</td>
<td>536</td>
<td>oral <span class="compound_color">treprostinil</span> for the treatment of <span class="disease_color">pulmonary arterial hypertension</span> in patients receiving background endothelin receptor antagonist and phosphodiesterase type 5 inhibitor therapy ( the freedom-c2 study ) : a randomized controlled trial.</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Colchicine</td>
<td>gout</td>
<td>1.000</td>
<td>0.911</td>
<td>78</td>
<td>this is the first in vivo data to provide a biological rationale that supports the implementation of low dose , non-toxic , <span class="compound_color">colchicine</span> therapy for the treatment of <span class="disease_color">gouty arthritis</span>.</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Propranolol</td>
<td>stomach cancer</td>
<td>1.000</td>
<td>0.898</td>
<td>45</td>
<td>74 cirrhotic patients with a history of variceal or <span class="disease_color">gastric bleeding</span> were randomly assigned to treatment with <span class="compound_color">propranolol</span> ( 40 to 360 mg/day ) or placebo.</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Reboxetine</td>
<td>endogenous depression</td>
<td>1.000</td>
<td>0.894</td>
<td>439</td>
<td>data were obtained from four short-term ( 4-8-week ) , randomized , placebo-controlled trials of <span class="compound_color">reboxetine</span> for the treatment of <span class="disease_color">mdd</span>.</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Diclofenac</td>
<td>ankylosing spondylitis</td>
<td>1.000</td>
<td>0.892</td>
<td>61</td>
<td>comparison of two different dosages of celecoxib with <span class="compound_color">diclofenac</span> for the treatment of active <span class="disease_color">ankylosing spondylitis</span> : results of a 12-week randomised , double-blind , controlled study.</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>t<span class="disease_color">D</span></td>
<td>Tapentadol</td>
<td>osteoarthritis</td>
<td>1.000</td>
<td>0.880</td>
<td>29</td>
<td>driving ability in patients with severe chronic low back or <span class="disease_color">osteoarthritis</span> knee pain on stable treatment with <span class="compound_color">tapentadol</span> prolonged release : a multicenter , open-label , phase 3b trial.</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>Dexamethasone</td>
<td>NR3C1</td>
<td>1.000</td>
<td>0.850</td>
<td>1119</td>
<td>submicromolar free calcium modulates <span class="compound_color">dexamethasone</span> binding to the <span class="gene_color">glucocorticoid receptor</span>.</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>Vitamin A</td>
<td>RBP4</td>
<td>1.000</td>
<td>0.807</td>
<td>5512</td>
<td>the authors give serum <span class="compound_color">retinol</span> binding protein ( <span class="gene_color">rbp )</span> normal values , established by immunonephelometry , for two healthy populations in their hospital laboratory.</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>D-Proline</td>
<td>IGFBP4</td>
<td>1.000</td>
<td>0.790</td>
<td>1</td>
<td>the insulin-like growth factor-i-stimulated <span class="compound_color">l-proline</span> uptake was inhibited by one of its binding protein , <span class="gene_color">insulin-like growth factor binding protein-4</span> , in a concentration-dependent manner.</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>Sucrose</td>
<td>AR</td>
<td>0.996</td>
<td>0.789</td>
<td>37</td>
<td>the amount ( maximal binding capacity of 24 to 30 fmol/mg protein ) and hormone binding affinity ( half-maximal saturation of 0.2 nm ) of the <span class="gene_color">androgen receptor</span> in cultured skin fibroblasts was normal , but the receptor was qualitatively abnormal as evidenced by instability on <span class="compound_color">sucrose</span> density gradient centrifugation.</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>D-Lysine</td>
<td>PLG</td>
<td>1.000</td>
<td>0.787</td>
<td>403</td>
<td>in both elisa and rocket immunoelectrophoresis systems , complex formation was inhibited by 10 mm epsilon-amino-n-caproic acid , implying that there is a role for the <span class="compound_color">lysine</span> binding sites of <span class="gene_color">plg</span> in mediating the interaction.</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>Adenosine</td>
<td>INSR</td>
<td>1.000</td>
<td>0.785</td>
<td>129</td>
<td>these findings demonstrate basal state binding of <span class="compound_color">atp</span> to the ckd leading to cis-autophosphorylation and novel basal state regulatory interactions among the subdomains of the <span class="gene_color">insulin receptor</span> kinase.</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>Adenosine</td>
<td>PLK1</td>
<td>1.000</td>
<td>0.783</td>
<td>104</td>
<td>most kinase inhibitors interact with the <span class="compound_color">atp</span> binding site on <span class="gene_color">plk1</span> , which is highly conserved.</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>Calcium Chloride</td>
<td>ITPR3</td>
<td>0.995</td>
<td>0.777</td>
<td>1954</td>
<td>control of <span class="compound_color">ca2</span> + influx in human neutrophils by inositol 1,4,5-trisphosphate ( ip3 ) binding : differential effects of micro-injected <span class="gene_color">ip3 receptor</span> antagonists.</td>
</tr>
<tr class="odd">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>D-Arginine</td>
<td>C5AR1</td>
<td>1.000</td>
<td>0.775</td>
<td>808</td>
<td>thus , selected out of a multiplicity of possibilities by the natural binding partner , <span class="compound_color">arg37</span> as well as arg40 appear to be anchor residues in binding to the <span class="gene_color">c5a receptor</span>.</td>
</tr>
<tr class="even">
<td><span class="compound_color">C</span>b<span class="gene_color">G</span></td>
<td>Ticagrelor</td>
<td>P2RY12</td>
<td>1.000</td>
<td>0.773</td>
<td>322</td>
<td>purpose : <span class="compound_color">ticagrelor</span> is a reversibly binding <span class="gene_color">p2y12</span> receptor antagonist used clinically for the prevention of atherothrombotic events in patients with acute coronary syndromes ( acs ).</td>
</tr>
<tr class="odd">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>ABL1</td>
<td>ABL1</td>
<td>0.999</td>
<td>0.600</td>
<td>9572</td>
<td>the acquired resistance in patients who failed to respond to imatinib seemed to be induced by several point mutations in the <span class="gene_color">bcr-abl</span> gene , which were likely to affect the binding of imatinib with <span class="gene_color">bcr-abl</span>.</td>
</tr>
<tr class="even">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>TP63</td>
<td>TP53</td>
<td>1.000</td>
<td>0.595</td>
<td>2557</td>
<td><span class="gene_color">tp63</span> , a member of the <span class="gene_color">p53</span> gene family gene , encodes the np63 protein and is one of the most frequently amplified genes in squamous cell carcinomas ( scc ) of the head and neck ( hnscc ) and lungs ( lusc ).</td>
</tr>
<tr class="odd">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>FERMT1</td>
<td>FERMT1</td>
<td>0.004</td>
<td>0.590</td>
<td>194</td>
<td>ks is caused by mutations in the <span class="gene_color">fermt1</span> gene encoding <span class="gene_color">kindlin-1</span>.</td>
</tr>
<tr class="even">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>GRN</td>
<td>GRN</td>
<td>1.000</td>
<td>0.590</td>
<td>3842</td>
<td>background : mutations in the <span class="gene_color">progranulin</span> gene ( <span class="gene_color">pgrn )</span> have recently been identified as a cause of frontotemporal lobar degeneration with ubiquitin-positive inclusions ( ftld-u ) in some families.</td>
</tr>
<tr class="odd">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>FASN</td>
<td>EP300</td>
<td>0.999</td>
<td>0.589</td>
<td>6</td>
<td>here , we demonstrated that <span class="gene_color">p300</span> binds to and increases histone h3 lysine 27 acetylation ( h3k27ac ) in the <span class="gene_color">fasn</span> gene promoter.</td>
</tr>
<tr class="even">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>SETBP1</td>
<td>SETBP1</td>
<td>1.000</td>
<td>0.588</td>
<td>354</td>
<td>the critical deleted region contains <span class="gene_color">setbp1</span> gene ( <span class="gene_color">set binding protein 1 )</span>.</td>
</tr>
<tr class="odd">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>BCL2</td>
<td>BAK1</td>
<td>0.118</td>
<td>0.587</td>
<td>1220</td>
<td>different expression patterns of <span class="gene_color">bcl-2</span> family genes in breast cancer by estrogen receptor status with special reference to pro-apoptotic <span class="gene_color">bak</span> gene.</td>
</tr>
<tr class="even">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>SP1</td>
<td>INSR</td>
<td>0.948</td>
<td>0.587</td>
<td>23</td>
<td>thus , the efficient expression of the human <span class="gene_color">insulin receptor</span> gene possibly requires the binding of transcriptional factor <span class="gene_color">sp1</span> to four g-c boxes located -593 to -618 base pairs upstream of the atg translation initiation codon.</td>
</tr>
<tr class="odd">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>ABCD1</td>
<td>ABCD1</td>
<td>1.000</td>
<td>0.586</td>
<td>410</td>
<td>x-linked adrenoleukodystrophy ( x-ald ) is caused by mutations in the <span class="gene_color">abcd1</span> gene encoding the peroxisomal abc transporter adrenoleukodystrophy protein ( <span class="gene_color">aldp )</span>.</td>
</tr>
<tr class="even">
<td><span class="gene_color">G</span>i<span class="gene_color">G</span></td>
<td>CYP1A1</td>
<td>AHR</td>
<td>0.996</td>
<td>0.586</td>
<td>1940</td>
<td>the liganded <span class="gene_color">ah receptor</span> activates transcription by binding to a specific dna-recognition motif within a dioxin-responsive enhancer upstream of the <span class="gene_color">cyp1a1</span> gene.</td>
</tr>
</tbody>
</table>
</div>
<section class="footnotes" role="doc-endnotes">
<hr />
<ol>
<li id="fn1" role="doc-endnote"><p>Labeled sentences are available <a href="https://github.com/greenelab/text_mined_hetnet_manuscript/tree/master/supplementary_materials/annotated_sentences">here</a>.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>
<!-- default theme -->

<style>
    /* import google fonts */
    @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
    @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

    /* -------------------------------------------------- */
    /* global */
    /* -------------------------------------------------- */

    /* all elements */
    * {
        /* force sans-serif font unless specified otherwise */
        font-family: "Open Sans", "Helvetica", sans-serif;

        /* prevent text inflation on some mobile browsers */
        -webkit-text-size-adjust: none !important;
        -moz-text-size-adjust: none !important;
        -o-text-size-adjust: none !important;
        text-size-adjust: none !important;
    }

    @media only screen {
        /* "page" element */
        body {
            position: relative;
            box-sizing: border-box;
            font-size: 12pt;
            line-height: 1.5;
            max-width: 8.5in;
            margin: 20px auto;
            padding: 40px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* "page" element */
        body {
            padding: 20px;
            margin: 0;
            border-radius: 0;
            border: none;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
            background: none;
        }
    }

    /* -------------------------------------------------- */
    /* headings */
    /* -------------------------------------------------- */

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
        margin: 20px 0;
        padding: 0;
        font-weight: bold;
    }

    /* biggest heading */
    h1 {
        margin: 40px 0;
        text-align: center;
    }

    /* second biggest heading */
    h2 {
        margin-top: 30px;
        padding-bottom: 5px;
        border-bottom: solid 1px #bdbdbd;
    }

    /* -------------------------------------------------- */
    /* manuscript header */
    /* -------------------------------------------------- */

    /* manuscript title */
    header > h1 {
        margin: 0;
    }

    /* manuscript title caption text (ie "automatically generated on") */
    header + p {
        text-align: center;
        margin-top: 10px;
    }

    /* -------------------------------------------------- */
    /* text elements */
    /* -------------------------------------------------- */

    /* links */
    a {
        color: #2196f3;
        overflow-wrap: break-word;
    }

    /* normal links (not empty, not button link, not syntax highlighting link) */
    a:not(:empty):not(.button):not(.sourceLine) {
        padding-left: 1px;
        padding-right: 1px;
    }

    /* superscripts and subscripts */
    sub,
    sup {
        /* prevent from affecting line height */
        line-height: 0;
    }

    /* unordered and ordered lists*/
    ul,
    ol {
        padding-left: 20px;
    }

    /* class for styling text semibold */
    .semibold {
        font-weight: 600;
    }

    /* class for styling elements horizontally left aligned */
    .left {
        display: block;
        text-align: left;
        margin-left: auto;
        margin-right: 0;
        justify-content: left;
    }

    /* class for styling elements horizontally centered */
    .center {
        display: block;
        text-align: center;
        margin-left: auto;
        margin-right: auto;
        justify-content: center;
    }

    /* class for styling elements horizontally right aligned */
    .right {
        display: block;
        text-align: right;
        margin-left: 0;
        margin-right: auto;
        justify-content: right;
    }

    /* -------------------------------------------------- */
    /* section elements */
    /* -------------------------------------------------- */

    /* horizontal divider line */
    hr {
        border: none;
        height: 1px;
        background: #bdbdbd;
    }

    /* paragraphs, horizontal dividers, figures, tables, code */
    p,
    hr,
    figure,
    table,
    pre {
        /* treat all as "paragraphs", with consistent vertical margins */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* figures */
    /* -------------------------------------------------- */

    /* figure */
    figure {
        max-width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure caption */
    figcaption {
        padding: 0;
        padding-top: 10px;
    }

    /* figure image element */
    figure img {
        max-width: 100%;
        display: block;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure auto-number */
    img + figcaption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* tables */
    /* -------------------------------------------------- */

    /* table */
    table {
        border-collapse: collapse;
        border-spacing: 0;
        width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* table cells */
    th,
    td {
        border: solid 1px #bdbdbd;
        padding: 10px;
        /* squash table if too wide for page by forcing line breaks */
        overflow-wrap: break-word;
    }

    /* header row and even rows */
    th,
    tr:nth-child(2n) {
        background-color: #fafafa;
    }

    /* odd rows */
    tr:nth-child(2n + 1) {
        background-color: #ffffff;
    }

    /* table caption */
    caption {
        text-align: left;
        padding: 0;
        padding-bottom: 10px;
    }

    /* table auto-number */
    table > caption > span:first-of-type,
    div.table_wrapper > table > caption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* code */
    /* -------------------------------------------------- */

    /* multi-line code block */
    pre {
        padding: 10px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
        break-inside: avoid;
        text-align: left;
    }

    /* inline code, ie code within normal text */
    :not(pre) > code {
        padding: 0 4px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
    }

    /* code text */
    /* apply all children, to reach syntax highlighting sub-elements */
    code,
    code * {
        /* force monospace font */
        font-family: "Source Code Pro", "Courier New", monospace;
    }

    /* -------------------------------------------------- */
    /* quotes */
    /* -------------------------------------------------- */

    /* quoted text */
    blockquote {
        margin: 0;
        padding: 0;
        border-left: 4px solid #bdbdbd;
        padding-left: 16px;
        break-inside: avoid;
    }

    /* -------------------------------------------------- */
    /* banners */
    /* -------------------------------------------------- */

    /* info banners */
    .banner {
        box-sizing: border-box;
        display: block;
        position: relative;
        width: 100%;
        margin-top: 20px;
        margin-bottom: 20px;
        padding: 20px;
        text-align: center;
    }

    /* paragraph in banner */
    .banner > p {
        margin: 0;
    }

    /* -------------------------------------------------- */
    /* highlight colors */
    /* -------------------------------------------------- */

    .white {
        background: #ffffff;
    }
    .lightgrey {
        background: #eeeeee;
    }
    .grey {
        background: #757575;
    }
    .darkgrey {
        background: #424242;
    }
    .black {
        background: #000000;
    }
    .lightred {
        background: #ffcdd2;
    }
    .lightyellow {
        background: #ffecb3;
    }
    .lightgreen {
        background: #dcedc8;
    }
    .lightblue {
        background: #e3f2fd;
    }
    .lightpurple {
        background: #f3e5f5;
    }
    .red {
        background: #f44336;
    }
    .orange {
        background: #ff9800;
    }
    .yellow {
        background: #ffeb3b;
    }
    .green {
        background: #4caf50;
    }
    .blue {
        background: #2196f3;
    }
    .purple {
        background: #9c27b0;
    }
    .white,
    .lightgrey,
    .lightred,
    .lightyellow,
    .lightgreen,
    .lightblue,
    .lightpurple,
    .orange,
    .yellow,
    .white a,
    .lightgrey a,
    .lightred a,
    .lightyellow a,
    .lightgreen a,
    .lightblue a,
    .lightpurple a,
    .orange a,
    .yellow a {
        color: #000000;
    }
    .grey,
    .darkgrey,
    .black,
    .red,
    .green,
    .blue,
    .purple,
    .grey a,
    .darkgrey a,
    .black a,
    .red a,
    .green a,
    .blue a,
    .purple a {
        color: #ffffff;
    }

    /* -------------------------------------------------- */
    /* buttons */
    /* -------------------------------------------------- */

    /* class for styling links like buttons */
    .button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        margin: 5px;
        padding: 10px 20px;
        font-size: 0.75em;
        font-weight: 600;
        text-transform: uppercase;
        text-decoration: none;
        letter-spacing: 1px;
        background: none;
        color: #2196f3;
        border: solid 1px #bdbdbd;
        border-radius: 5px;
    }

    /* buttons when hovered */
    .button:hover:not([disabled]),
    .icon_button:hover:not([disabled]) {
        cursor: pointer;
        background: #f5f5f5;
    }

    /* buttons when disabled */
    .button[disabled],
    .icon_button[disabled] {
        opacity: 0.35;
        pointer-events: none;
    }

    /* class for styling buttons containg only single icon */
    .icon_button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        text-decoration: none;
        margin: 0;
        padding: 0;
        background: none;
        border-radius: 5px;
        border: none;
        width: 20px;
        height: 20px;
        min-width: 20px;
        min-height: 20px;
    }

    /* icon button inner svg image */
    .icon_button > svg {
        height: 16px;
    }

    /* -------------------------------------------------- */
    /* icons */
    /* -------------------------------------------------- */

    /* class for styling icons inline with text */
    .inline_icon {
        height: 1em;
        position: relative;
        top: 0.125em;
    }

    /* -------------------------------------------------- */
    /* print control */
    /* -------------------------------------------------- */

    @media print {
        @page {
            /* suggested printing margin */
            margin: 0.5in;
        }

        /* document and "page" elements */
        html, body {
            margin: 0;
            padding: 0;
            width: 100%;
            height: 100%;
        }

        /* "page" element */
        body {
            font-size: 11pt !important;
            line-height: 1.35;
        }

        /* all headings */
        h1,
        h2,
        h3,
        h4,
        h5,
        h6 {
            margin: 15px 0;
        }

        /* heading 1 */
        h1 {
            font-size: 1.75em;
        }

        /* heading 2 */
        h2 {
            font-size: 1.25em;
            margin-top: 0;
        }

        /* heading 3 */
        h3 {
            font-size: 1.10em;
        }

        /* figures and tables */
        figure, table {
            font-size: 0.85em;
        }

        /* table cells */
        th,
        td {
            padding: 5px;
        }

        /* shrink font awesome icons */
        i.fas,
        i.fab,
        i.far,
        i.fal {
            transform: scale(0.85);
        }

        /* decrease banner margins */
        .banner {
            margin-top: 15px;
            margin-bottom: 15px;
            padding: 15px;
        }

        /* class for centering an element vertically on its own page */
        .page_center {
            margin: auto;
            width: 100%;
            height: 100%;
            display: flex;
            align-items: center;
            vertical-align: middle;
            break-before: page;
            break-after: page;
        }

        /* always insert a page break before the element */
        .page_break_before {
            break-before: page;
        }

        /* always insert a page break after the element */
        .page_break_after {
            break-after: page;
        }

        /* avoid page break before the element */
        .page_break_before_avoid {
            break-before: avoid;
        }

        /* avoid page break after the element */
        .page_break_after_avoid {
            break-after: avoid;
        }

        /* avoid page break inside the element */
        .page_break_inside_avoid {
            break-inside: avoid;
        }
    }

    /* -------------------------------------------------- */
    /* override pandoc css quirks */
    /* -------------------------------------------------- */

    .sourceCode {
        /* prevent unsightly overflow in wide code blocks */
        overflow: auto !important;
    }

    div.sourceCode {
        /* prevent background fill on top-most code block  container */
        background: none !important;
    }

    .sourceCode * {
        /* force consistent line spacing */
        line-height: 1.5 !important;
    }

    div.sourceCode {
        /* style code block margins same as <pre> element */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* mathjax */
    /* -------------------------------------------------- */

    /* mathjax containers */
    .math.display > span:not(.MathJax_Preview) {
        /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
        display: flex !important;
        overflow-x: auto !important;
        overflow-y: hidden !important;
        justify-content: center;
        align-items: center;
        margin: 0 !important;
    }

    /* right click menu */
    .MathJax_Menu {
        border-radius: 5px !important;
        border: solid 1px #bdbdbd !important;
        box-shadow: none !important;
    }

    /* equation auto-number */
    span[id^="eq:"] > span.math.display + span {
        font-weight: 600;
    }

    /* equation */
    span[id^="eq:"] > span.math.display > span {
        /* nudge to make room for equation auto-number */
        margin-right: 40px !important;
    }

    /* -------------------------------------------------- */
    /* table scroll plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* table wrapper */
        .table_wrapper {
            /* show scrollbar on tables if necessary to prevent overflow */
            overflow: auto;
            width: 100%;
            margin: 20px 0;
        }

        /* table within table wrapper */
        .table_wrapper table,
        .table_wrapper table * {
            /* don't break table words */
            word-break: normal !important;
        }

        .table_wrapper > table {
            /* move margins from table to table_wrapper to allow margin collapsing */
            margin: 0;
        }
    }

    /* -------------------------------------------------- */
    /* anchors plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anchor button */
        .anchor {
            opacity: 0;
            margin-left: 5px;
        }

        /* anchor buttons within <h2>'s */
        h2 .anchor {
            margin-left: 10px;
        }

        /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
        *:hover > .anchor,
        .anchor:hover,
        .anchor:focus {
            opacity: 1;
        }

        /* anchor button when hovered */
        .anchor:hover {
            cursor: pointer;
        }
    }

    /* always show anchor button on devices with no mouse/hover ability */
    @media (hover: none) {
        .anchor {
            opacity: 1;
        }
    }

    /* always hide anchor button on print */
    @media only print {
        .anchor {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* accordion plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* accordion arrow button */
        .accordion_arrow {
            margin-right: 10px;
        }

        /* arrow icon when <h2> data-collapsed attribute true */
        h2[data-collapsed="true"] > .accordion_arrow > svg {
            transform: rotate(-90deg);
        }

        /* all elements (except <h2>'s) when data-collapsed attribute true */
        *:not(h2)[data-collapsed="true"] {
            display: none;
        }

        /* accordion arrow button when hovered and <h2>'s when hovered */
        .accordion_arrow:hover,
        h2[data-collapsed="true"]:hover,
        h2[data-collapsed="false"]:hover {
            cursor: pointer;
        }
    }

    /* always hide accordion arrow button on print */
    @media only print {
        .accordion_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* tooltips plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* tooltip container */
        #tooltip {
            position: absolute;
            width: 50%;
            min-width: 240px;
            max-width: 75%;
            z-index: 1;
        }

        /* tooltip content */
        #tooltip_content {
            margin-bottom: 5px;
            padding: 20px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
            overflow-wrap: break-word;
        }

        /* tooltip copy of paragraphs and figures */
        #tooltip_content > p,
        #tooltip_content > figure {
            margin: 0;
            max-height: 320px;
            overflow-y: auto;
        }

        /* tooltip copy of <img> */
        #tooltip_content > figure > img {
            max-height: 260px;
        }

        /* navigation bar */
        #tooltip_nav_bar {
            margin-top: 10px;
            text-align: center;
        }

        /* navigation bar previous/next buton */
        #tooltip_nav_bar > .icon_button {
            position: relative;
            top: 3px;
        }

        /* navigation bar previous button */
        #tooltip_nav_bar > .icon_button:first-of-type {
            margin-right: 5px;
        }

        /* navigation bar next button */
        #tooltip_nav_bar > .icon_button:last-of-type {
            margin-left: 5px;
        }
    }

    /* always hide tooltip on print */
    @media only print {
        #tooltip {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* jump to first plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* jump button */
        .jump_arrow {
            position: relative;
            top: 0.125em;
            margin-right: 5px;
        }
    }

    /* always hide jump button on print */
    @media only print {
        .jump_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* link highlight plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anything with data-highlighted attribute true */
        [data-highlighted="true"] {
            background: #ffeb3b;
        }

        /* anything with data-selected attribute true */
        [data-selected="true"] {
            background: #ff8a65 !important;
        }

        /* animation definition for glow */
        @keyframes highlight_glow {
            0% {
                background: none;
            }
            10% {
                background: #bbdefb;
            }
            100% {
                background: none;
            }
        }

        /* anything with data-glow attribute true */
        [data-glow="true"] {
            animation: highlight_glow 2s;
        }
    }

    /* -------------------------------------------------- */
    /* table of contents plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* toc panel */
        #toc_panel {
            box-sizing: border-box;
            position: fixed;
            top: 0;
            left: 0;
            background: #ffffff;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            z-index: 2;
        }

        /* toc panel when closed */
        #toc_panel[data-open="false"] {
            min-width: 60px;
            width: 60px;
            height: 60px;
            border-right: solid 1px #bdbdbd;
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc panel when open */
        #toc_panel[data-open="true"] {
            min-width: 260px;
            max-width: 480px;
            /* keep panel edge consistent distance away from "page" edge */
            width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
            bottom: 0;
            border-right: solid 1px #bdbdbd;
        }

        /* toc panel header */
        #toc_header {
            box-sizing: border-box;
            display: flex;
            flex-direction: row;
            align-items: center;
            height: 60px;
            margin: 0;
            padding: 20px;
        }

        /* toc panel header when hovered */
        #toc_header:hover {
            cursor: pointer;
        }

        /* toc panel header when panel open */
        #toc_panel[data-open="true"] > #toc_header {
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc open/close header button */
        #toc_button {
            margin-right: 20px;
        }

        /* hide toc list and header text when closed */
        #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
        #toc_panel[data-open="false"] > #toc_list {
            display: none;
        }

        /* toc list of entries */
        #toc_list {
            box-sizing: border-box;
            width: 100%;
            padding: 20px;
            position: absolute;
            top: calc(60px + 1px);
            bottom: 0;
            overflow: auto;
        }

        /* toc entry, link to section in document */
        .toc_link {
            display: block;
            padding: 5px;
            position: relative;
            font-weight: 600;
            text-decoration: none;
        }

        /* toc entry when hovered or when "viewed" */
        .toc_link:hover,
        .toc_link[data-viewing="true"] {
            background: #f5f5f5;
        }

        /* toc entry, level 1 indentation */
        .toc_link[data-level="1"] {
            margin-left: 0;
        }

        /* toc entry, level 2 indentation */
        .toc_link[data-level="2"] {
            margin-left: 20px;
        }

        /* toc entry, level 3 indentation */
        .toc_link[data-level="3"] {
            margin-left: 40px;
        }

        /* toc entry, level 4 indentation */
        .toc_link[data-level="4"] {
            margin-left: 60px;
        }

        /* toc entry bullets */
        #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
            position: absolute;
            left: -15px;
            top: -1px;
            font-size: 1.5em;
        }

        /* toc entry, level 2 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
            content: "\2022";
        }

        /* toc entry, level 3 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
            content: "\25AB";
        }

        /* toc entry, level 4 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
            content: "-";
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* push <body> ("page") element down to make room for toc icon */
        .toc_body_nudge {
            padding-top: 60px;
        }

        /* toc icon when panel closed and not hovered */
        #toc_panel[data-open="false"]:not(:hover) {
            background: rgba(255, 255, 255, 0.75);
        }
    }

    /* always hide toc panel on print */
    @media only print {
        #toc_panel {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* lightbox plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* regular <img> in document when hovered */
        .lightbox_document_img:hover {
            cursor: pointer;
        }

        .body_no_scroll {
            overflow: hidden !important;
        }

        /* screen overlay */
        #lightbox_overlay {
            display: flex;
            flex-direction: column;
            position: fixed;
            left: 0;
            top: 0;
            right: 0;
            bottom: 0;
            background: rgba(0, 0, 0, 0.75);
            z-index: 3;
        }

        /* middle area containing lightbox image */
        #lightbox_image_container {
            flex-grow: 1;
            display: flex;
            justify-content: center;
            align-items: center;
            overflow: hidden;
            position: relative;
            padding: 20px;
        }

        /* bottom area containing caption */
        #lightbox_bottom_container {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 100px;
            min-height: 100px;
            max-height: 100px;
            background: rgba(0, 0, 0, 0.5);
        }

        /* image number info text box */
        #lightbox_number_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            left: 2px;
            top: 0;
            z-index: 4;
        }

        /* zoom info text box */
        #lightbox_zoom_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            right: 2px;
            top: 0;
            z-index: 4;
        }

        /* copy of image caption */
        #lightbox_caption {
            box-sizing: border-box;
            display: inline-block;
            width: 100%;
            max-height: 100%;
            padding: 10px 0;
            text-align: center;
            overflow-y: auto;
            color: #ffffff;
        }

        /* navigation previous/next button */
        .lightbox_button {
            width: 100px;
            height: 100%;
            min-width: 100px;
            min-height: 100%;
            color: #ffffff;
        }

        /* navigation previous/next button when hovered */
        .lightbox_button:hover {
            background: none !important;
        }

        /* navigation button icon */
        .lightbox_button > svg {
            height: 25px;
        }

        /* figure auto-number */
        #lightbox_caption > span:first-of-type {
            font-weight: bold;
            margin-right: 5px;
        }

        /* lightbox image when hovered */
        #lightbox_img:hover {
            cursor: grab;
        }

        /* lightbox image when grabbed */
        #lightbox_img:active {
            cursor: grabbing;
        }
    }

    /* when on screen < 480px wide */
    @media only screen and (max-width: 480px) {
        /* make navigation buttons skinnier on small screens to make more room for caption text */
        .lightbox_button {
            width: 50px;
            min-width: 50px;
        }
    }

    /* always hide lightbox on print */
    @media only print {
        #lightbox_overlay {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* hypothesis (annotations) plugin */
    /* -------------------------------------------------- */

    /* side panel */
    .annotator-frame {
        width: 280px !important;
        z-index: 0 !important;
    }

    /* match highlight color to rest of theme */
    .annotator-highlights-always-on .annotator-hl {
        background-color: #ffeb3b !important;
    }

    /* match focused color to rest of theme */
    .annotator-hl.annotator-hl-focused {
        background-color: #ff8a65 !important;
    }

    /* match bucket bar color to rest of theme */
    .annotator-bucket-bar {
        background: #f5f5f5 !important;
    }

    /* always hide toolbar and tooltip on print */
    @media only print {
        .annotator-frame {
            display: none !important;
        }

        hypothesis-adder {
            display: none !important;
        }
    }
</style>
<!-- table scroll plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows tables that are too wide to fit within
        // the page to have a scrollbar instead of being squashed to fit.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tableScroll';

        // default plugin options
        const options = {
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // wrap each table in a container
            const tables = document.querySelectorAll('table');
            for (const table of tables)
                wrapElement(table).classList.add('table_wrapper');
            // table_wrapper CSS class in theme file provides scroll
            // functionality
        }

        // wrap element in div and return div
        function wrapElement(element) {
            const parent = element.parentNode;
            const wrapper = document.createElement('div');
            parent.replaceChild(wrapper, element);
            wrapper.appendChild(element);
            return wrapper;
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- anchors plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds an anchor next to each of a certain type
        // of element that provides a human-readable url to that specific
        // item/position in the document (eg "manuscript.html#abstract"). It
        // also makes it such that scrolling out of view of a target removes
        // its identifier from the url.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'anchors';

        // default plugin options
        const options = {
            // which types of elements to add anchors next to, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3, figure, table',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // add anchor to each element of specified types
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements)
                addAnchor(element);

            // attach scroll listener to window
            window.addEventListener('scroll', onScroll);
        }

        // when window is scrolled
        function onScroll() {
            // if url has hash and user has scrolled out of view of hash
            // target, remove hash from url
            const target = getHashTarget();
            if (target) {
                if (
                    target.getBoundingClientRect().top > window.innerHeight ||
                    target.getBoundingClientRect().bottom < 0
                )
                    history.pushState(null, null, ' ');
            }
        }

        // add anchor to element
        function addAnchor(element) {
            let withId; // element with unique id
            let addTo; // element to add anchor button to

            // if figure or table, modify withId and addTo to get expected
            // elements
            if (element.tagName.toLowerCase() === 'figure') {
                withId = element.querySelector('img');
                addTo = element.querySelector('figcaption');
            } else if (element.tagName.toLowerCase() === 'table') {
                withId =
                    element.previousElementSibling ||
                    element.parentNode.previousElementSibling;
                addTo = element.querySelector('caption');
            }

            withId = withId || element;
            addTo = addTo || element;
            const id = withId.id || withId.name || null;

            // do not add anchor if element doesn't have assigned id.
            // id is generated by pandoc and is assumed to be unique and
            // human-readable
            if (!id)
                return;

            // create anchor button
            const anchor = document.createElement('a');
            anchor.innerHTML = document.querySelector('.icon_link').innerHTML;
            anchor.title = 'Link to this part of the document';
            anchor.classList.add('icon_button', 'anchor');
            anchor.dataset.ignore = 'true';
            anchor.href = '#' + id;
            addTo.appendChild(anchor);
        }

        // get element that is target of link or url hash
        function getHashTarget() {
            const hash = window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector(
                '[id="' + id + '"], [name="' + id + '"]'
            );
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (hash.indexOf('#fig:') === 0)
                target = target.parentNode;
            if (hash.indexOf('#tbl:') === 0)
                target = target.nextElementSibling;

            return target;
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- link icon -->

<template class="icon_link">
    <!-- modified from: https://fontawesome.com/icons/link -->
    <svg width="16" height="16" viewBox="0 0 512 512">
        <path
            fill="currentColor"
            d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
        ></path>
    </svg>
</template>
<!-- accordion plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows sections of content under <h2> headings
        // to be collapsible.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'accordion';

        // default plugin options
        const options = {
            // whether to always start expanded ('false'), always start
            // collapsed ('true'), or start collapsed when screen small ('auto')
            startCollapsed: 'auto',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <h2> heading
            const headings = document.querySelectorAll('h2');
            for (const heading of headings) {
                addArrow(heading);

                // start expanded/collapsed based on option
                if (
                    options.startCollapsed === 'true' ||
                    (options.startCollapsed === 'auto' && isSmallScreen())
                )
                    collapseHeading(heading);
                else
                    expandHeading(heading);
            }

            // attach hash change listener to window
            window.addEventListener('hashchange', onHashChange);
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                goToElement(target);
        }

        // add arrow to heading
        function addArrow(heading) {
            // add arrow button
            const arrow = document.createElement('button');
            arrow.innerHTML = document.querySelector(
                '.icon_angle_down'
            ).innerHTML;
            arrow.classList.add('icon_button', 'accordion_arrow');
            heading.insertBefore(arrow, heading.firstChild);

            // attach click listener to heading and button
            heading.addEventListener('click', onHeadingClick);
            arrow.addEventListener('click', onArrowClick);
        }

        // determine if on mobile-like device with small screen
        function isSmallScreen() {
            return Math.min(window.innerWidth, window.innerHeight) < 480;
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get element that is target of hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector(
                '[id="' + id + '"], [name="' + id + '"]'
            );
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (hash.indexOf('#fig:') === 0)
                target = target.parentNode;
            if (hash.indexOf('#tbl:') === 0)
                target = target.nextElementSibling;

            return target;
        }

        // when <h2> heading is clicked
        function onHeadingClick(event) {
            // only collapse if <h2> itself is target of click (eg, user did
            // not click on anchor within <h2>)
            if (event.target === this)
                toggleCollapse(this);
        }

        // when arrow button is clicked
        function onArrowClick() {
            toggleCollapse(this.parentNode);
        }

        // collapse section if expanded, expand if collapsed
        function toggleCollapse(heading) {
            if (heading.dataset.collapsed === 'false')
                collapseHeading(heading);
            else
                expandHeading(heading);
        }

        // elements to exclude from collapse, such as table of contents panel,
        // hypothesis panel, etc
        const exclude = '#toc_panel, div.annotator-frame, #lightbox_overlay';

        // collapse section
        function collapseHeading(heading) {
            heading.setAttribute('data-collapsed', 'true');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'true');
        }

        // expand section
        function expandHeading(heading) {
            heading.setAttribute('data-collapsed', 'false');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'false');
        }

        // get list of elements between this <h2> and next <h2> or <h1>
        // ("children" of the <h2> section)
        function getChildren(heading) {
            return nextUntil(heading, 'h2, h1', exclude);
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get list of elements after a start element up to element matching
        // query
        function nextUntil(element, query, exclude) {
            const elements = [];
            while (element = element.nextElementSibling, element) {
                if (element.matches(query))
                    break;
                if (!element.matches(exclude))
                    elements.push(element);
            }
            return elements;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
    <!-- modified from: https://fontawesome.com/icons/angle-down -->
    <svg width="16" height="16" viewBox="0 0 448 512">
        <path
            fill="currentColor"
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        ></path>
    </svg>
</template>
<!-- tooltips plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when the user hovers or
        // focuses a link to a citation or figure, a tooltip appears with a
        // preview of the reference content, along with arrows to navigate
        // between instances of the same reference in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tooltips';

        // default plugin options
        const options = {
            // whether user must click off to close tooltip instead of just
            // un-hovering
            clickClose: 'false',
            // delay (in ms) between opening and closing tooltip
            delay: '100',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach hover and focus listeners to link
                link.addEventListener('mouseover', onLinkHover);
                link.addEventListener('mouseleave', onLinkUnhover);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('touchend', onLinkTouch);
            }

            // attach mouse, key, and resize listeners to window
            window.addEventListener('mousedown', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('keyup', onKeyUp);
            window.addEventListener('resize', onResize);
        }

        // when link is hovered
        function onLinkHover() {
            // function to open tooltip
            const delayOpenTooltip = function() {
                openTooltip(this);
            }.bind(this);

            // run open function after delay
            this.openTooltipTimer = window.setTimeout(
                delayOpenTooltip,
                options.delay
            );
        }

        // when mouse leaves link
        function onLinkUnhover() {
            // cancel opening tooltip
            window.clearTimeout(this.openTooltipTimer);

            // don't close on unhover if option specifies
            if (options.clickClose === 'true')
                return;

            // function to close tooltip
            const delayCloseTooltip = function() {
                // if tooltip open and if mouse isn't over tooltip, close
                const tooltip = document.getElementById('tooltip');
                if (tooltip && !tooltip.matches(':hover'))
                    closeTooltip();
            };

            // run close function after delay
            this.closeTooltipTimer = window.setTimeout(
                delayCloseTooltip,
                options.delay
            );
        }

        // when link is focused (tabbed to)
        function onLinkFocus(event) {
            openTooltip(this);
        }

        // when link is touched on touch screen
        function onLinkTouch(event) {
            // attempt to force hover state on first tap always, and trigger
            // regular link click (and navigation) on second tap
            if (event.target === document.activeElement)
                event.target.click();
            else {
                document.activeElement.blur();
                event.target.focus();
            }
            if (event.cancelable)
                event.preventDefault();
            event.stopPropagation();
            return false;
        }

        // when mouse is clicked anywhere in window
        function onClick(event) {
            closeTooltip();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'tooltip_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'tooltip_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeTooltip();
                    break;
            }
        }

        // when window is resized or zoomed
        function onResize() {
            closeTooltip();
        }

        // get all links of types we wish to handle
        function getLinks() {
            const queries = [];
            // exclude buttons, anchor links, toc links, etc
            const exclude =
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            queries.push('a[href^="#ref-"]' + exclude); // citation links
            queries.push('a[href^="#fig:"]' + exclude); // figure links
            const query = queries.join(', ');
            return document.querySelectorAll(query);
        }

        // get links with same target, get index of link in set, get total
        // same links
        function getSameLinks(link) {
            const sameLinks = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    sameLinks.push(otherLink);
            }

            return {
                elements: sameLinks,
                index: sameLinks.indexOf(link),
                total: sameLinks.length
            };
        }

        // open tooltip
        function openTooltip(link) {
            // delete tooltip if it exists, start fresh
            closeTooltip();

            // make tooltip element
            const tooltip = makeTooltip(link);

            // if source couldn't be found and tooltip not made, exit
            if (!tooltip)
                return;

            // make navbar elements
            const navBar = makeNavBar(link);
            if (navBar)
                tooltip.firstElementChild.appendChild(navBar);

            // attach tooltip to page
            document.body.appendChild(tooltip);

            // position tooltip
            const position = function() {
                positionTooltip(link);
            };
            position();

            // if tooltip contains images, position again after they've loaded
            const imgs = tooltip.querySelectorAll('img');
            for (const img of imgs)
                img.addEventListener('load', position);
        }

        // close (delete) tooltip
        function closeTooltip() {
            const tooltip = document.getElementById('tooltip');
            if (tooltip)
                tooltip.remove();
        }

        // make tooltip
        function makeTooltip(link) {
            // get target element that link points to
            const source = getSource(link);

            // if source can't be found, exit
            if (!source)
                return;

            // create new tooltip
            const tooltip = document.createElement('div');
            tooltip.id = 'tooltip';
            const tooltipContent = document.createElement('div');
            tooltipContent.id = 'tooltip_content';
            tooltip.appendChild(tooltipContent);

            // make copy of source node and put in tooltip
            const sourceCopy = makeCopy(source);
            tooltipContent.appendChild(sourceCopy);

            // attach mouse event listeners
            tooltip.addEventListener('click', onTooltipClick);
            tooltip.addEventListener('mousedown', onTooltipClick);
            tooltip.addEventListener('touchstart', onTooltipClick);
            tooltip.addEventListener('mouseleave', onTooltipUnhover);

            // (for interaction with lightbox plugin)
            // transfer click on tooltip copied img to original img
            const sourceImg = source.querySelector('img');
            const sourceCopyImg = sourceCopy.querySelector('img');
            if (sourceImg && sourceCopyImg) {
                const clickImg = function() {
                    sourceImg.click();
                    closeTooltip();
                };
                sourceCopyImg.addEventListener('click', clickImg);
            }

            return tooltip;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // when tooltip is clicked
        function onTooltipClick(event) {
            // when user clicks on tooltip, stop click from transferring
            // outside of tooltip (eg, click off to close tooltip, or eg click
            // off to unhighlight same refs)
            event.stopPropagation();
        }

        // when tooltip is unhovered
        function onTooltipUnhover(event) {
            if (options.clickClose === 'true')
                return;

            // make sure new mouse/touch/focus no longer over tooltip or any
            // element within it
            const tooltip = document.getElementById('tooltip');
            if (!tooltip)
                return;
            if (this.contains(event.relatedTarget))
                return;

            closeTooltip();
        }

        // make nav bar to go betwen prev/next instances of same reference
        function makeNavBar(link) {
            // find other links to the same source
            const sameLinks = getSameLinks(link);

            // don't show nav bar when singular reference
            if (sameLinks.total <= 1)
                return;

            // find prev/next links with same target
            const prevLink = getPrevLink(link, sameLinks);
            const nextLink = getNextLink(link, sameLinks);

            // create nav bar
            const navBar = document.createElement('div');
            navBar.id = 'tooltip_nav_bar';
            const text = sameLinks.index + 1 + ' of ' + sameLinks.total;

            // create nav bar prev/next buttons
            const prevButton = document.createElement('button');
            const nextButton = document.createElement('button');
            prevButton.id = 'tooltip_prev_button';
            nextButton.id = 'tooltip_next_button';
            prevButton.title =
                'Jump to the previous occurence of this item in the document [←]';
            nextButton.title =
                'Jump to the next occurence of this item in the document [→]';
            prevButton.classList.add('icon_button');
            nextButton.classList.add('icon_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;
            navBar.appendChild(prevButton);
            navBar.appendChild(document.createTextNode(text));
            navBar.appendChild(nextButton);

            // attach click listeners to buttons
            prevButton.addEventListener('click', function() {
                onPrevNextClick(link, prevLink);
            });
            nextButton.addEventListener('click', function() {
                onPrevNextClick(link, nextLink);
            });

            return navBar;
        }

        // get previous link with same target
        function getPrevLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if < 1
            let index;
            if (sameLinks.index - 1 >= 0)
                index = sameLinks.index - 1;
            else
                index = sameLinks.total - 1;
            return sameLinks.elements[index];
        }

        // get next link with same target
        function getNextLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if > total
            let index;
            if (sameLinks.index + 1 <= sameLinks.total - 1)
                index = sameLinks.index + 1;
            else
                index = 0;
            return sameLinks.elements[index];
        }

        // get element that is target of link or url hash
        function getSource(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector(
                '[id="' + id + '"], [name="' + id + '"]'
            );
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (hash.indexOf('#ref-') === 0)
                target = target.querySelector('p');
            else if (hash.indexOf('#fig:') === 0)
                target = target.parentNode;
            else if (hash.indexOf('#tbl:') === 0)
                return;

            return target;
        }

        // when prev/next arrow button is clicked
        function onPrevNextClick(link, prevNextLink) {
            if (link && prevNextLink)
                goToElement(prevNextLink, window.innerHeight * 0.5);
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // determine position to place tooltip based on link position in
        // viewport and tooltip size
        function positionTooltip(link, left, top) {
            const tooltipElement = document.getElementById('tooltip');
            if (!tooltipElement)
                return;

            // get convenient vars for position/dimensions of
            // link/tooltip/page/view
            link = getRectInPage(link);
            const tooltip = getRectInPage(tooltipElement);
            const view = getRectInPage();

            // horizontal positioning
            if (left)
                // use explicit value
                left = left;
            else if (link.left + tooltip.width < view.right)
                // fit tooltip to right of link
                left = link.left;
            else if (link.right - tooltip.width > view.left)
                // fit tooltip to left of link
                left = link.right - tooltip.width;
            // center tooltip in view
            else
                left = (view.right - view.left) / 2 - tooltip.width / 2;

            // vertical positioning
            if (top)
                // use explicit value
                top = top;
            else if (link.top - tooltip.height > view.top)
                // fit tooltip above link
                top = link.top - tooltip.height;
            else if (link.bottom + tooltip.height < view.bottom)
                // fit tooltip below link
                top = link.bottom;
            else {
                // center tooltip in view
                top = view.top + view.height / 2 - tooltip.height / 2;
                // nudge off of link to left/right if possible
                if (link.right + tooltip.width < view.right)
                    left = link.right;
                else if (link.left - tooltip.width > view.left)
                    left = link.left - tooltip.width;
            }

            tooltipElement.style.left = left + 'px';
            tooltipElement.style.top = top + 'px';
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get position of element relative to page
        function getRectInPage(element) {
            const rect = getRectInView(element);
            const body = getRectInView(document.body);

            const newRect = {};
            newRect.left = rect.left - body.left;
            newRect.top = rect.top - body.top;
            newRect.right = rect.right - body.left;
            newRect.bottom = rect.bottom - body.top;
            newRect.width = rect.width;
            newRect.height = rect.height;

            return newRect;
        }

        // (for interaction with accordion plugin)
        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // (for interaction with accordion plugin)
        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- jump to first plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds a button next to each reference entry,
        // figure, and table that jumps the page to the first occurrence of a
        // link to that item in the manuscript.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'jumpToFirst';

        // default plugin options
        const options = {
            // whether to add buttons next to reference entries
            references: 'true',
            // whether to add buttons next to figures
            figures: 'true',
            // whether to add buttons next to tables
            tables: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            if (options.references !== 'false')
                makeReferenceButtons();
            if (options.figures !== 'false')
                makeFigureButtons();
            if (options.tables !== 'false')
                makeTableButtons();
        }

        // when jump button clicked
        function onButtonClick() {
            const first = getFirstOccurrence(this.dataset.id);
            if (!first)
                return;

            // update url hash so navigating "back" in history will return
            // user to jump button
            window.location.hash = this.dataset.id;
            // scroll to link
            window.setTimeout(function() {
                goToElement(first, window.innerHeight * 0.5);
            }, 0);
        }

        // get first occurence of link to item in document
        function getFirstOccurrence(id) {
            let query = 'a';
            query += '[href="#' + id + '"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelector(query);
        }

        // add button next to each reference entry
        function makeReferenceButtons() {
            const references = document.querySelectorAll('div[id^="ref-"]');
            for (const reference of references) {
                // get reference id and element to add button to
                const id = reference.id;
                const container = reference.firstElementChild;
                const first = getFirstOccurrence(id);

                // if can't find link to reference, ignore
                if (!first)
                    continue;

                // make jump button
                let button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this reference in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.innerHTML = button.outerHTML + container.innerHTML;
                button = container.firstElementChild;
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeFigureButtons() {
            const figures = document.querySelectorAll('img[id^="fig:"]');
            for (const figure of figures) {
                // get figure id and element to add button to
                const id = figure.id;
                const container = figure.nextElementSibling;
                const first = getFirstOccurrence(id);

                // if can't find link to figure, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this figure in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeTableButtons() {
            const tables = document.querySelectorAll('a[name^="tbl:"]');
            for (const table of tables) {
                // get ref id and element to add button to
                const id = table.name;
                const container = table.nextElementSibling.querySelector(
                    'caption'
                );
                const first = getFirstOccurrence(id);

                // if can't find link to table, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this table in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
    <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
    <svg width="16" height="16" viewBox="0 0 320 512">
        <path
            fill="currentColor"
            d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
        ></path>
    </svg>
</template>
<!-- link highlight plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user hovers or
        // focuses a link, other links that have the same target will be
        // highlighted. It also makes it such that when clicking a link, the
        // target of the link (eg reference, figure, table) is briefly
        // highlighted.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'linkHighlight';

        // default plugin options
        const options = {
            // whether to also highlight links that go to external urls
            externalLinks: 'false',
            // whether user must click off to unhighlight instead of just
            // un-hovering
            clickUnhighlight: 'false',
            // whether to also highlight links that are unique
            highlightUnique: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach mouse and focus listeners to link
                link.addEventListener('mouseenter', onLinkFocus);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('mouseleave', onLinkUnhover);
            }

            // attach click and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('hashchange', onHashChange);

            // run hash change on window load in case user has navigated
            // directly to hash
            onHashChange();
        }

        // when link is focused (tabbed to) or hovered
        function onLinkFocus() {
            highlight(this);
        }

        // when link is unhovered
        function onLinkUnhover() {
            if (options.clickUnhighlight !== 'true')
                unhighlightAll();
        }

        // when the mouse is clicked anywhere in window
        function onClick(event) {
            unhighlightAll();
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                glowElement(target);
        }

        // get element that is target of link or url hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector(
                '[id="' + id + '"], [name="' + id + '"]'
            );
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (hash.indexOf('#fig:') === 0)
                target = target.parentNode;
            else if (hash.indexOf('#tbl:') === 0)
                target = target.nextElementSibling.querySelector('caption');

            return target;
        }

        // start glow sequence on an element
        function glowElement(element) {
            const startGlow = function() {
                onGlowEnd();
                element.dataset.glow = 'true';
                element.addEventListener('animationend', onGlowEnd);
            };
            const onGlowEnd = function() {
                element.removeAttribute('data-glow');
                element.removeEventListener('animationend', onGlowEnd);
            };
            startGlow();
        }

        // highlight link and all others with same target
        function highlight(link) {
            // force unhighlight all to start fresh
            unhighlightAll();

            // get links with same target
            if (!link)
                return;
            const sameLinks = getSameLinks(link);

            // if link unique and option is off, exit and don't highlight
            if (sameLinks.length <= 1 && options.highlightUnique !== 'true')
                return;

            // highlight all same links, and "select" (special highlight) this
            // one
            for (const sameLink of sameLinks) {
                if (sameLink === link)
                    sameLink.setAttribute('data-selected', 'true');
                else
                    sameLink.setAttribute('data-highlighted', 'true');
            }
        }

        // unhighlight all links
        function unhighlightAll() {
            const links = getLinks();
            for (const link of links) {
                link.setAttribute('data-selected', 'false');
                link.setAttribute('data-highlighted', 'false');
            }
        }

        // get links with same target
        function getSameLinks(link) {
            const results = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    results.push(otherLink);
            }
            return results;
        }

        // get all links of types we wish to handle
        function getLinks() {
            let query = 'a';
            if (options.externalLinks !== 'true')
                query += '[href^="#"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelectorAll(query);
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- table of contents plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin provides a "table of contents" (toc) panel on
        // the side of the document that allows the user to conveniently
        // navigate between sections of the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tableOfContents';

        // default plugin options
        const options = {
            // which types of elements to add links for, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3',
            // whether default behavior is to be closed ('false'), open
            // ('true'), or only open when screen wide enough to fit panel
            // ('auto'). note: still always starts closed when page loads.
            open: 'auto',
            // if list item is more than this many characters, text will be
            // truncated
            charLimit: '50',
            // whether or not to show bullets next to each toc item
            bullets: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // make toc panel and populate with entries (links to document
            // sections)
            const panel = makePanel();
            if (!panel)
                return;
            makeEntries(panel);
            document.body.insertBefore(panel, document.body.firstChild);

            closePanel();

            // attach click, scroll, and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('scroll', onScroll);
            window.addEventListener('hashchange', onScroll);
            window.addEventListener('keyup', onKeyUp);
            onScroll();

            // add class to push document body down out of way of toc button
            document.body.classList.add('toc_body_nudge');
        }

        // determine if screen wide enough to fit toc panel
        function isSmallScreen() {
            // in default theme:
            // 816px = 8.5in = width of "page" (<body>) element
            // 260px = min width of toc panel (*2 for both sides of <body>)
            return window.innerWidth < 816 + 260 * 2;
        }

        // open/close panel based on option and screen size
        function openOrClosePanel() {
            if (
                options.open === 'true' ||
                (options.open === 'auto' && !isSmallScreen())
            )
                openPanel();
            else
                closePanel();
        }

        // when mouse is clicked anywhere in window
        function onClick() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                openOrClosePanel();
        }

        // when window is scrolled or hash changed
        function onScroll() {
            highlightViewed();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            // close on esc
            if (event.key === 'Escape')
                closePanel();
        }

        // find entry of currently viewed document section in toc and highlight
        function highlightViewed() {
            const firstId = getFirstInView(options.typesQuery);

            // get toc entries (links), unhighlight all, then highlight viewed
            const list = document.getElementById('toc_list');
            if (!firstId || !list)
                return;
            const links = list.querySelectorAll('a');
            for (const link of links)
                link.dataset.viewing = 'false';
            const link = list.querySelector('a[href="#' + firstId + '"]');
            if (!link)
                return;
            link.dataset.viewing = 'true';
        }

        // get first or previous toc listed element in top half of view
        function getFirstInView(query) {
            // get all elements matching query and with id
            const elements = document.querySelectorAll(query);
            const elementsWithIds = [];
            for (const element of elements) {
                if (element.id)
                    elementsWithIds.push(element);
            }


            // get first or previous element in top half of view
            for (let i = 0; i < elementsWithIds.length; i++) {
                const element = elementsWithIds[i];
                const prevElement = elementsWithIds[Math.max(0, i - 1)];
                if (element.getBoundingClientRect().top >= 0) {
                    if (
                        element.getBoundingClientRect().top <
                        window.innerHeight / 2
                    )
                        return element.id;
                    else
                        return prevElement.id;
                }
            }
        }

        // make panel
        function makePanel() {
            // create panel
            const panel = document.createElement('div');
            panel.id = 'toc_panel';
            if (options.bullets === 'true')
                panel.dataset.bullets = 'true';

            // create header
            const header = document.createElement('div');
            header.id = 'toc_header';

            // create toc button
            const button = document.createElement('button');
            button.id = 'toc_button';
            button.innerHTML = document.querySelector(
                '.icon_th_list'
            ).innerHTML;
            button.classList.add('icon_button');

            // create header text
            const text = document.createElement('h3');
            text.innerHTML = 'Table of Contents';

            // create container for toc list
            const list = document.createElement('div');
            list.id = 'toc_list';

            // attach click listeners
            panel.addEventListener('click', onPanelClick);
            header.addEventListener('click', onHeaderClick);
            button.addEventListener('click', onButtonClick);

            // attach elements
            header.appendChild(button);
            header.appendChild(text);
            panel.appendChild(header);
            panel.appendChild(list);

            return panel;
        }

        // create toc entries (links) to each element of the specified types
        function makeEntries(panel) {
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements) {
                // do not add link if element doesn't have assigned id
                if (!element.id)
                    continue;

                // create link/list item
                const link = document.createElement('a');
                link.classList.add('toc_link');
                switch (element.tagName.toLowerCase()) {
                    case 'h1':
                        link.dataset.level = '1';
                        break;
                    case 'h2':
                        link.dataset.level = '2';
                        break;
                    case 'h3':
                        link.dataset.level = '3';
                        break;
                    case 'h4':
                        link.dataset.level = '4';
                        break;
                }
                link.title = element.innerText;
                let text = element.innerText;
                if (text.length > options.charLimit)
                    text = text.slice(0, options.charLimit) + '...';
                link.innerHTML = text;
                link.href = '#' + element.id;
                link.addEventListener('click', onLinkClick);

                // attach link
                panel.querySelector('#toc_list').appendChild(link);
            }
        }

        // when panel is clicked
        function onPanelClick(event) {
            // stop click from propagating to window/document and closing panel
            event.stopPropagation();
        }

        // when header itself is clicked
        function onHeaderClick(event) {
            togglePanel();
        }

        // when button is clicked
        function onButtonClick(event) {
            togglePanel();
            // stop header underneath button from also being clicked
            event.stopPropagation();
        }

        // when link is clicked
        function onLinkClick() {
            openOrClosePanel();
        }

        // open panel if closed, close if opened
        function togglePanel() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                closePanel();
            else
                openPanel();
        }

        // open panel
        function openPanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'true';
        }

        // close panel
        function closePanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'false';
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- th list icon -->

<template class="icon_th_list">
    <!-- modified from: https://fontawesome.com/icons/th-list -->
    <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
        <path
            fill="currentColor"
            d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- lightbox plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user clicks on an
        // image, the image fills the screen and the user can pan/drag/zoom
        // the image and navigate between other images in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'lightbox';

        // default plugin options
        const options = {
            // list of possible zoom/scale factors
            zoomSteps:
                '0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1,' +
                '1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8',
            // whether to fit image to view ('fit'), display at 100% and shrink
            // if necessary ('shrink'), or always display at 100% ('100')
            defaultZoom: 'fit',
            // whether to zoom in/out toward center of view ('true') or mouse
            // ('false')
            centerZoom: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <img> element
            const imgs = document.querySelectorAll('figure > img');
            let count = 1;
            for (const img of imgs) {
                img.classList.add('lightbox_document_img');
                img.dataset.number = count;
                img.dataset.total = imgs.length;
                img.addEventListener('click', openLightbox);
                count++;
            }

            // attach mouse and key listeners to window
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('keyup', onKeyUp);
        }

        // when mouse is moved anywhere in window
        function onWindowMouseMove(event) {
            window.mouseX = event.clientX;
            window.mouseY = event.clientY;
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'lightbox_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'lightbox_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeLightbox();
                    break;
            }
        }

        // open lightbox
        function openLightbox() {
            const lightbox = makeLightbox(this);
            if (!lightbox)
                return;

            blurBody(lightbox);
            document.body.appendChild(lightbox);
        }

        // make lightbox
        function makeLightbox(img) {
            // delete lightbox if it exists, start fresh
            closeLightbox();

            // create screen overlay containing lightbox
            const overlay = document.createElement('div');
            overlay.id = 'lightbox_overlay';

            // create image info boxes
            const numberInfo = document.createElement('div');
            const zoomInfo = document.createElement('div');
            numberInfo.id = 'lightbox_number_info';
            zoomInfo.id = 'lightbox_zoom_info';

            // create container for image
            const imageContainer = document.createElement('div');
            imageContainer.id = 'lightbox_image_container';
            const lightboxImg = makeLightboxImg(
                img,
                imageContainer,
                numberInfo,
                zoomInfo
            );
            imageContainer.appendChild(lightboxImg);

            // create bottom container for caption and navigation buttons
            const bottomContainer = document.createElement('div');
            bottomContainer.id = 'lightbox_bottom_container';
            const caption = makeCaption(img);
            const prevButton = makePrevButton(img);
            const nextButton = makeNextButton(img);
            bottomContainer.appendChild(prevButton);
            bottomContainer.appendChild(caption);
            bottomContainer.appendChild(nextButton);

            // attach top middle and bottom to overlay
            overlay.appendChild(numberInfo);
            overlay.appendChild(zoomInfo);
            overlay.appendChild(imageContainer);
            overlay.appendChild(bottomContainer);

            return overlay;
        }

        // make <img> object that is intuitively draggable and zoomable
        function makeLightboxImg(
            sourceImg,
            container,
            numberInfoBox,
            zoomInfoBox
        ) {
            // create copy of source <img>
            const img = sourceImg.cloneNode(true);
            img.classList.remove('lightbox_document_img');
            img.removeAttribute('id');
            img.removeAttribute('width');
            img.removeAttribute('height');
            img.style.position = 'unset';
            img.style.margin = '0';
            img.style.padding = '0';
            img.style.width = '';
            img.style.height = '';
            img.style.minWidth = '';
            img.style.minHeight = '';
            img.style.maxWidth = '';
            img.style.maxHeight = '';
            img.id = 'lightbox_img';

            // build sorted list of unique zoomSteps, always including a 100%
            let zoomSteps = [];
            const optionsZooms = options.zoomSteps.split(/[^0-9.]/);
            for (const optionZoom of optionsZooms) {
                const newZoom = parseFloat(optionZoom);
                if (newZoom && !zoomSteps.includes(newZoom))
                    zoomSteps.push(newZoom);
            }
            if (!zoomSteps.includes(1))
                zoomSteps.push(1);
            zoomSteps = zoomSteps.sort(function sortNumber(a, b) {
                return a - b;
            });

            // <img> object property variables
            let zoom = 1;
            let translateX = 0;
            let translateY = 0;
            let clickMouseX = undefined;
            let clickMouseY = undefined;
            let clickTranslateX = undefined;
            let clickTranslateY = undefined;

            updateNumberInfo();

            // update image numbers displayed in info box
            function updateNumberInfo() {
                numberInfoBox.innerHTML =
                    sourceImg.dataset.number + ' of ' + sourceImg.dataset.total;
            }

            // update zoom displayed in info box
            function updateZoomInfo() {
                let zoomInfo = zoom * 100;
                if (!Number.isInteger(zoomInfo))
                    zoomInfo = zoomInfo.toFixed(2);
                zoomInfoBox.innerHTML = zoomInfo + '%';
            }

            // move to closest zoom step above current zoom
            const zoomIn = function() {
                for (const zoomStep of zoomSteps) {
                    if (zoomStep > zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                updateTransform();
            };

            // move to closest zoom step above current zoom
            const zoomOut = function() {
                zoomSteps.reverse();
                for (const zoomStep of zoomSteps) {
                    if (zoomStep < zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                zoomSteps.reverse();

                updateTransform();
            };

            // update display of <img> based on scale/translate properties
            const updateTransform = function() {
                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                // get new width/height after scale
                const rect = img.getBoundingClientRect();
                // limit translate
                translateX = Math.max(translateX, -rect.width / 2);
                translateX = Math.min(translateX, rect.width / 2);
                translateY = Math.max(translateY, -rect.height / 2);
                translateY = Math.min(translateY, rect.height / 2);

                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                updateZoomInfo();
            };

            // fit <img> to container
            const fit = function() {
                // no x/y offset, 100% zoom by default
                translateX = 0;
                translateY = 0;
                zoom = 1;

                // widths of <img> and container
                const imgWidth = img.naturalWidth;
                const imgHeight = img.naturalHeight;
                const containerWidth = parseFloat(
                    window.getComputedStyle(container).width
                );
                const containerHeight = parseFloat(
                    window.getComputedStyle(container).height
                );

                // how much zooming is needed to fit <img> to container
                const xRatio = imgWidth / containerWidth;
                const yRatio = imgHeight / containerHeight;
                const maxRatio = Math.max(xRatio, yRatio);
                const newZoom = 1 / maxRatio;

                // fit <img> to container according to option
                if (options.defaultZoom === 'shrink') {
                    if (maxRatio > 1)
                        zoom = newZoom;
                } else if (options.defaultZoom === 'fit')
                    zoom = newZoom;

                updateTransform();
            };

            // when mouse wheel is rolled anywhere in container
            const onContainerWheel = function(event) {
                if (!event)
                    return;

                // let ctrl + mouse wheel to zoom behave as normal
                if (event.ctrlKey)
                    return;

                // prevent normal scroll behavior
                event.preventDefault();
                event.stopPropagation();

                // point around which to scale img
                const viewRect = container.getBoundingClientRect();
                const viewX = (viewRect.left + viewRect.right) / 2;
                const viewY = (viewRect.top + viewRect.bottom) / 2;
                const originX = options.centerZoom === 'true' ? viewX : mouseX;
                const originY = options.centerZoom === 'true' ? viewY : mouseY;

                // get point on image under origin
                const oldRect = img.getBoundingClientRect();
                const oldPercentX = (originX - oldRect.left) / oldRect.width;
                const oldPercentY = (originY - oldRect.top) / oldRect.height;

                // increment/decrement zoom
                if (event.deltaY < 0)
                    zoomIn();
                if (event.deltaY > 0)
                    zoomOut();

                // get offset between previous image point and origin
                const newRect = img.getBoundingClientRect();
                const offsetX =
                    originX - (newRect.left + newRect.width * oldPercentX);
                const offsetY =
                    originY - (newRect.top + newRect.height * oldPercentY);

                // translate image to keep image point under origin
                translateX += offsetX;
                translateY += offsetY;

                // perform translate
                updateTransform();
            };

            // when container is clicked
            function onContainerClick(event) {
                // if container itself is target of click, and not other
                // element above it
                if (event.target === this)
                    closeLightbox();
            }

            // when mouse button is pressed on image
            const onImageMouseDown = function(event) {
                // store original mouse position relative to image
                clickMouseX = window.mouseX;
                clickMouseY = window.mouseY;
                clickTranslateX = translateX;
                clickTranslateY = translateY;
                event.stopPropagation();
                event.preventDefault();
            };

            // when mouse button is released anywhere in window
            const onWindowMouseUp = function(event) {
                // reset original mouse position
                clickMouseX = undefined;
                clickMouseY = undefined;
                clickTranslateX = undefined;
                clickTranslateY = undefined;

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mouseup', onWindowMouseUp);
            };

            // when mouse is moved anywhere in window
            const onWindowMouseMove = function(event) {
                if (
                    clickMouseX === undefined ||
                    clickMouseY === undefined ||
                    clickTranslateX === undefined ||
                    clickTranslateY === undefined
                )
                    return;

                // offset image based on original and current mouse position
                translateX = clickTranslateX + window.mouseX - clickMouseX;
                translateY = clickTranslateY + window.mouseY - clickMouseY;
                updateTransform();
                event.preventDefault();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mousemove', onWindowMouseMove);
            };

            // when window is resized
            const onWindowResize = function(event) {
                fit();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('resize', onWindowResize);
            };

            // attach the necessary event listeners
            img.addEventListener('dblclick', fit);
            img.addEventListener('mousedown', onImageMouseDown);
            container.addEventListener('wheel', onContainerWheel);
            container.addEventListener('mousedown', onContainerClick);
            container.addEventListener('touchstart', onContainerClick);
            window.addEventListener('mouseup', onWindowMouseUp);
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('resize', onWindowResize);

            // run fit() after lightbox atttached to document and <img> Loaded
            // so needed container and img dimensions available
            img.addEventListener('load', fit);

            return img;
        }

        // make caption
        function makeCaption(img) {
            const caption = document.createElement('div');
            caption.id = 'lightbox_caption';
            const captionSource = img.nextElementSibling;
            if (captionSource.tagName.toLowerCase() === 'figcaption') {
                const captionCopy = makeCopy(captionSource);
                caption.innerHTML = captionCopy.innerHTML;
            }

            caption.addEventListener('touchstart', function(event) {
                event.stopPropagation();
            });

            return caption;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // make button to jump to previous image in document
        function makePrevButton(img) {
            const prevButton = document.createElement('button');
            prevButton.id = 'lightbox_prev_button';
            prevButton.title = 'Jump to the previous image in the document [←]';
            prevButton.classList.add('icon_button', 'lightbox_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;

            // attach click listeners to button
            prevButton.addEventListener('click', function() {
                getPrevImg(img).click();
            });

            return prevButton;
        }

        // make button to jump to next image in document
        function makeNextButton(img) {
            const nextButton = document.createElement('button');
            nextButton.id = 'lightbox_next_button';
            nextButton.title = 'Jump to the next image in the document [→]';
            nextButton.classList.add('icon_button', 'lightbox_button');
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;

            // attach click listeners to button
            nextButton.addEventListener('click', function() {
                getNextImg(img).click();
            });

            return nextButton;
        }

        // get previous image in document
        function getPrevImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if < 1
            if (index - 1 >= 0)
                index--;
            else
                index = imgs.length - 1;
            return imgs[index];
        }

        // get next image in document
        function getNextImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if > total
            if (index + 1 <= imgs.length - 1)
                index++;
            else
                index = 0;
            return imgs[index];
        }

        // close lightbox
        function closeLightbox() {
            focusBody();

            const lightbox = document.getElementById('lightbox_overlay');
            if (lightbox)
                lightbox.remove();
        }

        // make all elements behind lightbox non-focusable
        function blurBody(overlay) {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.tabIndex = -1;
            document.body.classList.add('body_no_scroll');
        }

        // make all elements focusable again
        function focusBody() {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.removeAttribute('tabIndex');
            document.body.classList.remove('body_no_scroll');
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- attributes plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows arbitrary HTML attributes to be attached
        // to (almost) any element. Place an HTML comment inside or next to the
        // desired element in the format <!-- $attribute="value" -->

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'attributes';

        // default plugin options
        const options = {
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // get list of comments in document
            const comments = findComments();

            for(const comment of comments)
                if (comment.parentElement)
                    addAttributes(
                        comment.parentElement,
                        comment.nodeValue.trim()
                    );
        }

        // add html attributes to specified element based on string of 
        // html attributes and values
        function addAttributes(element, text) {
            // regex's for finding attribute/value pairs in the format of
            // attribute="value" or attribute='value
            const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
            const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

            // loop through attribute/value pairs
            let match;
            while(match = text.match(regex2) || text.match(regex1)) {
                // get attribute and value from regex capture groups
                let attribute = match[1];
                let value = match[2];

                // remove from string
                text = text.substring(match.index + match[0].length);

                if (!attribute || !value)
                    break;

                // set attribute of parent element
                try {
                    element.setAttribute(attribute, value);
                } catch(error) {
                    console.log(error);
                }

                // special case for colspan
                if (attribute === 'colspan')
                    removeTableCells(element, value);
            }
        }

        // get list of comment elements in document
        function findComments() {
            const comments = [];

            // iterate over comment nodes in document
            function acceptNode(node) {
                return NodeFilter.FILTER_ACCEPT;
            }
            const iterator = document.createNodeIterator(
                document.body,
                NodeFilter.SHOW_COMMENT,
                acceptNode
            );
            let node;
            while(node = iterator.nextNode())
                comments.push(node);

            return comments;
        }

        // remove certain number of cells after specified cell
        function removeTableCells(cell, number) {
            number = parseInt(number);
            if (!number)
                return;

            // remove elements
            for(; number > 1; number--) {
                if (cell.nextElementSibling)
                    cell.nextElementSibling.remove();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- math plugin configuration -->

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        "CommonHTML": { linebreaks: { automatic: true } },
        "HTML-CSS": { linebreaks: { automatic: true } },
        "SVG": { linebreaks: { automatic: true } },
        "fast-preview": { disabled: true }
    });
</script>

<!-- math plugin -->

<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML'>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'MathJax' allows the proper rendering of
    // math/equations written in LaTeX.

    // https://www.mathjax.org/
</script>
<!-- annotations plugin configuration -->

<script>
    window.hypothesisConfig = function() {
        return {
            branding: {
                accentColor: '#2196f3',
                appBackgroundColor: '#f8f8f8',
                ctaBackgroundColor: '#f8f8f8',
                ctaTextColor: '#000000',
                selectionFontFamily: 'Open Sans, Helvetica, sans serif',
                annotationFontFamily: 'Open Sans, Helvetica, sans serif'
            }
        };
    };
</script>

<!-- annotations plugin -->

<script src='https://hypothes.is/embed.js'>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'Hypothesis' allows public annotation of the
    // manuscript.

    // https://web.hypothes.is/
</script>
<!-- analytics plugin -->

<!-- copy and paste code from Google Analytics or similar service here -->
</body>
</html>
